---
title: "PEGA for pine"
author: "Katie Lotterhos"
date: "August 24, 2015"
output: html_document
---

This log is in Blog-style format (newest at top)

## August 30, 2015
Today I was working with Sam's raw (uncorrected for population structure) Spearman's rho for GWAS and GEAs, but I realized that I don't know how he coded his alleles.  Was it alphabetical (like bayenv) or by minor allele (like GWAS)????  UGGGGGGGGGGG!  This is driving me crazy.

## August 29, 2015
I realized that one problem with my approach is that, even though I am comparing architectures for super outliers based on raw values or values corrected for population structure, I am only plotting the values of rho and GWAS that have been corrected by population structure.  Sam updated the RESULTS files today, time to re-download.
```
scp klott@rogue.zoology.ubc.ca:/data/seqcap/adaptree2_seqcap/spruce/var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS .
scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS .
```

## August 28, 2015
Pine code is done, time to compare it to spruce.  

``` 
scp klott@rogue.zoology.ubc.ca:/data/seqcap/adaptree2_seqcap/spruce/flip.cor .
scp klott@rogue.zoology.ubc.ca:/data/seqcap/adaptree2_seqcap/spruce/var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS .

scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/super_outliers_per_variable_spruce_both_methods* .
```

## August 24, 2015
Download pine data to laptop.  I am in the directory /data of the project folder, and downloading these files the the "large_files" subfolder.  Files in that folder will not be synced to github, potentially causing it to crash.  You will have to download these files in order to run code on your laptop.  I'm not running it through hulk because of lack of visualization.

```
Katies-MacBook-Air:data katie$ scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/flip.cor large_files

Katies-MacBook-Air:data katie$ scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS large_files

Katies-MacBook-Air:data katie$ scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/super_outliers_per_variable_pine_both_methods_* large_files
```

So, the GWAS phenotype labels are different from the PE phenotype labels.  I think I figured them all out, except 
1) for the GWAS there is 
root_wt__shoot_wt_p
root_wt__shoot_wt_p_1

2) and for the PE there is
RootShoot_Ratio (matching the two above)
TotalWeight (not matching any GWAS)

From Sam:
* One is a ratio (root wt / shoot wt) and the other is a sum (root wt + shoot wt).  The two root-shoot things clearly should have been called something more meaningful. The order has been maintained for Haktan, so the first root_wt__shoot_wt_p is the ratio and the second one (_p1) is the sum. As for total weight I think that was Laura's way of getting rid of the root_shoot problem, and it probably means root wt + shoot wt. Clear as mud?

A similar problem exists for the GEs - 
in addition to the problem that some GEs were log-trasformed for the PE but not log-transformed for the GEAs.  Specifically log(MAP) and log(MSP).  Some of the PEs for MAP were comparably large, which makes me worry a bit because the GEAs were not very significant for this variable.  Should we have log transformed it?

I made files 
* EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGEA.csv
* EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGWAS.csv
that will have to be used to match names (ug.)

## August 24, 2015

This is the beginning of my attempt to integrate the phenotypic, genetic, and environmental data for the pine Adaptree results.  The rationale follows like this:

* Organisms adapt to their environment, and a measure of this is the correlation between measured phenotypes grown in a common garden and the environments that they came from (the PEA: phenotype-environment-association, Spearman's correlation)
* We can estimate the effect that a SNP has on that phenotype with GWAS (the genotype-phenotype association, a slope)
* With knowledge of the PEA and the GPA, we can visualize the genetic architecture of the genetic-environment-association (Spearman's rho from Bayenv2).

For each phenotype-environment association, here is the conceptual idea:

![PEGA sketch](201508PEGA.png)

If the GEA effects were polarized/adjusted based on the direction of the PEA and GWAS, then we could visualize this in the same way for all traits and environments.

### Some initial issues
* __From Kay around August 20__: So for the bayenv formatting I printed out the alleles in alphabetical order. This made the most sense at the time because it was the easiest to implement (I did not need to determine overall SNP frequency for any reason in the script) and the SNP table did not have the reference allele denoted any place.  But Haktan coded things in terms of allele frequency, so we will have to flip the signs for those that differed. (Kay made the file "flip" for that purpose, this will be an additional sign change prior to the polarization discussed above)

* __From Sam on August 24__: Turns out that the way that budset was measured in spruce in season 1 wasn't biologically representative (they all set bud in such a short window that the timing of sampling didn't catch much of it), but that was the trait we'd been using for gwas and phenotype-environment correlations, and that was probably responsible for the reason why spruce and pine had opposite correlations between phenotype and environment for budset. Pia figured that a better way to calculate budset for spruce would be to use the season 2 measurements, which were calculated a little differently. Using these values, pine and spruce have very similar correlations between budset and environment. So now Haktan is redoing the GWAS on this new budset trait for spruce (again) and I'll redo the super-outlier things (again), and the analysis (again). 