---
title: "PEGA for pine"
author: "Katie Lotterhos"
date: "August 24, 2015"
output: html_document
---

This log is in Blog-style format (newest at top)

## May 6, 2016

I had a breakthrough this week in how to analyze and visualize everything.  I identified two groups of genes - a temperature group (RED) and a moisture group (BLUE), that have sometimes similar and sometimes contrasting behavior to each other and to population structure.

These groups also help explain apparent trade-offs between cold injury and height.  In the phenotype data, we find that fall/winter cold injury (FWCI) decreases with mean annual temperature (MAT) and height (season 2, HS2) also decreases with MAT.  However, if we plot population averages of FWCI and HS2 against each other, we see an apparent negative correlation between these two traits, which suggests a tradeoff.  These two observations seem at odds with one another, because the correlation between the traits is negative, but both traits have positive associations with the same environmental variable.

However, the RED and BLUE groups show no such trade off - alleles at higher frequency in colder latitudes are also found at higher frequency in trees with less CI and smaller heights.  These groups are evolving against the genome-wide pattern for FWCI and HS2.



Overview:
- took Sam's list of super outliers - genes that have more outlier snps than predicted by random based on raw or corrected correlations at p9 level
- reduced list further to genes with SNPs in > 0.999 on raw rho (Group 1)
- I noticed two groups of behavior.  Chose variables that had strongest PE associations.
  - Group temp: subset of Group 1 that are raw outliers in lat and DD0
  - Group moisture: subset of Group 1 that are raw outliers in MAP and CMD

Red genes (LAT and DD0) Blue genes (MAP and CMD) in Bayenv2:
* Both read and blue decrease FWCI with latitude and DD0 (true in raw)
* Both decrease height and MGR with DD0 (true in raw)
* Red decrease FWCI with elevation, while blue increase FWCI with elevation (true in raw for WCI in elevation)
* Red increase FWSCI, MGR, height, and diam with MAP, while blue do the opposite (true for raw)
* red increase FWCI, MCG, height, and diam with FFP, while blue do the opposite (in raw associations blue do nothing)
* red and blue increase FWCI with CMD, and increase height and diam 

## April 20, 2016

I looked into one group of super outliers a bit more: the outliers on the FCI-Lat association in Bayenv.  A majority of them are from one contig that is on your super outlier list at p9:


tscaffold1962_232203_233472            comp60064_c0_seq1 (9 / 76 SNPs)
tscaffold1962_233534_237733            comp60064_c0_seq1 (21 / 76 SNPs)


And here is what they do in different comparisons (note that I didn’t flip for the negative PE association, so it looks like a reflection of what I showed you before).  If you look at the Bayenv Lat vs. Raw Lat plot, you can really see how these SNPs behave differently after Bayenv corrects for population structure.  

## April 19, 2016
Last night I had a skype meeting with Sam and Mike and Kay to bounce some ideas around.  I came out of that meeting with a plan, and some thoughts for moving forward:

1. Edit dataframe so all SNP effects are alphabetical

* Thought I did this before, but didn't - I was doing the cacluations as I went.  Need to do this first.
  
2. Identify groups of outliers. Focus on these groups for analysis.

* Group 1 a priori: top 1% of SNPs from each of Sam's super contigs
* Group 2 reinforcing: top outliers in Fall Cold Hardiness - Lat
* Group 3 reinforcing: second group of outliers in Fall Cold Hardiness and Eref/Ext/CMD, excluding group 2
* Group 4 antagonistic: antagonistic in Fall Cold Hardiness and MAP.  They might be from group 3, we'll see.
  
3. Output G2-4 outliers and share with Sam and Kay - are they interesting?

4.  Plot and analyze

   + order phenotypes and environments according to their abs(correlations) in a 17 x 22 grid, show dendogram
   + Do the following for Bayenv galaxies and Raw galaxies in each group (4 groups x 2 types = 8 graphs of 17 x 22 panels):
    + In each panel plot the galaxy plot with outliers highlighted, or with outlier bubbles (bubble size = # outliers / contig).  Add an x-axis to the plot indicating the direction of effects of the allele on the phenotype.For negative PE associations the x-axis is “flipped” for plotting. (Output plot to tiff and then plot tiff on grid). Add the PE association as text to panel.
    
   + Calculate the mean covariance in GE-PE among unlinked SNPs by boostrapping 1 SNP from each gcontig.  Add the mean covariance as text to panel.  Save covariance.
   
   + Calculate significance of covariance by sampling a group of SNPs the same number as gcontigs in the outlier SNPs and:

-sample a random gcontig
-sample a SNP from that random gcontig that is matched in allele frequency to outlier-SNP
-repeat above two steps creating a null distribution against which to test observed distribution.

## April 6, 2016
A couple weeks ago I presented results at an Adaptree meeting and people were interested, but it will take some convincing to show the approach is statistically valid.  I shared the "flipped" files on Dropbox with Ian and Kay, and realized that I have to document the flipping so we can determine the direction of allele effects.

First, the team sent me 'flip' files (for raw and GWAS, Bayenv was already coded this way) that the earliest alphabetical allele would be a 1 and the other allele would be 0.  I incorporated these into the dataframe with Sam's super outliers.  To get alleles coded in the same direction, first load the file and then multiply the relevant statistic column by the relevant "flip" column.  `flip_pine_gwas` is for Hatkan's gcta results, and `flip_pine_raw` is for Sam's raw correlations.

Example for pine:
```
### load dataframe
results_pine <- load("var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV_flipped_withSuper.Rdata")

### Example for eFFP
GWAS_effect <- results_pine$eFFP_snp_effect_GCTA*results_pine$flip_pine_gwas
  # this flip column applies to any GCTA results, including environment
Bayenv_effect_rho <- mean(results_pine$eFFP_rho1, results_pine$eFFP_rho2) 
  # no need to flip, but Bayenv outputs two rhos for some reason so I average them

Raw_effect_rho <- results_pine$eFFP_raw_rho*results_pine$flip_pine_raw
  # this flip column applies to any Bayenv results, including phenotype
```
After flipping,  if the SNP is an "A/T" and the association is positive, A would increase the phenotype/environment and T would decrease it. If the association is negative, A would decrease the phenotype/environment and T would increase it. If the SNP is a "C/G," the same logic applies.


## Feb 23, 2016
Lots of major updates, have been working on this on and off the last couple months.
Analyses have been focusing on the Bayenv Gen-Phen and Gen-Envi, because effect sizes (rho) are standardized and this makes it easier to interpret.  Calculate a "Genome Score" for each Envi-Pheno combo, based on sum of GE*GP association.  To understand plieotropy, I also calculate the same score for subset of outlier SNPs (top 0.1%) in one Envi-Pheno combo in each Envi-Pheno combos, to reveal if they also have positive, or potentially negative, effects on that phenotype in that environment.

## Dec 27, 2015
Sam and Haktan decided to run the phenotypes with bayenv and the environments with gcta (originally we just did the converse) because of the strong evidence that they were correcting for population structure in different ways.  

I downloaded the new files:

`var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV`

`var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV`

And ran `src/incorporateFlip.R` to merge the data frames with the flip variables and re-wrote them to:

`var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV_flipped`

`var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV_flipped`

I've spent the last week or so re-working the code to make `figures_new` plots of all correlations among variables.

## Sept 30, 2015
### More on the allele coding problem
It became evident that alleles for GWAS were not coded the way Hatkan thought.  After a series of emails I think we figure it out.
From Hatkan: 
```
Spruce
(i) rogue:/data/seqcap/adaptree2_seqcap/spruce/final_tables_filt/var_out_598ind_allhet_ALL.table.contig_filt10_filtGATK2_p95_h0.7.haktan
(ii) rogue:/data/hsuren/GWAS/rawdata_hapmap.plk.ped
(iii) rogue:/data/hsuren/GWAS/data.bed (Input file for GCTA)

Pine
(i) hulk:/data/seqcap/pine/bwa_pseudo/round2_bams/final_tables_filt/var_out_pine688_allhet_ALL.table.contig.filt10.passFILT2_p95_h0.7.finalPops279.haktan
(ii) hulk:/data/hsuren/GWAS/data.plk.ped
(iii) hulk:/data/hsuren/GWAS/data.bed (Input file for GCTA)

To convert from HapMap (i) to Plink (ii) (used TASSEL program w/ following code)
./run_pipeline.pl -fork1 -h /data/hsuren/GWAS/rawdata_hapmap -export -exportType Plink -runfork1

To convert from Plink (ii) to binary plink (bed) (iii) (used plink program w/ following code)
./plink --file ../TASSEL/data.plk --out data --make-bed

Just to confirm again, GCTA converted the data.bed file before GWAS analysis (11 for the homozygote of the first allele (the first allele in the *.bim file) and 00 for the homozygote of the second allele). The code (GCTA) is on github https://github.com/cooljeanius/gcta
```

* the *.bim file was used to code alleles for gcta (Hatkan)
* So if the first column in the *bim file is a “G” and the second column is an “A" it coded a GG as a 11 and a AA as a 00.
* So if the first letter is also first alphabetically, the coding is correct. else, the coding is incorrect

##### I downloaded the *.bim files and converted them myself:
Pine: from`hulk:/data/hsuren/GWAS/data.bim` to `data/large_files/pine_gwas.bim`, corresponding flip file is `flip_pine_v2.cor`

Spruce: from`rogue:/data/hsuren/GWAS/data.bim` to `data/large_files/spruce_gwas.bim`, corresponding flip file is `flip_spruce_v2.cor`

### Checksums on uploaded data files
The pine results are weird, several thousand SNPs have the same value in the data.
```
# On my mac
Katies-MacBook-Air:large_files katie$ shasum var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS
9551324646546c4814760703f09dd26b56f08555  

# On hulk
shasum /data/seqcap/pine/bwa_pseudo/round2_bams/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS
9551324646546c4814760703f09dd26b56f08555

The two are identical.  Checking if problem comes from merger.
```

## Sept 26, 2015
### More on the allele coding problem
GWAS (Hatkan) 1 for major allele, 0 for minor allele.  “Major allele” means the allele in higher frequency for given SNP.

Bayenv/LFMM (Kay) 1 for earliest letter in alphabet, 0 for other allele

flip (Kay) 1 if major allele was first letter and -1 if major allele came second in alphabetical order.  Kay's files are the "flip.cor" files downloaded below.

raw correlations (Sam) "I read through the genotypes for a given SNP, and the first homozygote encountered was given a value of "0", while the second homozygote encountered was given a value of "2", (hets = "1"). This means that it wasn't alphabetical or reference/alternate or minor/major or anything else that made sense but purely random based on the first genotype encountered. "

For example:

situation  	gwas	bayenv	flip
A/T A-major  	1/0  		1/0		1
A/T T-major	0/1		1/0		-1
C/G C-major	1/0		1/0		1
C/G G-major	0/1		1/0		-1

Sam's flip files are in (downloaded today):
`/data/seqcap/pine/bwa_pseudo/round2_bams/var_out_GATK3_allhet_pine688_ALL.table_filt10_p95_het7_passFILT2.2_LDformat_TOFLIP_alphabet_lowest_eq_1`

`/data/seqcap/adaptree2_seqcap/spruce/var_out_GATK3_spruce_ALL.table_filt10_p95_het7_passFILT2.3_LDformat_TOFLIP_alphabet_lowest_eq_1`

(Sam: Whenever there is a "1", it means that the coding of the SNP is currently in the format of {lowest alphabetical == 1, higest alphabetical == 0}. Whenvever there is a "-1", it means that the coding is opposite to this. 

NOTE: these SNPs are not necessarily in the same order as any tables, so a "merge" or similar command should be used to get the order right.)


## August 30, 2015
Today I was working with Sam's raw (uncorrected for population structure) Spearman's rho for GWAS and GEAs, but I realized that I don't know how he coded his alleles.  Was it alphabetical (like bayenv) or by minor allele (like GWAS)????  UGGGGGGGGGGG!  This is driving me crazy.

I still managed to get some cool results, see email to Sam and Kay sent today.

## August 29, 2015
I realized that one problem with my approach is that, even though I am comparing architectures for super outliers based on raw values or values corrected for population structure, I am only plotting the values of rho and GWAS that have been corrected by population structure.  Sam updated the RESULTS files today, time to re-download.
```
scp klott@rogue.zoology.ubc.ca:/data/seqcap/adaptree2_seqcap/spruce/var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS .
scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS .
```

## August 28, 2015
Pine code is done, time to compare it to spruce.  

``` 
scp klott@rogue.zoology.ubc.ca:/data/seqcap/adaptree2_seqcap/spruce/flip.cor .
scp klott@rogue.zoology.ubc.ca:/data/seqcap/adaptree2_seqcap/spruce/var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS .

scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/super_outliers_per_variable_spruce_both_methods* .
```

## August 24, 2015
Download pine data to laptop.  I am in the directory /data of the project folder, and downloading these files the the "large_files" subfolder.  Files in that folder will not be synced to github, potentially causing it to crash.  You will have to download these files in order to run code on your laptop.  I'm not running it through hulk because of lack of visualization.

```
Katies-MacBook-Air:data katie$ scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/flip.cor large_files

Katies-MacBook-Air:data katie$ scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS large_files

Katies-MacBook-Air:data katie$ scp klott@hulk.zoology.ubc.ca:/data/seqcap/pine/bwa_pseudo/round2_bams/super_outliers_per_variable_pine_both_methods_* large_files
```

So, the GWAS phenotype labels are different from the PE phenotype labels.  I think I figured them all out, except 
1) for the GWAS there is 
root_wt__shoot_wt_p
root_wt__shoot_wt_p_1

2) and for the PE there is
RootShoot_Ratio (matching the two above)
TotalWeight (not matching any GWAS)

From Sam:
* One is a ratio (root wt / shoot wt) and the other is a sum (root wt + shoot wt).  The two root-shoot things clearly should have been called something more meaningful. The order has been maintained for Haktan, so the first root_wt__shoot_wt_p is the ratio and the second one (_p1) is the sum. As for total weight I think that was Laura's way of getting rid of the root_shoot problem, and it probably means root wt + shoot wt. Clear as mud?

A similar problem exists for the GEs - 
in addition to the problem that some GEs were log-trasformed for the PE but not log-transformed for the GEAs.  Specifically log(MAP) and log(MSP).  Some of the PEs for MAP were comparably large, which makes me worry a bit because the GEAs were not very significant for this variable.  Should we have log transformed it?

I made files 
* EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGEA.csv
* EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGWAS.csv
that will have to be used to match names (ug.)

## August 24, 2015

This is the beginning of my attempt to integrate the phenotypic, genetic, and environmental data for the pine Adaptree results.  The rationale follows like this:

* Organisms adapt to their environment, and a measure of this is the correlation between measured phenotypes grown in a common garden and the environments that they came from (the PEA: phenotype-environment-association, Spearman's correlation)
* We can estimate the effect that a SNP has on that phenotype with GWAS (the genotype-phenotype association, a slope)
* With knowledge of the PEA and the GPA, we can visualize the genetic architecture of the genetic-environment-association (Spearman's rho from Bayenv2).

For each phenotype-environment association, here is the conceptual idea:

![PEGA sketch](201508PEGA.png)

If the GEA effects were polarized/adjusted based on the direction of the PEA and GWAS, then we could visualize this in the same way for all traits and environments.

### Some initial issues
* __From Kay around August 20__: So for the bayenv formatting I printed out the alleles in alphabetical order. This made the most sense at the time because it was the easiest to implement (I did not need to determine overall SNP frequency for any reason in the script) and the SNP table did not have the reference allele denoted any place.  But Haktan coded things in terms of allele frequency, so we will have to flip the signs for those that differed. (Kay made the file "flip" for that purpose, this will be an additional sign change prior to the polarization discussed above)

* __From Sam on August 24__: Turns out that the way that budset was measured in spruce in season 1 wasn't biologically representative (they all set bud in such a short window that the timing of sampling didn't catch much of it), but that was the trait we'd been using for gwas and phenotype-environment correlations, and that was probably responsible for the reason why spruce and pine had opposite correlations between phenotype and environment for budset. Pia figured that a better way to calculate budset for spruce would be to use the season 2 measurements, which were calculated a little differently. Using these values, pine and spruce have very similar correlations between budset and environment. So now Haktan is redoing the GWAS on this new budset trait for spruce (again) and I'll redo the super-outlier things (again), and the analysis (again). 