---
title: "Galaxy Plots on Sims"
author: "Katie Lotterhos"
date: "1/16/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Evaluating Galaxy plots with known simulation data

Recently I have been exploring the idea of visualizing SNP effect sizes to multiple environmental variables. Here, I am extending the results from the I am using the Lotterhos and Whitlock (2015) simulations to explore this concept in a known scenario. Basically, we want to know if the two distinct groups of behavior we observed in the dataset could be generated by chance if there is only one environmental variable driving the adapatation.

I have chosen three replicate datasets from the LW2015 simulations, that were simulated range expansions in a 1 refugia model, with 90 populations sampled and 20 individuals per population. This demography is similar to the range expansion we expect in lodgepole pine, which is the species and dataset I am applying the concept to (Adaptree data). The total sample size is also similar - in the paper I think we have ~600 diploids (1200 chromosomes), whereas here we have 1800 haploid chromosomes.

In LW2015, we simulated adaptation of loci to a single environmental variable (E_main).

Here, I am going to create two more environmental variables: one that is positively correlated with the selective environment (E_falsePlus), and one that is negatively correlated with the selective environment (E_falseMinus).

With this data, I am going to essentially repeat the steps that I took for the Adaptree data:

1) calculate uncorrected Spearman's rho association between allele frequency ($f$) and environment ($E_i$) for E_main, E_falsePlus, and E_falseMinus.

2) choose outliers based on uncorrected Spearman's rho as those that have P-values less than the Bonferroni correction. Hereafter these will be called the "candidate set".

3) use Galaxy plots to visualize the candidate set in 2 dimensional space with 3 plots:

* E_main vs. E_falsePlus
* E_main vs. E_falseMinus
* E_falsePlus vs. E_falseMinus

I will conduct this exercise for datasets with just neutral loci (9900) and for datasets with neutral loci and loci simulated under varying strengths of selection (9900 and 100).

```{r, eval=FALSE}
setwd("Users/katie/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/simulations/")
```


### Load data

```{r}
snps <- read.table("data/LW2015data/1R_R90_1351142954_453_1_NumPops=90_NumInd=20.lfmm")
env <- read.table("data/LW2015data/1R_R90_1351142954_453_1_NumPops=90_NumInd=20.env")
df <- read.table("data/LW2015data/1R_R90_1351142954_453_1_NumPops=90_NumInd=20Bayenv2LFMMpca.Cpval", header=TRUE)

dim(snps)
dim(env)

snps <- read.table("data/LW2015data/IBD_R90_1351142954_453_3_NumPops=90_NumInd=20.lfmm")
env <- read.table("data/LW2015data/IBD_R90_1351142954_453_3_NumPops=90_NumInd=20.env")
df <- read.table("data/LW2015data/IBD_R90_1351142954_453_3_NumPops=90_NumInd=20Bayenv2LFMMpca.Cpval", header=TRUE)

dim(snps)
dim(env)
```


### Create correlated environments

Here, E_falsePlus and E_falseMinus both have Spearman's correlations with E_Main that are about 0.5. E_falsePlus and E_falseMinus have a correlation of about 0.3 with each other.

```{r}
  head(env[,1], 40)
  # note that this is individual level data, but each population only has one environment measure

  pops <- seq(1, 1800-19, by=20)
  length(pops)
  
  E_main <- env[pops,1]
  
  ### Falsely positively correlated variable
  E_falsePlus <- E_main * 2 + rnorm(length(E_main), sd=8)
  plot(E_main, E_falsePlus)
  cor.test(E_main, E_falsePlus, method="spearman", exact=FALSE)
  
  ### Falsely negatively correlated variable
  E_falseMinus <- E_main * -2 + rnorm(length(E_main), sd=8)
  plot(E_main, E_falseMinus)
  cor.test(E_main, E_falseMinus, method="spearman", exact=FALSE)
 
  env[,2] <- rep(E_falsePlus, each=20)
  env[,3] <- rep(E_falseMinus, each=20)
  
 
  # The two false variables are also correlated:
  cor.test(E_falsePlus, E_falseMinus, method="spearman", exact=FALSE)
  
  # Plot of the correlation structure among environments:
   plot(env)
```


### Calculate rho for all 3 environments

Create a matrix to hold Spearman's rho and another matrix to hold P-values, where each row is a SNP and each column is an environmental variable

```{r}
TrueEnv.rho <- matrix(NA, ncol(snps), ncol(env))
TrueEnv.P <- matrix(NA, ncol(snps), ncol(env))

for (i in 1:ncol(snps)){
  for (j in 1:ncol(env)){
    test <- cor.test(snps[,i], env[,j], method="spearman", exact=FALSE)
    TrueEnv.rho[i,j] <- test$estimate
    TrueEnv.P[i,j] <- test$p.value
  }
}
```

### Get indices of neutral and selected SNPs
(not shown)
```{r, echo=FALSE}
Neut = df$s_high==0 & df$UseSNP==TRUE
sum(Neut)

Sel = df$s_high>0 & df$UseSNP==TRUE
sum(Sel)
which(Sel)

Neut_ind <- 1:9900
Sel_ind <- 9901:ncol(snps)
```

```{r, echo=FALSE}

par(mfcol=c(2,3))
hist(TrueEnv.rho[Neut_ind, 1], xlim=c(-1, 1), main="Neut True env")
hist(TrueEnv.rho[Sel_ind, 1], xlim=c(-1, 1), main="Neut True env")

hist(TrueEnv.rho[Neut_ind, 2], xlim=c(-1, 1), main="Neut False env Plue")
hist(TrueEnv.rho[Sel_ind, 2], xlim=c(-1, 1), main="Sel False env Plus")

hist(TrueEnv.rho[Neut_ind, 3], xlim=c(-1, 1), main="Neut False env Minus")
hist(TrueEnv.rho[Sel_ind, 3], xlim=c(-1, 1), main="Sel False env Plus")
```

The above plot shows the distribution of Spearman's rho for the three environments, with neutral loci in the top row and selected loci in the bottom row.


### How many are significant

I determined which loci were "outliers" by the criteria that they had to have p-values from Spearman's rho less than expected by Bonferroni correction.

The false positive rate is really high in the "main" environment (See: "Number false positives in landscape sims" in R code). This might be because the main environment correlates with the population structure; e.g. the range expansion was south the north, and so is the environmental cline. (This makes me think we should use more stringent criteria for the pine paper, since the Temperature-related variables are the ones that correlate the strongest with population structure and those are the ones we see the most outliers in.)

```{r}
(Bonferroni_P_N <- -log10(0.05/(10000*3)))

### Number false positives in landscape sims
colSums(-log10(TrueEnv.P[Neut_ind,])>Bonferroni_P_N)
  # The false positive rate is really high in the "main" environment (first column). This might be because the causal environment correlates with the population structure; e.g. the range expansion was south the north, and so is the environmental cline.


### Number true positives in landscape sims
colSums(-log10(TrueEnv.P[Sel_ind,])>Bonferroni_P_N)
  #(total selected loci in this dataset is 96)

Is.Sig <- -log10(TrueEnv.P)>Bonferroni_P_N
dim(Is.Sig)

# As a first pass, we call a snp an outlier if it is an outlier in any one environmental variable
Is.Out <- (rowSums(Is.Sig)>0)

# The number of "outliers" that are neutral is slightly larger than the number of outliers in the "main" environment:
sum(Is.Out[Neut_ind])

# Number of "outliers" that are under selection is the same as the number in the "main" environment
sum(Is.Out[Sel_ind])
```


### Galaxy plots

```{r, fig.height=8, echo=FALSE}
par(mfcol=c(2,2))
#TrueEnv.rho <- data.frame(TrueEnv.rho)

### Main vs. Envi_Plus

plot(TrueEnv.rho[Neut_ind,1],TrueEnv.rho[Neut_ind,2], col=adjustcolor(Is.Sig[Neut_ind,]+1, alpha.f=0.1), xlim=c(-0.5,0.5), ylim=c(-0.5,0.5), pch=20, xlab="Main environment", ylab="False envi 1")

#points(TrueEnv.rho[Sel_ind,1], TrueEnv.rho[Sel_ind,2], col="blue", pch=25, bg="lightblue")

### Main vs. Envi_Minus

plot(TrueEnv.rho[Neut_ind,1],TrueEnv.rho[Neut_ind,3], col=adjustcolor(Is.Sig[Neut_ind,]+1, alpha.f=0.1), xlim=c(-0.5,0.5), ylim=c(-0.5,0.5), pch=20, xlab="Main environment", ylab="False envi 2")

#points(TrueEnv.rho[Sel_ind,1], TrueEnv.rho[Sel_ind,3], col="blue", pch=25, bg="lightblue")

### Envi_Plus vs. Envi_Minus

plot(TrueEnv.rho[Neut_ind,2],TrueEnv.rho[Neut_ind,3], col=adjustcolor(Is.Sig[Neut_ind,]+1, alpha.f=0.1), xlim=c(-0.5,0.5), ylim=c(-0.5,0.5), pch=20, xlab="False envi 1", ylab="False envi 2")

#points(TrueEnv.rho[Sel_ind,2], TrueEnv.rho[Sel_ind,3], col="blue", pch=25, bg="lightblue")

plot(1,1, col="white")
legend(x = 0.8, y=1.2, legend = c("Neutral", "Neutral false positive"), pch=c(20,20,25), col=c("black", "red"))

#legend(x = 0.8, y=1.2, legend = c("Neutral", "Neutral false positive", "Selected locus"), pch=c(20,20,25), col=c("black", "red", "blue"), bg=c(NA, NA, "lightblue"))

```

The main result is the false positives are expected to follow the orientation of the genome-wide null distribution, no matter what combinations of environmental variables we look at.

This makes me think maybe an eigenvector would be a better test of behavior rather than a covariance, with the null hypothesis that the eigenvector of a focal group of SNPs is significantly different from the genome-wide expected direction. This null hypothesis should hold true in any combination of environmental variables.

**In the Adaptree dataset, we see two distinct groups of behavior at non-overlapping groups of SNPs, which was not observed here** 

In these simulations, true positives also follow the orientation of the genome-wide null distribution, but are even moreso 
"outliers". This illustrates a potential problem with the eigenvector approach, which would overlook "true" positives that are only adapting to a single environmental variable.

```{r, fig.height=8, echo=FALSE}
par(mfcol=c(2,2))
#TrueEnv.rho <- data.frame(TrueEnv.rho)

### Main vs. Envi_Plus

plot(TrueEnv.rho[Neut_ind,1],TrueEnv.rho[Neut_ind,2], col=adjustcolor(Is.Sig[Neut_ind,]+1, alpha.f=0.1), xlim=c(-0.5,0.5), ylim=c(-0.5,0.5), pch=20, xlab="Main environment", ylab="False envi 1")

points(TrueEnv.rho[Sel_ind,1], TrueEnv.rho[Sel_ind,2], col="blue", pch=25, bg="lightblue")

### Main vs. Envi_Minus

plot(TrueEnv.rho[Neut_ind,1],TrueEnv.rho[Neut_ind,3], col=adjustcolor(Is.Sig[Neut_ind,]+1, alpha.f=0.1), xlim=c(-0.5,0.5), ylim=c(-0.5,0.5), pch=20, xlab="Main environment", ylab="False envi 2")

points(TrueEnv.rho[Sel_ind,1], TrueEnv.rho[Sel_ind,3], col="blue", pch=25, bg="lightblue")

### Envi_Plus vs. Envi_Minus

plot(TrueEnv.rho[Neut_ind,2],TrueEnv.rho[Neut_ind,3], col=adjustcolor(Is.Sig[Neut_ind,]+1, alpha.f=0.1), xlim=c(-0.5,0.5), ylim=c(-0.5,0.5), pch=20, xlab="False envi 1", ylab="False envi 2")

points(TrueEnv.rho[Sel_ind,2], TrueEnv.rho[Sel_ind,3], col="blue", pch=25, bg="lightblue")

plot(1,1, col="white")

legend(x = 0.8, y=1.2, legend = c("Neutral", "Neutral false positive", "Selected locus"), pch=c(20,20,25), col=c("black", "red", "blue"), bg=c(NA, NA, "lightblue"))

```
