---
title: "Analysis_PEGA_2"
author: "Katie Lotterhos"
date: "1/10/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load data
setwd("/Users/katie/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysisV2/")
```{r, label="load data"}
load("../data/large_files/Pine_Alpha_AveRho_WithSuperLogical_andAncestralFlip.RData")
```

This loads:

* PE
* results_pine3 (data frame with SNPs in rows and results in columns)
* gtcontig_array 
* gtcontig_array_names
* PE_matchGWAS
* PE_matchGEA

```{r}
if(!("ash" %in% installed.packages())){install.packages("ash")}
if(!("corrr" %in% installed.packages())){install.packages("corrr")}
if(!("igraph" %in% installed.packages())){install.packages("igraph")}
if(!("ggplot2" %in% installed.packages())){install.packages("ggplot2")}
if(!("ggraph" %in% installed.packages())){install.packages("ggraph")}
if(!("dendextend" %in% installed.packages())){install.packages("dendextend")}
if(!("superheat" %in% installed.packages())){install.packages("superheat")}
if(!("VennDiagram" %in% installed.packages())){install.packages("VennDiagram")}
if(!("limma" %in% installed.packages())){install.packages("limma")}
if(!("fields" %in% installed.packages())){install.packages("fields")}
if(!("dplyr" %in% installed.packages())){install.packages("dplyr")}
if(!("limma" %in% installed.packages())){install.packages("limma")}
if(!("cluster" %in% installed.packages())){install.packages("cluster")}
if(!("extrafont" %in% installed.packages())){install.packages("extrafont")}
if(!("RColorBrewer" %in% installed.packages())){install.packages("RColorBrewer")}
#install.packages("VennDiagram")
#install.packages('dendextend')
#install.packages("devtools")
#devtools::install_github("rlbarter/superheat")

library(RColorBrewer)

#source("http://www.bioconductor.org/biocLite.R")
#biocLite("limma")
  library(corrr)
  library(igraph)
  library(ggplot2)
  library(ggraph)
library(dendextend)
library(superheat)
require(VennDiagram)
require(limma)
  require(ash)
  require(fields)
  require(dplyr)
  require(cluster)

  library(extrafont) 
    #font_import() 
  loadfonts()
      
source("plot2Dcov.R")
source("galaxy2.R")
```

Make galaxy plots for Kay's paper
One thought might be to show the results for winter cold injury (and height season 2?) since these are the ones with the most contrasting effects of population structure. 
```{r}
par(mfcol=c(3,2), oma=c(0,0,2,0))
## Winter cold injury
plot2Dcov(x=results_pine3$Winter_cold_injury_p_raw_rho, 
          y= results_pine3$Winter_cold_injury_snp_effect, 
          nbin=100, PElab = c("x", "y"), 
          PE = cor(results_pine3$Winter_cold_injury_p_raw_rho, 
                   results_pine3$Winter_cold_injury_snp_effect, use="pairwise"),
          xlim=c(-0.6,0.6), ylim=c(-0.8,0.8),
          xlab="Uncorrected correlation",
          ylab="GCTA effect size"
          )
plot2Dcov(x=results_pine3$Winter_cold_injury_p_raw_rho,
          results_pine3$Winter_cold_injury_rhoave_BAYENV, 
          nbin=100, PElab = c("x", "y"), 
          PE = cor(results_pine3$Winter_cold_injury_p_raw_rho, 
                   results_pine3$Winter_cold_injury_rhoave_BAYENV, use="pairwise"),
          xlim=c(-0.6,0.6), ylim=c(-0.3,0.3),
          xlab="Uncorrected correlation",
          ylab="Bayenv2-corrected correlation"
          )
plot2Dcov(x=results_pine3$Winter_cold_injury_snp_effect,
          results_pine3$Winter_cold_injury_rhoave_BAYENV, 
          nbin=100, PElab = c("x", "y"), 
          PE = cor(results_pine3$Winter_cold_injury_snp_effect, 
                   results_pine3$Winter_cold_injury_rhoave_BAYENV, use="pairwise"),
          xlim=c(-1,1), ylim=c(-0.3,0.3),
          xlab="GCTA effect size",
          ylab="Bayenv2-corrected correlation"
          )
          
## Winter cold injury
plot2Dcov(x=results_pine3$Height_season_2_p_raw_rho, 
          y= results_pine3$Height_season_2_snp_effect, 
          nbin=100, PElab = c("x", "y"), 
          PE = cor(results_pine3$Height_season_2_p_raw_rho, 
                   results_pine3$Height_season_2_snp_effect, use="pairwise"),
          xlim=c(-0.6,0.6), ylim=c(-0.8,0.8),
          xlab="Uncorrected correlation",
          ylab="GCTA effect size"
          )
plot2Dcov(x=results_pine3$Height_season_2_p_raw_rho,
          results_pine3$Height_season_2_rhoave_BAYENV, 
          nbin=100, PElab = c("x", "y"), 
          PE = cor(results_pine3$Height_season_2_p_raw_rho, 
                   results_pine3$Height_season_2_rhoave_BAYENV, use="pairwise"),
          xlim=c(-0.6,0.6), ylim=c(-0.3,0.3),
          xlab="Uncorrected correlation",
          ylab="Bayenv2-corrected correlation"
          )
plot2Dcov(x=results_pine3$Height_season_2_snp_effect,
          results_pine3$Height_season_2_rhoave_BAYENV, 
          nbin=100, PElab = c("x", "y"), 
          PE = cor(results_pine3$Height_season_2_snp_effect, 
                   results_pine3$Height_season_2_rhoave_BAYENV, use="pairwise"),
          xlim=c(-1,1), ylim=c(-0.3,0.3),
          xlab="GCTA effect size",
          ylab="Bayenv2-corrected correlation"
          )

mtext("Winter cold injury", adj=0.2, side=3, outer=TRUE)
mtext("Height season 2", adj=0.85, side=3, outer=TRUE)

```

Get incides for different groups of columns in the dataframe

```{r}
  names(results_pine3)[grep("raw_rho", names(results_pine3))]
  names(results_pine3)[grep("raw_p", names(results_pine3))]
  
### Spearman's rho (not structure-corrected or "uncorrected")
  env_cols <- grep("raw_rho", names(results_pine3))[1:22]
  rawrho_names <- names(results_pine3)[env_cols] # raw rho columns
  
### P-value for Uncorrected Spearman's rho  
  cols_raw <- grep("raw_p", names(results_pine3))[1:22]
  rawp_names <- names(results_pine3)[cols_raw ] # raw p columns

### Check to make sure they line up    
  cbind(names(results_pine3)[env_cols],names(results_pine3)[cols_raw])
    
### Bayes Factor from Bayenv2    
  bayenvBF_names <- sub("_raw_p","",names(results_pine3)[cols_raw])
  cols_bayenvBF <- which(names(results_pine3) %in% bayenvBF_names)

### Structure-correcte rho from Bayenv2      
  bayenvrho_names <- paste(bayenvBF_names, "rhoave", sep="_")
  cols_bayenvrho <-which(names(results_pine3) %in% bayenvrho_names)

````


### Step 1: Identify top candidate contigs from Yeaman et al. 2016 and identify top candidate SNPs within those contigs

```{r,label="Top SNP's from Sam's top candidates" }

### Subset entire data to top contigs  
  superdf <- results_pine3[results_pine3$pine_super_p9 | 
                           results_pine3$pine_super_raw_p9,]
### Set up logical dataframe for identifying outliers     
  is.superdf <- matrix(NA, nrow(superdf), length(cols_raw))
  colnames(is.superdf) <- names(results_pine3)[cols_raw]


### Identify top candidate SNPs within those contigs
  ### Top SNPs have a significant Bonferroni-corrected P value for uncorrected Spearmans rho and a Bayes Factor from Bayenv2 > 2
  ### This is the code that decides top outliers:
    
    P_cutoff <- -log10(0.05/(22*nrow(results_pine3)))
    P_cutoff
    
    for (i in 1:length(cols_raw)){
      cd <- cols_raw[i]
      cd_bayenv <- cols_bayenvBF[i]
      P_focal <- -log10(abs(superdf[,cd]))
      BF_focal <- superdf[,cd_bayenv]
      is.superdf[,i] <- P_focal>P_cutoff & BF_focal > 2
      is.superdf[is.na(is.superdf[,i]),i] <- FALSE
      #sum(is.na(is.superdf[,i]))
      print( c(i,names(results_pine3)[cd], names(results_pine3)[cd_bayenv]))
    }
  num.var.raw.out <- rowSums(is.superdf,na.rm = TRUE)
  
  ### total number of SNPs in super contigs
  nrow(superdf) 
  
  ### number of all Sam's super contigs
  length(unique(superdf$gtcontig)) 
  
  ### number of super SNPs in super contigs
  sum(num.var.raw.out>0) 
    ## 1241 SNPs without BF cutoff, 801 SNPS with BF cutoff

  ### number of super contigs with superSNPs (# contigs in Group 1)
  length(unique(superdf$gtcontig[num.var.raw.out>0])) 
    ## 133 contigs without BF cutoff, 108 contigs with BF cutoff
  
  
  ### Dataframe of Group 1
  Group1_allSuperSNPs <- superdf[num.var.raw.out>0,]
  dim(Group1_allSuperSNPs)
  ### Logical Group 1
  results_pine3$Group1_logical <- results_pine3$gcontig__gcontig_pos %in% Group1_allSuperSNPs$gcontig__gcontig_pos
  
  ### Sanity check - number of SNPs in Group 1 the same
  (Group1_nSNPs <- sum(results_pine3$Group1_logical))
  
  ### Sanity check - number of contigs in Group 1 the same
  (Group1_ncontigs <-length(levels(factor(results_pine3$gtcontig[results_pine3$Group1_logical]))))
  
  G1_hist <- sort(table(results_pine3$gtcontig[results_pine3$Group1_logical]),decreasing = TRUE)
  
  barplot(G1_hist, xlab="Contig", ylab="Number of SNPs", main=paste("Distribution of",Group1_nSNPs, "top SNPs", "in",Group1_ncontigs, "contigs"), las=2, cex.names=0.3)
  
  top_contigs <- names(G1_hist)
```


### Compare my list to Yeaman 2016 convergence candidates
These are genes we think are true positives because they adapt to climate in both pine and spruce.

Yeaman et al. 2016 identified 47 orthologs with FDR correction

```{r}
sy <- read.table("../data/orthologs/parallelism_genes_pine_spruce_q05.txt")
head(sy)
nrow(sy)
sy_top <- unique(sy$V1)
length(sy_top)
length(unique(sy$V2))

### Make sure all contigs in list are in my df
  for (i in 1:length(sy_top)){
    cd <- as.character(sy_top[i])
    if(length(which(sy_top[i]==results_pine3$gtcontig))==0){
      print(c(i, "missing",cd))
    }
  }
  ### If nothing printed that is a good thing!

### Which ones of Group 1 are super convergent outliers
  sum(top_contigs %in% sy$V1)
  sum(sy$V1 %in% top_contigs)
  
  top_contigs[which(top_contigs %in% sy$V1)]
  ### We get the same 10 contigs with or without the BF cutoff
```
According to the above results, 10 of those 47 are in my list of top candidates.

We can also use a less restrictive list, e.g. orthologs that are significant at P < 0.05 without FDR correction.

TO DO. Can do spruce but need to use orthology table.

```{r, eval=FALSE, echo=FALSE, results=FALSE}
### NOTE THAT THIS IS SPRUCE ORTHOLOGS AND NEEDS EDITING
sy2 <- read.table("../data/orthologs/pine_orthologs_to_spruce_top_candidates_p05.txt", header=TRUE)
head(sy2)
nrow(sy2)

orthoP05 <- unique(sy2$tcontig[sy2$pvalue < 0.05])
length(orthoP05)
  ### Which ones of Group 1 are super convergent outliers
  sum(top_contigs %in% orthoP05)
  
  top_contigs[which(top_contigs %in% orthoP05)]
```


### Group 1 Make some comparison plots

SNPs were coded for association analyses according to alphabetical order. However, we also have the ancestral SNP according to mapping our data onto the loblolly pine genome. The direction of the sign to change the SNP association to one relative to the ancestral state is given by multiplying the SNP association by the column "Ances_flipNum".

Here, I am just comparing the patterns between the alphabetical coding and the ancestral coding.

```{r}
sum(results_pine3$Group1_logical)
sum(results_pine3$Group1_logical & !is.na(results_pine3$Ances_flipNum), na.rm=TRUE)
  # number of top candidate SNPs that we have ancestral state for

names(results_pine3)[grep("MAT", names(results_pine3))]
```


## Co-association network analysis
The objective of the next section is to identify groups of SNPs that show similar patterns of association across the multivariate environment.

This analysis consists of several steps:

1) Cluster the environmental data according to their abs(correlation) with each other.

2) Cluster the SNPs according to their abs(associations) with environments, while holding the order of the environments according to the environmental clustering. Through data visualizations and toy data I've determined that the absolute value here is important, because the sign of the SNP association is arbitrary anyway.

3) Identify groups of SNPs according to how they cluster together.

4) Use undirected network graphs to visualize how contigs and SNPs cluster together.

5) Use galaxy plots to interpret individual SNP correlations in multiple environments relative to the ancestral state.

Load the Pine environmental data and determine cluster relationships among environments
```{r}
### 
  load("../data/PineHeatmap.Rdata")
  # loads env_clust - this cluster object was not based on signed associations and led to some weird clustering
    # env_clust$rowInd is the order of environments based on signed clustering
  env_clust_old <- env_clust
  env_clust_old$rowInd
  # raw correlations
    head(env_clust_old$carpet)
  # Redo clustering based on abs()
    cormat <- t(env_clust_old$carpet)
    cormat <- cormat[,22:1]
    head(cormat)
    longnames <- data.frame(rownames(cormat),colnames(cormat))
    #rownames(cormat) <- colnames(cormat)
    head(cormat)
    env_clust <- hclust(abs(dist(as.matrix(cormat))))
    env_clust$order
    env_clust$labels
    env_clust$labels[env_clust$order]
    rownames(env_clust_old$carpet)
    rownames(t(env_clust_old$carpet))
  # heatmap of environments
    env_clust$labels
    colnames(cormat)
    cormat2 <- cormat[env_clust$order, ]
    #colnames(cormat2)
    cormat2 <- cormat2[, env_clust$order]
    head(cormat2)
    par(mar=c(4,1,1,1))
    png("../results/dendogram_environment.png", width=12, height=9, units = "in", res=500)
    superheat(cormat2, col.dendrogram = TRUE,#row.dendrogram=TRUE,
              order.rows = 1:22,
              order.cols = 1:22,#legend.height = 0.05,
              left.label.size = 1, bottom.label.size =  0.25,
              bottom.label.text.angle = 90)
    dev.off()
```


First, take absolute value of the SNP effect sizes b/c of signed nature of alleles
```{r}
results_pine3$gtcontig_pos <-  paste(results_pine3$gtcontig, results_pine3$pos_gcontig, sep="_")
labs <- results_pine3$gtcontig_pos
labs2 <- gsub("_c0_seq1",replacement = "",labs)
labs3 <- gsub("_c1_seq1",replacement = "",labs2)
labs4 <- gsub("_c4_seq1",replacement = "",labs3)
results_pine3$gtcontig_pos <- labs4

# Distance matrix and hierarchical cluster based on association patterns
  snp_mat <- abs(results_pine3[results_pine3$Group1_logical, env_cols])
  rownames(snp_mat) <- results_pine3$gtcontig_pos[results_pine3$Group1_logical]
  head(snp_mat)
  snp_dist <- dist(snp_mat, diag=TRUE, upper=TRUE)
  str(snp_dist)
  snp_order <- hclust(snp_dist)
  head(snp_order$labels[snp_order$order])

# Assign cluster membership
  clusMember = cutree(snp_order, 4)
  table(clusMember)
  length(clusMember)
  results_pine3$cluster[results_pine3$Group1_logical] <- clusMember
    # 1 is Aridity, 2 is geography, 3 is freezing and 4 is multi


# Match order environments in SNPs to order of environments by their associations with each other
  sample1 = abs(results_pine3[results_pine3$Group1_logical, env_cols])
  colnames(sample1) <- sub("_raw_rho", "", colnames(sample1))
  colnames(sample1)
  #need to match this to the environmental clustering
  dim(sample1)
  
  m <- match(colnames(cormat2), colnames(sample1))
  sample <- sample1[,m]
  colnames(sample)

png("../results/dendogram_ordered.png", width=12, height=9, units = "in", res=500)
superheat(t(sample), 
          order.rows= 1:22,#rev(env_clust_old$rowInd),
          order.cols = snp_order$order,
          heat.pal = (brewer.pal(5, "Blues")), heat.na.col = "white",
          col.dendrogram = T, row.dendrogram = TRUE,
            left.label.size = 0.2)
dev.off()


### Compare pattern to structure-corrected rho
  sample_bayenv = (results_pine3[results_pine3$Group1_logical, cols_bayenvrho])
  colnames(sample_bayenv) <- sub("_rhoave", "", colnames(sample_bayenv))
  dim(sample_bayenv)
  m <- match(colnames(cormat2), colnames(sample_bayenv))
  sample_bv2 <- sample_bayenv[,m]
  colnames(sample_bv2)
  
  png("../results/dendogram_ordered_bayenv.png", width=12, height=9, units = "in", res=500)
    superheat(t(abs(sample_bv2)), 
            order.rows=1:22, #rev(env_clust_old$rowInd),
            heat.pal = (brewer.pal(5, "Blues")),
            heat.na.col = "white",
            col.dendrogram = TRUE, row.dendrogram = TRUE,
              left.label.size = 0.2,
          order.cols = snp_order$order)
  dev.off()
```
  
## Create Venn Table
```{r}
limit.df <- results_pine3[results_pine3$Group1_logical,]
dim(limit.df)

### Overlap among SNPs
a<- vennCounts(cbind(limit.df$cluster==1, # aridity
                     limit.df$cluster==2, # geography
                     limit.df$cluster==3, # freezing
                     limit.df$cluster==4)) # multi
# vennDiagram(a) note that this is meaningless b/c at SNP level

### Overlap among contigs
contig.group.df <- data.frame(
  Multi = table(limit.df$gtcontig,limit.df$cluster==4)[,2],
  Aridity = table(limit.df$gtcontig,limit.df$cluster==1)[,2],
  Freezing = table(limit.df$gtcontig,limit.df$cluster==3)[,2],
  Geography = table(limit.df$gtcontig,limit.df$cluster==2)[,2]
)

# "group" is for order
contig.group.df$group <- apply(contig.group.df[,1:4],1, FUN = function(x)(which(x==max(x))))
contig.group.df$nout <- rowSums(contig.group.df[,1:4])
# cluster is to correspond to SNP df
contig.group.df$cluster <- NA
contig.group.df$cluster[contig.group.df$group==1] <- "Multi-4"
contig.group.df$cluster[contig.group.df$group==2] <- "Aridity-1"
contig.group.df$cluster[contig.group.df$group==3] <- "Freezing-3"
contig.group.df$cluster[contig.group.df$group==4] <- "Geography-2"

###
contig.group.df<- contig.group.df[order(contig.group.df$group,contig.group.df$nout, decreasing=c(FALSE, TRUE),method="radix"),]
contig.group.df

contig.group.df$ContigID <- 1:nrow(contig.group.df)
  
contig.group.df$col <- NA
contig.group.df$gtcontig <- rownames(contig.group.df)

# Geog
  whichones <- contig.group.df$group==4
  ncol <- sum(whichones)
  ncol
   fills <- adjustcolor(two.colors(n=ncol, start="gold", middle= "peachpuff", 
                                   end="lightgoldenrod3"),alpha=0.9)
  plot(1:length(fills), col=fills, pch=19, cex=2)
  contig.group.df$col[whichones] = fills
  text(1:length(fills), labels=paste(contig.group.df$gtcontig[whichones], 
                                     contig.group.df$ContigID[whichones], sep="  "), cex=0.5)
  
# Aridity
  whichones <- contig.group.df$group==2
  ncol <- sum(whichones)
  ncol
  fills <- adjustcolor(two.colors(n=ncol, start="orange", middle= "brown2", end="darkorange4"),alpha=0.9)
  plot(1:length(fills), col=fills, pch=19, cex=2)
  contig.group.df$col[whichones] = fills
  text(1:length(fills), labels=paste(contig.group.df$gtcontig[whichones], 
                                     contig.group.df$ContigID[whichones], sep="  "), cex=0.5)
  
# Multi
  whichones <- contig.group.df$group==1
  ncol <- sum(whichones)
  ncol
  fills <- adjustcolor(c("#DAF7A6", "chartreuse4", "#28B463"), alpha=0.9)
  plot(1:length(fills), col=fills, pch=19, cex=2)
  contig.group.df$col[whichones] = fills
  text(1:length(fills), labels=paste(contig.group.df$gtcontig[whichones], 
                                     contig.group.df$ContigID[whichones], sep="  "), cex=0.5)
  

# Freezing
  whichones <- contig.group.df$group==3
  ncol <- sum(whichones)
  ncol
  fills <- adjustcolor(two.colors(n=ncol, start="darkturquoise", middle= "dodgerblue", end=,"darkslategrey"),alpha=0.9)
  plot(1:length(fills), col=fills, pch=19, cex=2)
  contig.group.df$col[whichones] = fills
  text(1:length(fills), labels=paste(contig.group.df$gtcontig[whichones], 
                                     contig.group.df$ContigID[whichones], sep="  "), cex=0.5)
  
  
contig.group.df
```

Add GO terms and super convergent status (Yeaman 2016) to contig.group.df
```{r}
### Super convergent status
  conv.ind <- which(contig.group.df$gtcontig %in% sy$V1)
  contig.group.df$is.convergent <- FALSE
  contig.group.df$is.convergent[conv.ind ] <- TRUE
  contig.group.df[conv.ind ,]

### GO terms
  go.df <- read.table("../GOanalysis/tair_pine_clusterall", header=TRUE)
  head(go.df)
  dim(go.df)
  go.df$pine_comp <- paste(go.df$pine_comp, "_seq1", sep="")
  m1<- match(go.df$pine_comp, contig.group.df$gtcontig)
  names(go.df)[1] <- "gtcontig"
  contig.group.df2<-merge(contig.group.df,go.df, all.x=TRUE)
  contig.group.df2 <- contig.group.df2[order(contig.group.df2$ContigID),]
write.csv(contig.group.df2, "../results/topCandidateContigs.csv")
```

### Examine LD based on $r^2$ among contigs  
```{r}
LD.df <- read.table("../data/large_files/pine_super_outlier_LD_p4_all_genemeans.txt", header=TRUE)
head(LD.df)

ind <- which(LD.df$tcontig1 %in% contig.group.df$gtcontig & LD.df$tcontig2 %in% contig.group.df$gtcontig)
length(ind)

head(LD.df[ind,])

LD2 <- merge(LD.df[ind,], contig.group.df[,c("gtcontig", "ContigID", "cluster")], by.x="tcontig2", by.y="gtcontig")

names(LD2)[c(4,5)] <- paste(names(LD2)[c(4,5)], ".2", sep="")

LD3 <- merge(LD2,  contig.group.df[,c("gtcontig", "ContigID", "cluster")], by.x="tcontig1", by.y="gtcontig")

LD3 <- LD3[order(LD3$ContigID, LD3$ContigID.2),]

head(LD3)
max(LD3$ContigID)*max(LD3$ContigID)
dim(LD3)
# perfect match up!

LD3mat <- matrix(LD3$r2, ncol=108)

png("../results/LD.png", width=6, height=5, units = "in", res=400)
  par(mar=c(1,1,1,6), oma=c(2,2,0,0))
  image(1:108, 1:108, LD3mat, xlab="Contig ID", ylab="Contig ID", col=tim.colors(50), xlim=c(-2,108), ylim=c(-2,108), xaxt="n", yaxt="n", bty="n")
  points(1:108, rep(-1, 108), pch=19, col=contig.group.df$col, cex=0.5)
  points(rep(-1, 108),1:108,  pch=19, col=contig.group.df$col, cex=0.5)
  image.plot(LD3mat, legend.only=TRUE, legend.lab = "Correlation in allele frequencies", legend.line = 2.5)
dev.off()
```

Make Pine Venn Barplot with contig colors
```{r}
toplot <- t(as.matrix(contig.group.df[,1:4]))
  head(toplot)
  dim(toplot)
  #dim(contigdat)
  colnames(toplot) <- contig.group.df$ContigID

pdf("../results/PineVenn_horiz.pdf", width=20, height=6)
  par(mar=c(6,6, 1,1))
  b<- barplot(toplot, beside = FALSE, horiz = FALSE, las=2, col=c("lightgreen", "orange", "lightblue",  "yellow"), cex.names=1, cex.axis = 2,  border = c("darkgreen", "brown", "blue",  "brown"), cex=2,  ylab="Number of SNPs", xlab= "Contig ID", ylim=c(-8,50), names.arg = rep("", 108), cex.lab=2)

  legend(100, 30, c("Multi", "Aridity", "Freezing",  "Geography"), bty="n", fill=c("lightgreen", "orange", "lightblue",  "yellow"), border = c("darkgreen", "brown", "blue",  "brown"), cex=2)

  points(b, rep(-1,108), pch=20, col=contig.group.df$col, cex=2)
  text(b, rep(-2, 108), colnames(toplot), col=contig.group.df$col, srt=90, adj=1) 
  convt <- rep("", 108)
  convt[contig.group.df$is.convergent] <- "*"
  text(b, rep(-5.5,108), convt)
dev.off()
```

Annotation barplot

```{r}
ann2 <- as.character(results_pine3$X_annotation)
ann2[grep("nonsyn*",ann2)] <- "nonsyn"

all_an <- table(ann2)
out_an <- table(ann2[results_pine3$Group1_logical])

compare_ann <- merge(data.frame( all_an/sum(all_an)), data.frame(out_an/sum(out_an)) , by.x="ann2", by.y="Var1", all.x=TRUE)

head(compare_ann)
compare_ann <- compare_ann[-which(compare_ann$ann2 %in% c("HAS_INDEL", "mismatch_altref", "multi-allelic")),]

compare_ann

pdf("../results/annotation.pdf", width=8, height=5)
  par(mar=c(8,4,1,1))
  barplot(t(as.matrix(compare_ann[,2:3])), beside=TRUE, ylab="Proportion", names.arg = compare_ann$ann2, las=2)
  legend(20, 0.5, c("All SNPs", "Top Candidates"), fill=c("grey30", "grey80"))
dev.off()
```

## Co-association network

#### Function to draw network based on distance matrix
```{r}
shortclust <- results_pine3$cluster[results_pine3$Group1_logical]
source("myFunDist.R")
### In contig.group.df, "group" is for clusterID and plotting order, but not equal to cluster given by the SNP

MakeDistNetwork <- function(clustnum, filename, disthreshold, full_name=FALSE){
  
  ### Based on correlation in associations (co-association)
  #names(results_pine3)[env_cols]
  #multidat <- abs(results_pine3[which(results_pine3$cluster==clustnum), env_cols])
  #str(multidat)
  #rownames(multidat) <- results_pine3$gtcontig_pos[which(results_pine3$cluster==clustnum)]
  #dim(multidat)
  
  snp_distmat_multi <- snp_distmat[shortclust ==clustnum, shortclust ==clustnum]
  print(dim(snp_distmat_multi))
  head(snp_distmat_multi[1:5, 1:5])
  
  #hist(snp_distmat_multi)
  # This histogram is bimodal with the larger values > 0.4

  d4_multi <- myFun(as.dist(snp_distmat_multi))
  head(d4_multi)
  #hist(d4_multi$value)
  
  (nedges <- nrow(d4_multi %>%
    filter(abs(d4_multi$value) < disthreshold)))
  
  print(c("Number of edges:", nedges))
  if(nedges > 30000){print("Too many edges"); break}
  
  
  graph_cors <- d4_multi %>%
    filter(value < disthreshold) %>% #filter to close distances
    graph_from_data_frame(directed = FALSE)
  
    # check node names and add nodes if they are missing
    V(graph_cors)$name
    toadd <- !(rownames(snp_distmat_multi) %in% V(graph_cors)$name)
    sum(toadd)
    graph_cors2 <- add_vertices(graph_cors, nv = sum(toadd, na.rm=TRUE), name= rownames(snp_distmat_multi)[toadd])
    V(graph_cors2)$name
    sum(!(rownames(snp_distmat_multi) %in% V(graph_cors2)$name))
      #should be 0 because nothing missing
    graph_cors <- graph_cors2
    
    # line up names
    (myo <- match(V(graph_cors)$name, results_pine3$gtcontig_pos))
    
    contigname = results_pine3$gtcontig[myo]
    
    myocol <- contig.group.df$col[match(contigname, contig.group.df$gtcontig)]
    
    if(full_name==FALSE){
    myoname <- contig.group.df$ContigID[match(contigname, contig.group.df$gtcontig)]
    }
    if(full_name==TRUE){
      myoname <- V(graph_cors)$name
    }
    

   #cluster membership
   mem=clusters(graph_cors)$membership
   
   # number edges for each node
   nodenumedge <- sapply(V(graph_cors)$name, function(x) length(E(graph_cors)[from(V(graph_cors)[x])]))
   
  
   png(file = paste("../results/NetworkDist2Hist", filename,".pdf", sep=""), width=15, height=15, units="in", res=300)
   hist(d4_multi$value)
   lines(x=c(disthreshold, disthreshold), y=c(0,1000), col="blue", lwd=5)
   dev.off()

    ggraph(graph_cors) +
    geom_edge_fan2(aes(edge_alpha = 0.01)) +
    guides(edge_alpha = "none", edge_width = "none") +
    #scale_edge_colour_gradientn(limits = c(-1, 1), 
    #                            colors = c("firebrick2", "dodgerblue2")) +
    geom_node_point(color = myocol, size = 15) +
    geom_node_text(aes(label = myoname, size=0.1)) +
    theme_graph() #+
    #labs(title = filename)
  
    options(warn=-1)

    ggsave(paste("../results/Network", filename,disthreshold, full_name,"dist.pdf", sep="_"), width=15, height=15, units="in")
  
     if(identical(names(nodenumedge), names(mem))){
    return(data.frame(gtcontig_pos=names(mem), subcluster=mem, numedge=nodenumedge))
   }else{
     print("error: node names and edge names not identical")
   }

} # end function
```

```{r}
colSums(table(results_pine3$gtcontig[results_pine3$Group1_logical], results_pine3$cluster[results_pine3$Group1_logical]))
snp_distmat <- as.matrix(snp_dist)
results_pine3$subcluster <- NA
results_pine3$numedges <- NA
results_pine3$allele <- NA

#### Multi Dist Network ####
mm <- MakeDistNetwork(4, "Multi", 0.1) 
  head(mm)
  hist(mm$subcluster)
  ### Assign subcluster membership and number edges
  mind <- match(mm$gtcontig_pos, results_pine3$gtcontig_pos)
  results_pine3$subcluster[mind] <- mm$subcluster
  results_pine3$numedges[mind] <- mm$numedge
  ### Assign allele ID based on sign of association determined from data vis
  ### For Multi group is it MAT
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]>0]<- "+"
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]<0]<- "-"
  which(names(results_pine3)=="MAT_raw_rho")      
  results_pine3[mind, c(197,401:410)]
  # Number of alleles "evolving" in each direction relative to ancestor
  table(results_pine3$allele[mind])  
  # direction of SNPs in each contig
  table(results_pine3$gtcontig[mind], results_pine3$allele[mind])

#### Aridity Dist Network ####
rm(mm)
mm <- MakeDistNetwork(1, "Aridity", 0.1) 
  head(mm)
  hist(mm$subcluster)
  ### Assign subcluster membership and number edges
  mind <- match(mm$gtcontig_pos, results_pine3$gtcontig_pos)
  results_pine3$subcluster[mind] <- mm$subcluster
  results_pine3$numedges[mind] <- mm$numedge
  ### Assign allele ID based on sign of association determined from data vis
  ### For Aridity group is it MAT
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]>0]<- "+"
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]<0]<- "-"
  results_pine3[mind, c(197,401:410)]
  table(results_pine3$allele[mind]) 
  # direction of SNPs in each contig
  table(results_pine3$gtcontig[mind], results_pine3$allele[mind])
  
#### Geography Dist Network ####
rm(mm)
mm <- MakeDistNetwork(2, "Geography", 0.1) 
  head(mm)
  hist(mm$subcluster)
  ### Assign subcluster membership and number edges
  mind <- match(mm$gtcontig_pos, results_pine3$gtcontig_pos)
  results_pine3$subcluster[mind] <- mm$subcluster
  results_pine3$numedges[mind] <- mm$numedge
  ### Assign allele ID based on sign of association determined from data vis
  ### For Geography group is it MAT ????
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]>0]<- "+"
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]<0]<- "-"
  #results_pine3[mind, c(197,401:410)]
  table(results_pine3$allele[mind]) 
  # direction of SNPs in each contig
  table(results_pine3$gtcontig[mind], results_pine3$allele[mind])
  
#### Freezing Dist Network ####
rm(mm)
mm <- MakeDistNetwork(3, "Freezing", 0.1)
  head(mm)
  ### Assign subcluster membership and number edges
  hist(mm$subcluster)
  mind <- match(mm$gtcontig_pos, results_pine3$gtcontig_pos)
  results_pine3$subcluster[mind] <- mm$subcluster
  results_pine3$numedges[mind] <- mm$numedge
  ### Assign allele ID based on sign of association determined from data vis
  ### For Freezing group is it LAT & Elevation
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]>0]<- "+"
    results_pine3$allele[mind][results_pine3$MAT_raw_rho[mind]*results_pine3$Ances_flipNum[mind]<0]<- "-"
  #results_pine3[mind, c(197,401:412)]
  # total num SNPs increasing in freq in each direction of MAT
  table(results_pine3$allele[mind]) 
  # direction of SNPs in each contig
  table(results_pine3$gtcontig[mind], results_pine3$allele[mind])
  
### Number of contigs in each cluster/subcluster combo
  tapply(results_pine3$gtcontig, list(results_pine3$cluster, results_pine3$subcluster), function(x){length(levels(as.factor(x)))})
```


## Add info

```{r}

#match SNP colors to contig dat
# Cluster color for outline
  #Aridity = table(limit.df$gtcontig,limit.df$cluster==1)[,2],
  #Geography = table(limit.df$gtcontig,limit.df$cluster==2)[,2],
  #Freezing = table(limit.df$gtcontig,limit.df$cluster==3)[,2],
  #Multi = table(limit.df$gtcontig,limit.df$cluster==4)[,2]
  #adjustcolor("yellow",0.5)("brown",0.5)
  #adjustcolor("lightgreen",0.5) ("darkgreen",0.5)
  #adjustcolor("orange",0.5) ("brown",0.5)
  #adjustcolor("lightblue",0.5) adjustcolor("blue",0.5)
  results_pine3$clustercol[results_pine3$cluster==1] = "orange"
  results_pine3$clustercol[results_pine3$cluster==2] = "yellow"
  results_pine3$clustercol[results_pine3$cluster==3] = "blue"
  results_pine3$clustercol[results_pine3$cluster==4] = "green"
  table(as.character(results_pine3$clustercol[results_pine3$Group1_logical]))
  
  # num snps
  (t1 <- table(as.character(results_pine3$cluster[results_pine3$Group1_logical])))
  sum(t1)
  # num contigs
  (t2<- sapply(1:4, function(x) length(table(results_pine3$gtcontig[results_pine3$cluster==x]))))
  sum(t2)
  
# Contig color for points
  m <- match(results_pine3$gtcontig[results_pine3$Group1_logical], contig.group.df$gtcontig)
  m
  results_pine3$contigcol[results_pine3$Group1_logical] <- contig.group.df$col[m]

  results_pine3$contigID <- NA
  results_pine3$contigID[results_pine3$Group1_logical] <- contig.group.df$ContigID[m]
  results_pine3$group[results_pine3$Group1_logical] <- contig.group.df$group[m]
  
  write.csv(results_pine3[results_pine3$Group1_logical,], file = "../results/PineTopCandidates.csv")


  table(results_pine3$cluster, results_pine3$pine_super_convergent)
    # super convergent SNPs found in elevation freezing group or geography group
  
```
   
```{r} 
pdf(file = "../results/GalaxyExample_inMATvMAP2.pdf", width=5, height=5)
x<- results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum
y <- results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum
par(mar=c(6,6,1,1))
galaxyplot(x=x, y=y, 
       xlab= "Spearman's rho between allele frequency and\nMean Annual Temperature", 
       ylab= "Spearman's rho between allele frequency and\nMean Annual Precipitation" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),PE=0.33, PElab=c("MAT", "MAP"))
dev.off()

pdf(file = "../results/GalaxyNew_inMATvMAP2.pdf", width=8, height=13)
par(mfrow=c(3,2), mar=c(6,6,1,1), oma = c(4, 0, 1, 0))

galaxyplot(x=x, y=y, 
       xlab= "Spearman's rho between allele frequency and\nMean Annual Temperature", 
       ylab= "Spearman's rho between allele frequency and\nMean Annual Precipitation" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub=x[results_pine3$Group1_logical], 
       y_sub=y[results_pine3$Group1_logical],
       sub_outlinecolor = results_pine3$clustercol[results_pine3$Group1_logical],
       sub_bgcolor = results_pine3$contigcol[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))
  text(-0.7, 0.7, "A", cex=2)
  mtext("Associations", side=3)

x<- results_pine3$MAT_rhoave*results_pine3$Ances_flipNum
y <- results_pine3$MAP_rhoave*results_pine3$Ances_flipNum
galaxyplot(x=x, y=y, 
       xlab= "Spearman's rho between structure-corrected\nallele frequency and Mean Annual Temperature", 
       ylab= "Spearman's rho between structure-corrected\nallele frequency and Mean Annual Precipitation" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
      x_sub=x[results_pine3$Group1_logical], 
       y_sub=y[results_pine3$Group1_logical],
       sub_outlinecolor = results_pine3$clustercol[results_pine3$Group1_logical],
       sub_bgcolor = results_pine3$contigcol[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))
  text(-0.25, 0.25, "B", cex=2)
  mtext("Structure-corrected associations", side=3)

### MAT VS ELVATION
x<- results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum
y <- results_pine3$ELEVATION_raw_rho*results_pine3$Ances_flipNum

galaxyplot(x=x, y=y, 
       xlab= "Spearman's rho between allele frequency and\nMean Annual Temperature", 
       ylab= "Spearman's rho between allele frequency and Elevation" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub=x[results_pine3$Group1_logical], 
       y_sub=y[results_pine3$Group1_logical],
       sub_outlinecolor = results_pine3$clustercol[results_pine3$Group1_logical],
       sub_bgcolor = results_pine3$contigcol[results_pine3$Group1_logical],
       PE=-0.48, PElab=c("MAT", "ELEVATION"))
  text(-0.7, 0.7, "C", cex=2)

x<- results_pine3$MAT_rhoave*results_pine3$Ances_flipNum
y <- results_pine3$ELEVATION_rhoave*results_pine3$Ances_flipNum
galaxyplot(x=x, y=y, 
       xlab= "Spearman's rho between structure-corrected\nallele frequency and Mean Annual Temperature", 
       ylab= "Spearman's rho between structure-corrected\nallele frequency and Elevation" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
      x_sub=x[results_pine3$Group1_logical], 
       y_sub=y[results_pine3$Group1_logical],
       sub_outlinecolor = results_pine3$clustercol[results_pine3$Group1_logical],
       sub_bgcolor = results_pine3$contigcol[results_pine3$Group1_logical],
      PE=-0.48, PElab=c("MAT", "ELEVATION"))
  text(-0.25, 0.25, "D", cex=2)

### MAT VS LAT
x<- results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum
y <- results_pine3$LAT_raw_rho*results_pine3$Ances_flipNum

galaxyplot(x=x, y=y, 
       xlab= "Spearman's rho between allele frequency and\nMean Annual Temperature", 
       ylab= "Spearman's rho between allele frequency and Latitude" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub=x[results_pine3$Group1_logical], 
       y_sub=y[results_pine3$Group1_logical],
       sub_outlinecolor = results_pine3$clustercol[results_pine3$Group1_logical],
       sub_bgcolor = results_pine3$contigcol[results_pine3$Group1_logical],
       PE=-0.48, PElab=c("MAT", "LAT"))
    text(-0.7, 0.7, "E", cex=2)

x<- results_pine3$MAT_rhoave*results_pine3$Ances_flipNum
y <- results_pine3$LAT_rhoave*results_pine3$Ances_flipNum
galaxyplot(x=x, y=y, 
       xlab= "Spearman's rho between structure-corrected\nallele frequency and Mean Annual Temperature", 
       ylab= "Spearman's rho between structure-corrected\nallele frequency and Latitude" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
      x_sub=x[results_pine3$Group1_logical], 
       y_sub=y[results_pine3$Group1_logical],
       sub_outlinecolor = results_pine3$clustercol[results_pine3$Group1_logical],
       sub_bgcolor = results_pine3$contigcol[results_pine3$Group1_logical],
      PE=-0.48, PElab=c("MAT", "LAT"))
  text(-0.25, 0.25, "F", cex=2)
  
  par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE)
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n")
legend("bottom", c("Multi", "Aridity", "Freezing", "Geography"), xpd = TRUE, horiz = TRUE, inset = c(0, 
    0), bty = "n", pch = 21, col= c("green", "orange", "blue", "yellow"), pt.bg = c("darkgreen", "brown2", "lightblue", "goldenrod"), cex = 2)
# xpd = TRUE tells R that it is OK to plot outside the region horiz = TRUE
# tells R that I want a horizontal legend inset = c(x,y) tells R how to move
# the legend relative to the 'bottom' location bty = 'n' means that 'no' box
# will be drawn around it pch and col are the types and colors of points cex
# = 2 makes the legend twice as large as other fonts

dev.off()
```


## Galaxy in multivariate environment raw 4 clusters TO DO
```{r}

whichvars <- c("LAT", "ELEVATION", "MAT","MAP", "CMD") #Eref

pdf("../results/multvariateEnviGalaxy4clusters.pdf", width=22, height=20)
par(mfrow=c(length(whichvars), length(whichvars)), mar=c(0.3,0.7,0,0.7), oma=c(10,10,0,0))
for (i in 1:length(whichvars)){
  for (j in 1:length(whichvars)){
    if(i==j){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      text(1,1, whichvars[i], cex=5)  
    }
    if(j>i){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      if(j==2 & i==1){
        legend(1,1, legend = c("Freezing", "Aridity", "Multi", "Geography"), 
               pch=c(22,24,21), cex=4, bty="n",xjust=0.5,yjust=0.5,
               col=c("blue", "brown", "darkgreen", "brown"), 
               pt.bg=c("lightblue", "orange", "lightgreen", "yellow"))
      }
    }
      
    if(i>j){
      print(c(i,j))
      print(c(whichvars[j], whichvars[i]))
      # Get EE corr
      ri <- grep(paste("^",whichvars[j],sep=""),rownames(cormat2)) #env_clust$carpet))
      cj <- grep(paste("^",whichvars[i],sep=""),colnames(cormat2))#env_clust$carpet))
      print(c(rownames(cormat2)[ri], colnames(cormat2)[cj]))
      PE<- round(cormat2[ri,cj],2)
      
      whichx <- which(names(results_pine3)==paste(whichvars[j], "_raw_rho", sep=""))
      whichy <- which(names(results_pine3)==paste(whichvars[i], "_raw_rho", sep=""))
      print(names(results_pine3)[c(whichx,whichy)])
      x <- results_pine3[,whichx]*results_pine3$Ances_flipNum
      y <- results_pine3[,whichy]*results_pine3$Ances_flipNum
      
      plot2Dcov(x=x,
       y=y, 
       xlab= "", 
       ylab= "" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=x[results_pine3$cluster==1], 
       y_sub_orange=y[results_pine3$cluster==1],
       x_sub_blue=x[results_pine3$cluster==3], 
       y_sub_blue=y[results_pine3$cluster==3],
       x_sub_green=x[results_pine3$cluster==4], 
       y_sub_green=y[results_pine3$cluster==4], 
       x_sub_yellow=x[results_pine3$cluster==2], 
       y_sub_yellow=y[results_pine3$cluster==2], 
       plot_greenCOV = FALSE, plot_orangeCOV = FALSE, 
       plot_blueCOV = FALSE, plot_yellowCOV = FALSE,
       PE=PE, PElab=c(whichvars[i], whichvars[j]))
    } # end if
  } # end j
} # end i
  mtext(text = "Correlation between allele frequency and environment (in column)", side=1, outer=TRUE, line=5, cex=3, adj=0.1)
  mtext(text = "Correlation between allele frequency and environment (in row)", side=2, outer=TRUE, line=5, cex=3, adj=0.1)
dev.off()
```