---
title: "Analysis_PEGA_2"
author: "Katie Lotterhos"
date: "1/10/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load data
setwd("/Users/katie/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysisV2/")
```{r, label="load data"}
load("../data/large_files/Pine_Alpha_AveRho_WithSuperLogical.RData")
```

```{r}
#if(!("ash" %in% installed.packages())){install.packages("ash")}
#install.packages("VennDiagram")
#source("http://www.bioconductor.org/biocLite.R")
#biocLite("limma")

require(VennDiagram)
require(limma)
  require(ash)
  require(fields)
  require(dplyr)
  require(cluster)
source("plot2Dcov.R")

load("../data/Ances_flip.Rdata")
head(Ances_df)
head(results_pine3[,1:20])

lineup2 <- match(paste(results_pine3$gcontig, results_pine3$pos_gcontig, sep="_"), paste(Ances_df$CHROM, Ances_df$POS, sep="_") )
sum(is.na(lineup2))/length(lineup2) # 100% matched
sum(!is.na(lineup2))/nrow(results_pine3)
nrow(Ances_df[lineup2,])/nrow(results_pine3)

head(results_pine3[,1:20])
head(Ances_df[lineup2,])
tail(results_pine3[,1:20])
tail(Ances_df[lineup2,])

results_pine3 <- data.frame(results_pine3, Ances_df[lineup2,])
ncol(results_pine3)
head(results_pine3[,c(1:20, 399:402)])
     
levels(factor(results_pine3$Ances_flip))
results_pine3$Ances_flipNum <- results_pine3$Ances_flip
results_pine3$Ances_flipNum[results_pine3$Ances_flip=="SNP"] <- NA

results_pine3$minor_allele <- as.character(results_pine3$minor_allele)
results_pine3$major_allele <- as.character(results_pine3$major_allele)

sum(results_pine3$Ances_flip %in% c(results_pine3$minor_allele, results_pine3$major_allele))/nrow(results_pine3)
  #80% of ancestral alleles match one of the SNPs in pine

flipnum <- rep(NA, nrow(results_pine3))
for (i in 1:nrow(results_pine3)){
  if(i%%100000==0){print(i)}
  bob <- which(c(results_pine3$minor_allele[i], results_pine3$major_allele[i])==results_pine3$Ances_flip[i])
  boborder <- order(c(results_pine3$minor_allele[i], results_pine3$major_allele[i]))
  # if the ancestral allele is one of alleles in pine SNP
  
  if(length(bob)>0){
    if (boborder[bob]==1){
      # the ancestral SNP is the allele that comes earlier in the alphabet
    flipnum[i] <- 1
    }
     # the ancestral SNP is NOT the allele that comes earlier in the alphabet
    if(boborder[bob]==2){
      flipnum[i] <- -1
    }
  }#end if length(bob) 
}

sum(!is.na(flipnum))/nrow(results_pine3)

results_pine3$Ances_flipNum <- flipnum
```
This loads:

* PE
* results_pine3 (data frame with SNPs in rows and results in columns)
* gtcontig_array 
* gtcontig_array_names
* PE_matchGWAS
* PE_matchGEA


### Group 1: Top SNP's from Sam's top candidates

```{r,label="Top SNP's from Sam's top candidates" }
  superdf <- results_pine3[results_pine3$pine_super_p9 | 
                           results_pine3$pine_super_raw_p9,]
  names(results_pine3)[grep("raw_rho", names(results_pine3))]
  names(results_pine3)[grep("raw_p", names(results_pine3))]
  
    env_cols <- grep("raw_rho", names(results_pine3))[1:22]
    
    cols_raw <- grep("raw_p", names(results_pine3))[1:22]
    cbind(names(results_pine3)[env_cols], names(results_pine3)[cols_raw])
    
    is.superdf <- matrix(NA, nrow(superdf), length(cols_raw))
    colnames(is.superdf) <- names(results_pine3)[cols_raw]
    #P_cutoff <- -log10(0.05/nrow(results_pine3)) # 10-7 P-value Bonferroni
    P_cutoff <- -log10(0.05/(22*nrow(results_pine3)))
    P_cutoff
    for (i in 1:length(cols_raw)){
      cd <- cols_raw[i]
      P_focal <- -log10(abs(superdf[,cd]))
      is.superdf[,i] <- P_focal>P_cutoff
      is.superdf[is.na(is.superdf[,i]),i] <- FALSE
      #sum(is.na(is.superdf[,i]))
      print(i)
    }
  num.var.raw.out <- rowSums(is.superdf,na.rm = TRUE)
  
  #total number of SNPs in super contigs
  nrow(superdf) 
  
  #number of all Sam's super contigs
  length(unique(superdf$gtcontig)) 
  
  # number of super SNPs in super contigs
  sum(num.var.raw.out>0) 

  # number of super contigs with superSNPs (# contigs in Group 1)
  length(unique(superdf$gtcontig[num.var.raw.out>0])) 
  
   ### Dataframe of Group 1
  Group1_allSuperSNPs <- superdf[num.var.raw.out>0,]
  dim(Group1_allSuperSNPs)
  ### Logical Group 1
  results_pine3$Group1_logical <- results_pine3$gcontig__gcontig_pos %in% Group1_allSuperSNPs$gcontig__gcontig_pos
  
   ### Sanity check - number of SNPs in Group 1 the same
  Group1_nSNPs <- sum(results_pine3$Group1_logical)
  
  ### Sanity check - number of contigs in Group 1 the same
  (Group1_ncontigs <-length(levels(factor(results_pine3$gtcontig[results_pine3$Group1_logical]))))
  
  G1_hist <- sort(table(results_pine3$gtcontig[results_pine3$Group1_logical]),decreasing = TRUE)
  
  barplot(G1_hist, xlab="Contig", ylab="Number of SNPs", main=paste("Distribution of",Group1_nSNPs, "top SNPs", "in",Group1_ncontigs, "contigs"), las=2, cex.names=0.3)
  
  top_contigs <- names(G1_hist)
```


### Compare my list to Yeaman 2016 convergence candidates
These are genes we think are true positives because they adapt to climate in both pine and spruce.

Yeaman et al. 2016 identified 47 orthologs with FDR correction

```{r}
sy <- read.table("../data/orthologs/parallelism_genes_pine_spruce_q05.txt")
head(sy)
nrow(sy)
sy_top <- unique(sy$V1)
length(sy_top)
length(unique(sy$V2))
  ### Make sure all contigs in list are in my df
  for (i in 1:length(sy_top)){
    cd <- as.character(sy_top[i])
    if(length(which(sy_top[i]==results_pine3$gtcontig))==0){
      print(c(i, "missing",cd))
    }
  }

### Which ones of Group 1 are super convergent outliers
  sum(top_contigs %in% sy$V1)
  sum(sy$V1 %in% top_contigs)
  
  top_contigs[which(top_contigs %in% sy$V1)]
```
According to the above results, 10 of those 47 are in my list of top candidates.

We can also use a less restrictive list, e.g. orthologs that are significant at P < 0.05 without FDR correction.

TO DO. Can do spruce but need to use orthology table.

```{r, eval=FALSE, echo=FALSE}
sy2 <- read.table("../data/orthologs/pine_orthologs_to_spruce_top_candidates_p05.txt", header=TRUE)
head(sy2)
nrow(sy2)

orthoP05 <- unique(sy2$tcontig[sy2$pvalue < 0.05])
length(orthoP05)
  ### Which ones of Group 1 are super convergent outliers
  sum(top_contigs %in% orthoP05)
  
  top_contigs[which(top_contigs %in% orthoP05)]
```



### Group 1 Make some comparison plots
```{r}
sum(results_pine3$Group1_logical)
sum(results_pine3$Group1_logical & !is.na(results_pine3$Ances_flipNum), na.rm=TRUE)

names(results_pine3)[grep("MAT", names(results_pine3))]

par(mfrow=c(2,1))
plot2Dcov(x=results_pine3$MAT_rhoave,
       y=results_pine3$MAP_rhoave, 
       xlab= "MAT rho Bayenv", 
       ylab= "MAP rho Bayenv" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=results_pine3$MAT_rhoave[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_rhoave[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))

plot2Dcov(x=results_pine3$MAT_rhoave*results_pine3$Ances_flipNum,
       y=results_pine3$MAP_rhoave*results_pine3$Ances_flipNum, 
       xlab= "MAT rho Bayenv", 
       ylab= "MAP rho Bayenv" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=results_pine3$MAT_rhoave[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_rhoave[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))

plot2Dcov(x=results_pine3$MAT_raw_rho,
       y=results_pine3$MAP_raw_rho, 
       xlab= "MAT rho raw", 
       ylab= "MAP rho raw" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=results_pine3$MAT_raw_rho[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_raw_rho[results_pine3$Group1_logical])

plot2Dcov(x=results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum,
       y=results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum, 
       xlab= "MAT rho raw", 
       ylab= "MAP rho raw" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=results_pine3$MAT_raw_rho[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_raw_rho[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))
```


### Color outliers by various variables to determine variables that 
The last figure, especially the Bayenv-corrected one, suggests that there are two groups of SNPs that are behaving differently in the data. We used data visualization to determine which environmental variables (if any) could explain the groupings.

The rationale is that by subsetting the data to outliers in each environmental variable, and re-visualizing the MAP-MAT comparison, we should be able to "see" which variables determine the groupings.

put figures into results/groupingVisualizations

#### names
```{r}

rawp_names <- names(results_pine3)[cols_raw ] # raw p columns
rawrho_names <- names(results_pine3)[env_cols] # raw rho columns

bayenvBF_names <- sub("_raw_p","",names(results_pine3)[cols_raw])# bayenv BF cols
bayenvrho_names <- paste(bayenvBF_names, "rhoave", sep="_")

cols_bayenvBF <- which(names(results_pine3) %in% bayenvBF_names)
cols_bayenvrho <-which(names(results_pine3) %in% bayenvrho_names)
  
  # The first 22 variables are environmental variables. 
#env_cols <- cols_raw[c(1:22)]
#names(results_pine3)[env_cols]


pdf(file = "../results/1groupingVisualizations.pdf", width=8, height=8)
### Make the plots first will ALL snps
par(mfrow=c(2,2), mar=c(5,5, 1,1))


plot2Dcov(x=results_pine3$MAT_raw_rho,
       y=results_pine3$MAP_raw_rho, 
       xlab= "Spearman's rho between\nallele freq. and MAT", 
       ylab= "Spearman's rho between\nallele freq. and MAP" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=results_pine3$MAT_raw_rho[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_raw_rho[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))

plot2Dcov(x=results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum,
       y=results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum, 
       xlab= "Spearman's rho between\nallele freq. and MAT (derived)", 
       ylab= "Spearman's rho between\nallele freq. and MAP (derived)" ,
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=results_pine3$MAT_raw_rho[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_raw_rho[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))


plot2Dcov(x=results_pine3$MAT_rhoave,
       y=results_pine3$MAP_rhoave, 
       xlab= "Spearman's rho between structure-\ncorrected allele freq. and MAT",
       ylab= "Spearman's rho between structure-\ncorrected allele freq. and MAP" ,
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=results_pine3$MAT_rhoave[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_rhoave[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))

plot2Dcov(x=results_pine3$MAT_rhoave*results_pine3$Ances_flipNum,
       y=results_pine3$MAP_rhoave*results_pine3$Ances_flipNum, 
       xlab= "Spearman's rho between structure-\ncorrected allele freq. and MAT (derived)", 
       ylab= "Spearman's rho between structure-\ncorrected allele freq. and MAP (derived)" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=results_pine3$MAT_rhoave[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical], 
       y_sub_orange=results_pine3$MAP_rhoave[results_pine3$Group1_logical]*results_pine3$Ances_flipNum[results_pine3$Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))


for (i in 1:length(env_cols)){
  #focalvar <- results_pine3[,env_cols[i]]
  focalvarname <- names(results_pine3)[env_cols[i]]
  print(focalvarname)
  #a<- sort(abs(focalvar), na.last = NA)
  #sum(is.na(a))
  #sorted.index <- round(length(a)*.999)
  #rho_cutoff <- a[sorted.index]
  #rho_cutoff
  x_sub_bayenv <- superdf$MAT_rhoave[is.superdf[,i]]
  y_sub_bayenv <- superdf$MAP_rhoave[is.superdf[,i]]

  x_sub_raw <- superdf$MAT_raw_rho[is.superdf[,i]]
  y_sub_raw <- superdf$MAP_raw_rho[is.superdf[,i]]  
  
  anc_sub <- superdf$Ances_flipNum[is.superdf[,i]]
  
  print(length(x_sub_raw))
  
  par(mfrow=c(2,2))
  
  plot2Dcov(x=results_pine3$MAT_raw_rho,
         y=results_pine3$MAP_raw_rho, 
         xlab= "Spearman's rho between\nallele freq. and MAT", 
         ylab= "Spearman's rho between\nallele freq. and MAP" , 
         nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
         x_sub_green=x_sub_raw, 
         y_sub_green=y_sub_raw,
         PE=0.33, PElab=c("MAT", "MAP"))
  
    plot2Dcov(x=results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum,
         y=results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum, 
         xlab= "Spearman's rho between\nallele freq. and MAT (derived)", 
         ylab= "Spearman's rho between\nallele freq. and MAP (derived)" , 
         nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
         x_sub_green=x_sub_raw*anc_sub, 
         y_sub_green=y_sub_raw*anc_sub,
         PE=0.33, PElab=c("MAT", "MAP"))
    
  plot2Dcov(x=results_pine3$MAT_rhoave,
         y=results_pine3$MAP_rhoave, 
         xlab= "Spearman's rho between structure-\ncorrected allele freq. and MAT", 
         ylab= "Spearman's rho between structure-\ncorrected allele freq. and MAP" , 
         nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
         x_sub_green=x_sub_bayenv, 
         y_sub_green=y_sub_bayenv,
         PE=0.33, PElab=c("MAT", "MAP"))
  mtext(focalvarname,side = 3)
  
  plot2Dcov(x=results_pine3$MAT_rhoave*results_pine3$Ances_flipNum,
         y=results_pine3$MAP_rhoave*results_pine3$Ances_flipNum, 
         xlab= "Spearman's rho between structure-\ncorrected allele freq. and MAT (derived)", 
         ylab= "Spearman's rho between structure-\ncorrected allele freq. and MAP (derived)" , 
         nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
         x_sub_green=x_sub_bayenv*anc_sub, 
         y_sub_green=y_sub_bayenv*anc_sub,
         PE=0.33, PElab=c("MAT", "MAP"))
  mtext(focalvarname,side = 3)
  
}# end for loop
dev.off() #end pdf
```

```{r}
pdf(file = "../results/1groupingVisualizationsForSupp.pdf", width=10, height=7)
par(mfcol=c(2,3), mar=c(4,2, 1,1), oma=c(2,4,4,0))
for (i in c(1, 6,22)){ # LAT, MCMT, CMD
  focalvarname <- names(results_pine3)[env_cols[i]]
  print(focalvarname)
  #a<- sort(abs(focalvar), na.last = NA)
  #sum(is.na(a))
  #sorted.index <- round(length(a)*.999)
  #rho_cutoff <- a[sorted.index]
  #rho_cutoff
  x_sub_bayenv <- superdf$MAT_rhoave[is.superdf[,i]]
  y_sub_bayenv <- superdf$MAP_rhoave[is.superdf[,i]]

  x_sub_raw <- superdf$MAT_raw_rho[is.superdf[,i]]
  y_sub_raw <- superdf$MAP_raw_rho[is.superdf[,i]]  
  
  anc_sub <- superdf$Ances_flipNum[is.superdf[,i]]
    plot2Dcov(x=results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum,
         y=results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum, 
         xlab= "",#"Spearman's rho between\nallele freq. and MAT", 
         ylab= "",#"Spearman's rho between\nallele freq. and MAP" , 
         nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
         x_sub_green=x_sub_raw*anc_sub, 
         y_sub_green=y_sub_raw*anc_sub,
         PE=0.33, PElab=c("MAT", "MAP"), plot_greenCOV = FALSE)
      if(i==6){mtext("Spearman's rho between allele freq. and MAT", side=1, line=3)}
      if(i==1){mtext("Spearman's rho between\nallele freq. and MAP", side=2, line=2.5)}
    if(i==1){mtext("A) Latitude", side=3, line=1)}
    if(i==6){mtext("B) Mean temp. in coldest month", side=3, line=1)}
    if(i==22){mtext("C) Climate-moisture deficit", side=3, line=1)}
    
    plot2Dcov(x=results_pine3$MAT_rhoave*results_pine3$Ances_flipNum,
         y=results_pine3$MAP_rhoave*results_pine3$Ances_flipNum, 
         xlab= "", #"Spearman's rho between structure-\ncorrected allele freq. and MAT", 
         ylab= "", #"Spearman's rho between structure-\ncorrected allele freq. and MAP" , 
         nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
         x_sub_green=x_sub_bayenv*anc_sub, 
         y_sub_green=y_sub_bayenv*anc_sub,
         PE=0.33, PElab=c("MAT", "MAP"), plot_greenCOV = FALSE)
  #mtext(focalvarname,side = 3)
        if(i==6){mtext("Spearman's rho between structure-corrected allele freq. and MAT", side=1, line=3)}
        if(i==1){mtext("Spearman's rho between structure-\ncorrected allele freq. and MAP", side=2, line=2.5)}
}
mtext("")
dev.off()
```

### Inspection of the above plotting reveals two groups:

The Freezing Group:

Outliers from these variables fall into the 1st and 3rd quadrants of the MAT-MAP comparison:

LONG, Elevation, MAT,  MCMT, TD, DD0, DD5, NFFD, eFFP, EMT


The Aridity Group:

Outliers from these variables fall into the 2nd and 4th quadrants of the MAT-MAP comparison:

MSP, AHM, SHM, EXT, Eref, CMD

Variables with outliers in Both Groups:
LAT (all quadrants)
Other variables don't have a lot of outliers

## Determine more objectively which variables predict SNP behavior
The above deliniation of groups is based on visualizing the data. We can test more objectively whether outliers from any particular variable have SNPs that fall into more into the hot/wet & cold/dry (1st quadrant and 3rd quadrant) spectrum or more into the cold/wet & hot/dry (2nd and 4th quadrants) spectrum of the MAT-MAP comparison.

The null hypothesis is that 50% of outlier SNPs from any particular variable fall into the 1st/3rd quandrants and 50% fall into the 2nd/4th quadrants. We can test this hypothesis with a binomial test, with the number of trials $n$ equal to the the number of outlier SNPs in that variable, the number of successes equal to the number of SNPs that fall into the 1st/3rd quadrants, and the probability of success $p = 0.5$.

This this is just a metric for identifying whether a given variable should be included in one group or not, so although come contigs have multiple outlier SNPS, non-independence is not a big problem (weâ€™re not saying anything empirical about the genes within each variable at this stage). The only problem is if one big gene happens to be predominantly in one quadrant more often by chance, then we false-positive associate the variable with that group. This is a statistical issue we are aware of and at this stage a more complex solution is not particularly necessary. 

```{r}

dim(Group1_allSuperSNPs)
write.csv(Group1_allSuperSNPs[,c(1:35,cols_raw,cols_bayenvBF,env_cols,cols_bayenvrho)], file = "../results/Group1.csv")

summary(results_pine3$xtx)
quantile(results_pine3$xtx, probs = 0.999)

pdf("../results/XTXhist.pdf", width=4, height=7)
  names(Group1_allSuperSNPs[,env_cols])
  par(mfrow=c(2,1), mar=c(4,4,2,0))
  b<-hist(results_pine3$xtx, main="All loci", xlab="")
  hist(Group1_allSuperSNPs$xtx, breaks=b$breaks, main="GEA Outlier SNPs", xlab="XTX")
dev.off()

```

```{r}
is.superdf <- data.frame(is.superdf)
dim(is.superdf)
names(is.superdf)
names(superdf)

binom_result.df <- data.frame(index=1:22, env=NA, n=NA,  binom.raw.P=NA, binom.raw.est=NA, binom.bay.P=NA, binom.bay.est=NA)

for (i in 1:22){
  logical_var <- is.superdf[,i]
  
  binom_result.df$env[i] <- names(is.superdf)[i]
  
  ### number of SNPs in test
  n <- sum(logical_var, na.rm=TRUE)
  binom_result.df$n[i] <- n
  if(n>5){
  MAT <- superdf$MAT_raw_rho[logical_var==TRUE]
  MAP <- superdf$MAP_raw_rho[logical_var==TRUE]
  
  MAT_bay <- superdf$MAT_rhoave[logical_var]
  MAP_bay <- superdf$MAP_rhoave[logical_var]
  
  ### MAT x MAP
  ### If >0, SNP in 1st or 3rd quadrant
  ### If <0, SNP in 2nd or 4th quadrant
  success <- sum(MAT*MAP>0, na.rm=TRUE)
  fail <- sum(MAT*MAP<0, na.rm=TRUE)
  
  ### Sanity check
  n ==  (success + fail)
  
  raw_test <- binom.test(n=n, x=success, p=0.5)
  
  bayenv_test <- binom.test(n=n, x=sum(MAT_bay*MAP_bay>0, na.rm=TRUE), p=0.5)
  
  binom_result.df$binom.raw.P[i] <- log(raw_test$p.value,10)
  
  binom_result.df$binom.bay.P[i] <- log(bayenv_test$p.value,10)
  
  binom_result.df$binom.raw.est[i] <- raw_test$estimate
  binom_result.df$binom.bay.est[i] <- bayenv_test$estimate
  }
}# end loop


# With 22 variables and two types we'll take an alpha of (0.05/44)
0.05/22
# To compare to our log10-transformed P-values:
(alpha <- log10(0.05/44))

binom_result.df$binom.raw.sig <- binom_result.df$binom.raw.P < alpha

binom_result.df$binom.bay.sig <- binom_result.df$binom.bay.P < alpha


binom_result.df[,3:7] <- round(binom_result.df[,3:7], 2)
binom_result.df$env <- sub("_raw_rho","",binom_result.df$env)
binom_result.df
binom_result.df$Group <- "None"
binom_result.df$Group[binom_result.df$binom.raw.sig & binom_result.df$binom.bay.sig & binom_result.df$binom.bay.est > 0.5 & binom_result.df$binom.raw.est > 0.5] <- "Freezing"

binom_result.df$Group[binom_result.df$binom.raw.sig & binom_result.df$binom.bay.sig & binom_result.df$binom.bay.est < 0.5 & binom_result.df$binom.raw.est < 0.5] <- "Aridity"

binom_result.df

write.csv(binom_result.df, file = "../results/binom_results.csv")
```

The binomial test confirms our visual inspection of the data. In all cases the significance of the result before and after corecting for population structure is the same.

A binomial estimate > 0.50 means the SNPs fall into the 1st and 3rd quadrants in the MAT-MAP comparison. This is the Freezing Group:

```{r}
Fnames <- binom_result.df$env[binom_result.df$Group=="Freezing"]
```

A binomial estimate < 0.50 means the SNPs fall into the 2nd and 4th quadrants in the MAT-MAP comparison. This is the Aridity Group:

```{r}
Anames <- binom_result.df$env[binom_result.df$Group=="Aridity"]
```

## Create a barplot of contigs in environments

```{r}
barplot(colSums(is.superdf), col=as.numeric(as.factor(binom_result.df$Group)), las=2)

is.superdf$gtcontig <- superdf$gtcontig

envi_by_contig <- matrix(NA, nrow=length(unique(is.superdf$gtcontig)), ncol=22)
colnames(envi_by_contig) <- colnames(is.superdf)[1:22]
rownames(envi_by_contig) <- rownames(table(is.superdf$gtcontig, is.superdf[,1]))

for (i in 1:22){
  yo <- table(is.superdf$gtcontig, is.superdf[,i])
  if(ncol(yo)==2){
    envi_by_contig[,i] <- yo[,2]
  }else{
    envi_by_contig[,i] <- 0
  }
}
# Need to get total number of outlier SNPs in temp group or freezing group for each environmental variable
  # each row in is.superdf is a SNP
  # if there is a true in that row for a temp variable, count it:
  is.superdf$IsFreezing <- rowSums(is.superdf[,which(names(is.superdf) %in% Fnames)])>0
  is.superdf$IsAridity <- rowSums(is.superdf[,which(names(is.superdf) %in% Anames)])>0

  envi_by_contig <- data.frame(envi_by_contig)
  envi_by_contig$NumFreezing <- as.numeric(table(is.superdf$IsFreezing, is.superdf$gtcontig)[2,])
  
  envi_by_contig$NumAridity <- as.numeric(table(is.superdf$IsAridity, is.superdf$gtcontig)[2,])

## Limit contigs to those with outliers
envi_by_contig <- envi_by_contig[rowSums(envi_by_contig)>0,]
head(envi_by_contig)

## Rearrange columns into freezing, aridity, and other
state <- rep("Other", 22)
state[colnames(envi_by_contig) %in% Fnames] <- "Freezing"
state[colnames(envi_by_contig) %in% Anames] <- "Aridity"
state
cs <- colSums(envi_by_contig)
colorder <- order(state, cs, decreasing=TRUE)

envi_by_contig <- envi_by_contig[,colorder]

head(envi_by_contig)
colnames(envi_by_contig)
# 1st 4 rows are both groups (1:4)
# Next 5:15 are freezing
# 16:22 are aridity

### Order by rows
envi_by_contig$score <- envi_by_contig$NumFreezing-envi_by_contig$NumAridity
roworder <- order(envi_by_contig$score, decreasing=TRUE)

envi_by_contig2 <- envi_by_contig[roworder,]

barplot(as.matrix(t(envi_by_contig2[,23:24])))

envi_by_contig2$col1 <- two.colors(n=nrow(envi_by_contig2), start="blue", middle="lightblue", end="orange")

plot(rep(-0.90, length(envi_by_contig2$score)),envi_by_contig2$score, col=envi_by_contig2$col1, xaxt="n", xlab="", bty="n", xlim=c(-1,0), pch=20, xaxs="i", cex=2, ylab="")
abline(0,0)
#image.plot(1:length(score), 1, matrix(score,length(score),1), legend.only = TRUE, col = rev(envi_by_contig2$col1)) #not working

plot(envi_by_contig2$score, envi_by_contig2$NumFreezing, col=envi_by_contig2$col1)
plot(envi_by_contig2$score, envi_by_contig2$NumAridity, col=envi_by_contig2$col1)

plot(envi_by_contig2$NumFreezing, envi_by_contig2$NumAridity, col=envi_by_contig2$col1)
# perfect!
plot(envi_by_contig2$AHM_raw_p, col=envi_by_contig2$col1)
barplot(matrix(envi_by_contig2$AHM_raw_p, ncol=1), col=envi_by_contig2$col1, beside=FALSE)
head(envi_by_contig2)

## Plot
plotnames <- sub("_raw_p", "", names(envi_by_contig2[,1:22]))

pdf("../results/EnviBarPlotAwesome.pdf", width=8, height=5)
  par(mar=c(6,4,1,0))
  layout(matrix(c(1,2),nrow=1), widths=c(5,1))
  k<- barplot(as.matrix(envi_by_contig2[,1:22]), beside = FALSE, las=2, col = envi_by_contig2$col1, names.arg = plotnames, ylab="Number of Outlier SNPs")
  
  text(k[2], 600, "Other\nVariables")
  text(k[10], 600, "Freezing\nVariables", col="blue")
  text(k[19], 600, "Aridity\nVariables", col="darkorange")
  
  par(mar=c(6,2,1,0))
  plot(rep(-0.90, length(envi_by_contig2$score)),envi_by_contig2$score, col=envi_by_contig2$col1, xaxt="n", xlab="", bty="n", xlim=c(-1,0), pch=20, xaxs="i", cex=2, ylab="")
  
  mtext("(# Outlier SNPS across Freezing variables) -\n(# Outlier SNPS across Aridity variables) ", side=4, line=-1)
  dev.off()

```

## Subset the SNPs into 2 groups: Freezing, Aridity

### Freezing Group
```{r}
Freezing.cols <- colnames(is.superdf) %in% Fnames
which(Freezing.cols)
FreezingSNPs.df <- superdf[rowSums(is.superdf[,Freezing.cols],na.rm = TRUE)>0,]

  ## Num SNPs
  nrow(FreezingSNPs.df)
  length(unique(FreezingSNPs.df$gcontig__gcontig_pos))
  
  #write.csv(data.frame(FreezingSNPs.df$gcontig__gcontig_pos, FreezingSNPs.df$gtcontig), file = "../results/FreezingGroup.csv")
  
  ## Num Contigs
   length(levels(factor(FreezingSNPs.df$gtcontig)))
   length(unique(FreezingSNPs.df$gtcontig))

 ### Logical Freezing Group
  results_pine3$Freezing_logical <- results_pine3$gcontig__gcontig_pos %in% FreezingSNPs.df$gcontig__gcontig_pos
  
  ### Sanity check - number of SNPs in Group 1 the same
  sum(results_pine3$Freezing_logical)
  
  ### Sanity check - number of contigs in Group 1 the same
  length(levels(factor(results_pine3$gtcontig[results_pine3$Freezing_logical])))
  
  Freezing_hist <- sort(table(results_pine3$gtcontig[results_pine3$Freezing_logical]),decreasing = TRUE)
  
  barplot(Freezing_hist, xlab="Contig", ylab="Number of SNPs", main="Freezing group: X top SNPs in X contigs", las=2, cex.names=0.3)
  
  top_contigs_freezing <- names(Freezing_hist)

  # Which top contigs also in Yeaman 2016?
  top_contigs_freezing[which(top_contigs_freezing %in% sy$V1)]
  
  ## determine top candidates in temp group
  hist(-log10(abs(FreezingSNPs.df$MAT_raw_p)))
  plot(-log10(abs(FreezingSNPs.df$MAT_raw_p)), FreezingSNPs.df$MAT_raw_rho)
  plot(-log10(abs(FreezingSNPs.df$MAT_raw_p)), FreezingSNPs.df$MAT)
  # recall that out SNPs only have to have low P-values in one variable. So the ones with less significant P-values in this plot would be more significant in another plot.
```


#### Plot top freezing outliers
```{r}
    
pdf("../results/TopOutliersFreezing.pdf", width=10, height=10)
for (i in 1:length(Fnames)){
  x <- -log10(abs(FreezingSNPs.df[,which(names(FreezingSNPs.df)==Fnames[i])]))
    # p value
  y <- FreezingSNPs.df[,which(names(FreezingSNPs.df)==sub("_raw_p","",Fnames[i]))]
    # BF
  plot(x,y, main=Fnames[i], cex=4, col=as.numeric(as.factor(FreezingSNPs.df$gtcontig)), ylim=c(2,30), xlim=c(8,45))
  text(x,y,FreezingSNPs.df$gcontig__gcontig_pos, cex=0.2)
  text(x,y-0.2,paste(FreezingSNPs.df$major_allele,
                 FreezingSNPs.df$minor_allele,
                 FreezingSNPs.df$X_annotation,
                 round(FreezingSNPs.df$xtx,0),
                 round(FreezingSNPs.df$major_freq,2)), 
        cex=0.2)
}
dev.off()
```


### Aridity Group
```{r}
Aridity.cols <- colnames(is.superdf) %in% Anames
which(Aridity.cols)
AriditySNPs.df <- superdf[rowSums(is.superdf[,Aridity.cols],na.rm = TRUE)>0,]

  ## Num SNPs
  nrow(AriditySNPs.df)
  length(unique(AriditySNPs.df$gcontig__gcontig_pos))
  
  ## Num Contigs
  length(levels(factor(AriditySNPs.df$gtcontig)))
  length(unique(AriditySNPs.df$gtcontig))
  
 ### Logical Freezing Group
  results_pine3$Aridity_logical <- results_pine3$gcontig__gcontig_pos %in% AriditySNPs.df$gcontig__gcontig_pos
  
  write.csv(data.frame(AriditySNPs.df$gcontig__gcontig_pos, AriditySNPs.df$gtcontig), file = "../results/AridityGroup.csv")
  
  ### Sanity check - number of SNPs in Group 1 the same
  sum(results_pine3$Aridity_logical)
  
  ### Sanity check - number of contigs in Group 1 the same
  length(levels(factor(results_pine3$gtcontig[results_pine3$Aridity_logical])))
  
  Aridity_hist <- sort(table(results_pine3$gtcontig[results_pine3$Aridity_logical]),decreasing = TRUE)
  
  barplot(Aridity_hist, xlab="Contig", ylab="Number of SNPs", main="Aridity group: X top SNPs in X contigs", las=2, cex.names=0.3)
  
  top_contigs_aridity <- names(Aridity_hist)

  # Which top contigs also in Yeaman 2016?
  top_contigs_aridity[which(top_contigs_aridity %in% sy$V1)]
  # none!
```


#### Plot top aridity outliers
```{r}
    
pdf("../results/TopOutliersAridity.pdf", width=10, height=10)
for (i in 1:length(Anames)){
  x <- -log10(abs(AriditySNPs.df[,which(names(AriditySNPs.df)==Anames[i])]))
    # p value
  y <- AriditySNPs.df[,which(names(AriditySNPs.df)==sub("_raw_p","",Anames[i]))]
    # BF
  plot(x,y, main=Anames[i], cex=4, col=as.numeric(as.factor(AriditySNPs.df$gtcontig)), ylim=c(2,15), xlim=c(8,22))
  text(x,y,AriditySNPs.df$gcontig__gcontig_pos, cex=0.2)
  text(x,y-0.1,paste(AriditySNPs.df$major_allele,
                 AriditySNPs.df$minor_allele,
                 AriditySNPs.df$X_annotation,
                 round(AriditySNPs.df$xtx,0),
                 round(AriditySNPs.df$major_freq,2)), 
        cex=0.2)
}
dev.off()
```

### Compare Freezing and Aridity Groups
```{r}
pdf(file = "../results/TempVsAridity_inMATvMAP.pdf", width=4, height=8)

par(mfrow=c(3,1), mar=c(4,6,1,2.5))

plot2Dcov(x=results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum,
       y=results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum, 
       xlab= "Spearman's rho between allele frequency and MAT", 
       ylab= "Spearman's rho between\nallele frequency and MAP" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_green=results_pine3$MAT_raw_rho[results_pine3$Freezing_logical | results_pine3$Aridity_logical]*results_pine3$Ances_flipNum[results_pine3$Freezing_logical | results_pine3$Aridity_logical], 
       y_sub_green=results_pine3$MAP_raw_rho[results_pine3$Freezing_logical | results_pine3$Aridity_logical]*results_pine3$Ances_flipNum[results_pine3$Freezing_logical | results_pine3$Aridity_logical], plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))
  text(-0.7,0.7, "A",cex=1.5)

plot2Dcov(x=results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum,
       y=results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum, 
       xlab= "Spearman's rho between allele frequency and MAT", 
       ylab= "Spearman's rho between\nallele frequency and MAP" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=results_pine3$MAT_raw_rho[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical]*results_pine3$Ances_flipNum[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical], 
       y_sub_orange=results_pine3$MAP_raw_rho[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical]*results_pine3$Ances_flipNum[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical],
       x_sub_blue=results_pine3$MAT_raw_rho[!results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[!results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_blue=results_pine3$MAP_raw_rho[!results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[!results_pine3$Aridity_logical & results_pine3$Freezing_logical],
        x_sub_green=results_pine3$MAT_raw_rho[results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_green=results_pine3$MAP_raw_rho[results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[results_pine3$Aridity_logical & results_pine3$Freezing_logical], plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))
  text(-0.7,0.7, "B",cex=1.5)

plot2Dcov(x=results_pine3$MAT_rhoave*results_pine3$Ances_flipNum,
       y=results_pine3$MAP_rhoave*results_pine3$Ances_flipNum, 
       xlab= "Spearman's rho between structure-corrected allele frequency and MAT", 
       ylab= "Spearman's rho between structure-corrected\nallele frequency and MAP" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=results_pine3$MAT_rhoave[!results_pine3$Freezing_logical & results_pine3$Aridity_logical]*results_pine3$Ances_flipNum[!results_pine3$Freezing_logical & results_pine3$Aridity_logical], 
       y_sub_orange=results_pine3$MAP_rhoave[!results_pine3$Freezing_logical & results_pine3$Aridity_logical]*results_pine3$Ances_flipNum[!results_pine3$Freezing_logical & results_pine3$Aridity_logical],
       x_sub_blue=results_pine3$MAT_rhoave[!results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[!results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_blue=results_pine3$MAP_rhoave[!results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[!results_pine3$Aridity_logical & results_pine3$Freezing_logical],
       x_sub_green=results_pine3$MAT_rhoave[results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_green=results_pine3$MAP_rhoave[results_pine3$Aridity_logical & results_pine3$Freezing_logical]*results_pine3$Ances_flipNum[results_pine3$Aridity_logical & results_pine3$Freezing_logical], plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))
  text(-0.25,0.25, "C",cex=1.5)
dev.off()




pdf(file = "../results/TempVsAridity_inMATvMAP_OutButNotFreezArid.pdf", width=7, height=15)

par(mfrow=c(2,1), mar=c(4,4,1,1))
plot2Dcov(x=results_pine3$MAT_rhoave,
       y=results_pine3$MAP_rhoave, 
       xlab= "MAT rho Bayenv", 
       ylab= "MAP rho Bayenv" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_green=results_pine3$MAT_rhoave[results_pine3$Group1_logical & !results_pine3$Freezing_logical & !results_pine3$Aridity_logical], 
       y_sub_green=results_pine3$MAP_rhoave[results_pine3$Group1_logical & !results_pine3$Freezing_logical & !results_pine3$Aridity_logical],plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))

plot2Dcov(x=results_pine3$MAT_raw_rho,
       y=results_pine3$MAP_raw_rho, 
       xlab= "MAT rho raw", 
       ylab= "MAP rho raw" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
 x_sub_green=results_pine3$MAT_raw_rho[results_pine3$Group1_logical & !results_pine3$Freezing_logical & !results_pine3$Aridity_logical], 
       y_sub_green=results_pine3$MAP_raw_rho[results_pine3$Group1_logical & !results_pine3$Freezing_logical & !results_pine3$Aridity_logical],plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))

dev.off()
```

## Create Venn Tables
```{r}

limit.df <- results_pine3[results_pine3$Group1_logical,]
dim(limit.df)

### Overlap among SNPs
a<- vennCounts(cbind(limit.df$Freezing_logical, limit.df$Aridity_logical))
a
vennDiagram(a, names=c("Freezing", "Aridity", main = "Overlap among SNPs"))

### Overlap among contigs
contig.group.df <- data.frame(
  FreezingOnly = table(limit.df$gtcontig,limit.df$Freezing_logical & !limit.df$Aridity_logical )[,2],
  AridityOnly = table(limit.df$gtcontig,limit.df$Aridity_logical & !limit.df$Freezing_logical)[,2],
  Both = table(limit.df$gtcontig,limit.df$Aridity_logical & limit.df$Freezing_logical)[,2],
  Neither = table(limit.df$gtcontig,!limit.df$Aridity_logical & !limit.df$Freezing_logical)[,2]
)

contig.group.df$index=1:nrow(contig.group.df)

contig.group.df<- contig.group.df[order( contig.group.df$FreezingOnly, contig.group.df$Both, contig.group.df$AridityOnly, decreasing=TRUE),]

contig.group.df
nrow(contig.group.df)

# Number of SNPs in both groups
sum(contig.group.df$Both)
# Number of contigs with outlier SNPs in both groups
sum(contig.group.df$Both>0)

# Num SNPs
pdf("../results/PineVenn.pdf", width=6, height=10)
  par(mar=c(4,6, 0,1))
  barplot(t(as.matrix(contig.group.df[,c(1,3,2)])), beside = FALSE, horiz = TRUE, las=1, col=c("lightblue", "lightgreen", "orange"), cex.names=0.3, xlab="Number of SNPs", border = c("blue", "darkgreen", "brown"))
  legend(30, 120, c("Freezing", "Both", "Aridity"), bty="n", fill=c("lightblue", "lightgreen", "orange"), border = c("blue", "darkgreen", "brown"), cex=1.5)
dev.off()
```

## Galaxy in multivariate environment raw
```{r}
load("../data/PineHeatmap.Rdata")
# loads env_clust
head(env_clust$carpet)

rawp_names

whichvars <- c("LAT", "LONG", "ELEVATION", "MAT", "MAP", "CMD") #Eref

pdf("../results/multvariateEnviGalaxy.pdf", width=22, height=20)
par(mfrow=c(length(whichvars), length(whichvars)), mar=c(0.3,0.7,0,0.7), oma=c(10,10,0,0))
for (i in 1:length(whichvars)){
  for (j in 1:length(whichvars)){
    if(i==j){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      text(1,1, whichvars[i], cex=5)  
    }
    if(j>i){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      if(j==2 & i==1){
        legend(1,1, legend = c("Freezing", "Aridity", "Both"), 
               pch=c(22,24,21), cex=4, bty="n",xjust=0.5,yjust=0.5,
               col=c("blue", "brown", "darkgreen"), 
               pt.bg=c("lightblue", "orange", "lightgreen"))
        
      }
    }
      
    if(i>j){
      print(c(i,j))
      print(c(whichvars[j], whichvars[i]))
      # Get EE corr
      ri <- grep(paste("^",whichvars[j],sep=""),rownames(env_clust$carpet))
      cj <- grep(paste("^",whichvars[i],sep=""),colnames(env_clust$carpet))
      print(c(rownames(env_clust$carpet)[ri], colnames(env_clust$carpet)[cj]))
      PE<- round(env_clust$carpet[ri,cj],2)
      
      whichx <- which(names(results_pine3)==paste(whichvars[j], "_raw_rho", sep=""))
      whichy <- which(names(results_pine3)==paste(whichvars[i], "_raw_rho", sep=""))
      print(names(results_pine3)[c(whichx,whichy)])
      x <- results_pine3[,whichx]*results_pine3$Ances_flipNum
      y <- results_pine3[,whichy]*results_pine3$Ances_flipNum
      
      plot2Dcov(x=x,
       y=y, 
       xlab= "", 
       ylab= "" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
  x_sub_orange=x[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical], 
       y_sub_orange=y[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical],
       x_sub_blue=x[!results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_blue=y[!results_pine3$Aridity_logical & results_pine3$Freezing_logical],
        x_sub_green=x[results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_green=y[results_pine3$Aridity_logical & results_pine3$Freezing_logical], plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=PE, PElab=c(whichvars[i], whichvars[j]))
    } # end if
  } # end j
} # end i
  mtext(text = "Correlation between allele frequency and environment (in column)", side=1, outer=TRUE, line=5, cex=3, adj=0.1)
  mtext(text = "Correlation between allele frequency and environment (in row)", side=2, outer=TRUE, line=5, cex=3, adj=0.1)
dev.off()
```

## Galaxy in multivariate environment bayenv
```{r}

pdf("../results/multvariateEnviGalaxyBayenv.pdf", width=22, height=20)
par(mfrow=c(length(whichvars), length(whichvars)), mar=c(0.3,0.7,0,0.7), oma=c(10,10,0,0))
for (i in 1:length(whichvars)){
  for (j in 1:length(whichvars)){
    if(i==j){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      text(1,1, whichvars[i], cex=5)  
    }
    if(j>i){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      if(j==2 & i==1){
        legend(1,1, legend = c("Freezing", "Aridity", "Both"), 
               pch=c(22,24,21), cex=4, bty="n",xjust=0.5,yjust=0.5,
               col=c("blue", "brown", "darkgreen"), 
               pt.bg=c("lightblue", "orange", "lightgreen"))
        
      }
    }
      
    if(i>j){
      print(c(i,j))
      print(c(whichvars[j], whichvars[i]))
      # Get EE corr
      ri <- grep(paste("^",whichvars[j],sep=""),rownames(env_clust$carpet))
      cj <- grep(paste("^",whichvars[i],sep=""),colnames(env_clust$carpet))
      print(c(rownames(env_clust$carpet)[ri], colnames(env_clust$carpet)[cj]))
      PE<- round(env_clust$carpet[ri,cj],2)
      
      whichx <- which(names(results_pine3)==paste(whichvars[j], "_rhoave", sep=""))
      whichy <- which(names(results_pine3)==paste(whichvars[i], "_rhoave", sep=""))
      print(names(results_pine3)[c(whichx,whichy)])
      x <- results_pine3[,whichx]*results_pine3$Ances_flipNum
      y <- results_pine3[,whichy]*results_pine3$Ances_flipNum
      
      plot2Dcov(x=x,
       y=y, 
       xlab= "", 
       ylab= "" , 
       nbin=100, xlim=c(-0.4,0.4), ylim=c(-0.4,0.4),
  x_sub_orange=x[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical], 
       y_sub_orange=y[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical],
       x_sub_blue=x[!results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_blue=y[!results_pine3$Aridity_logical & results_pine3$Freezing_logical],
        x_sub_green=x[results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_green=y[results_pine3$Aridity_logical & results_pine3$Freezing_logical], plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=PE, PElab=c(whichvars[i], whichvars[j]))
    } # end if
  } # end j
} # end i
  mtext(text = "Correlation between allele frequency and environment (in column)", side=1, outer=TRUE, line=5, cex=3, adj=0.1)
  mtext(text = "Correlation between allele frequency and environment (in row)", side=2, outer=TRUE, line=5, cex=3, adj=0.1)
dev.off()
```


## Galaxy in multivariate environment raw WITH 3x3
```{r}


rawp_names
whichvars1 <- c("LAT", "LONG", "ELEVATION") #Eref #i columsn
whichvars2 <- c("MAT", "MAP", "CMD") # j rows

pdf("../results/multvariateEnviGalaxy3x3.pdf", width=22, height=20)
par(mfrow=c(length(whichvars1), length(whichvars2)), mar=c(0.3,0.7,0,0.7), oma=c(10,10,0,0))
for (i in 1:length(whichvars1)){
  for (j in 1:length(whichvars2)){
    #if(i==j){
    #  plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
    #  text(1,1, whichvars[i], cex=5)  
    # }
    # if(j>i){
    #   plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
    #   if(j==2 & i==1){
    #     legend(1,1, legend = c("Freezing", "Aridity", "Both"), 
    #            pch=c(22,24,21), cex=4, bty="n",xjust=0.5,yjust=0.5,
    #            col=c("blue", "brown", "darkgreen"), 
    #            pt.bg=c("lightblue", "orange", "lightgreen"))
    #     
    #  }
    #}
      
    #if(i>j){
      print(c(i,j))
      print(c(whichvars1[j], whichvars2[i]))
      # Get EE corr
      ri <- grep(paste("^",whichvars1[j],sep=""),rownames(env_clust$carpet))
      cj <- grep(paste("^",whichvars2[i],sep=""),colnames(env_clust$carpet))
      print(c(rownames(env_clust$carpet)[ri], colnames(env_clust$carpet)[cj]))
      PE<- round(env_clust$carpet[ri,cj],2)
      
      whichx <- which(names(results_pine3)==paste(whichvars1[j], "_raw_rho", sep=""))
      whichy <- which(names(results_pine3)==paste(whichvars2[i], "_raw_rho", sep=""))
      print(names(results_pine3)[c(whichx,whichy)])
      x <- results_pine3[,whichx]
      y <- results_pine3[,whichy]
      
      plot2Dcov(x=x,
       y=y, 
       xlab= "", 
       ylab= "" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
  x_sub_orange=x[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical], 
       y_sub_orange=y[!results_pine3$Freezing_logical  & results_pine3$Aridity_logical],
       x_sub_blue=x[!results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_blue=y[!results_pine3$Aridity_logical & results_pine3$Freezing_logical],
        x_sub_green=x[results_pine3$Aridity_logical & results_pine3$Freezing_logical], 
       y_sub_green=y[results_pine3$Aridity_logical & results_pine3$Freezing_logical], plot_greenCOV = FALSE, plot_orangeCOV = FALSE, plot_blueCOV = FALSE,
       PE=PE, PElab=c(whichvars2[i], whichvars1[j]))
    } # end if
  } # end j
} # end i
  mtext(text = "Correlation between allele frequency and environment (in column)", side=1, outer=TRUE, line=5, cex=3, adj=0.1)
  mtext(text = "Correlation between allele frequency and environment (in row)", side=2, outer=TRUE, line=5, cex=3, adj=0.1)
dev.off()
```


## Dendogram

```{r}
install.packages('dendextend')
library(dendextend)
# First, take absolute value of the SNP effect sizes b/c of signed nature of alleles

results_pine3$type <- NA
results_pine3$type[results_pine3$Freezing_logical & !results_pine3$Aridity_logical] <- "Freezing"
results_pine3$type[!results_pine3$Freezing_logical & results_pine3$Aridity_logical] <- "Aridity"
results_pine3$type[results_pine3$Freezing_logical & results_pine3$Aridity_logical] <- "Both"
levels(factor(results_pine3$type))
results_pine3$type[!results_pine3$Freezing_logical & !results_pine3$Aridity_logical] <- "Other"
levels(factor(results_pine3$type))

install.packages("devtools")
devtools::install_github("rlbarter/superheat")
library(superheat)
library(RColorBrewer)


#env_order<-hclust(dist(t(abs(results_pine3[results_pine3$Group1_logical, env_cols]))))
#str(env_order)
#env_order$labels[env_order$order]

snp_order <- hclust(dist(abs(results_pine3[results_pine3$Group1_logical, env_cols])))
head(snp_order$labels[snp_order$order])


groupCodes <- results_pine3$type[results_pine3$Group1_logical]
clusMember = cutree(h, 3)
table(results_pine3$type[results_pine3$Group1_logical], clusMember)
  # cluster 2 is mostly freezing
  # some freezing also in cluster 3

clusMember = cutree(h, 4)
table(results_pine3$type[results_pine3$Group1_logical], clusMember)
length(clusMember)
results_pine3$cluster[results_pine3$Group1_logical] <- clusMember

"The machine learns but we don't necessarily learn much"

sample = (results_pine3[results_pine3$Group1_logical, env_cols])
colnames(sample) <- sub("_raw_rho", "", colnames(sample))

#png("../results/dendogram.png", width=12, height=9, units = "in", res=500)
#superheat(t(sample), 
#          heat.pal = (brewer.pal(5, "Blues")), heat.na.col = "white",
#          col.dendrogram = T, row.dendrogram = TRUE,
#            left.label.size = 0.45)
#dev.off()

png("../results/dendogram_ordered.png", width=12, height=9, units = "in", res=500)
superheat(t(sample), 
          order.rows=rev(env_clust$rowInd),
          order.cols = snp_order$order,
          heat.pal = (brewer.pal(5, "Blues")), heat.na.col = "white",
          col.dendrogram = T, row.dendrogram = TRUE,
            left.label.size = 0.45)
dev.off()


sample_bayenv = (results_pine3[results_pine3$Group1_logical, cols_bayenvrho])
colnames(sample_bayenv) <- sub("_rhoave", "", colnames(sample))
clusMember = cutree(hclust(dist(abs(sample_bayenv))), 3)
head(clusMember)
table(results_pine3$type[results_pine3$Group1_logical], clusMember)


#png("../results/dendogram_bayenv.png", width=12, height=9, units = "in", res=500)
#superheat(t(sample_bayenv), 
#          heat.pal = (brewer.pal(5, "Blues")), heat.na.col = "white",
#          col.dendrogram = T, row.dendrogram = TRUE,
#            left.label.size = 0.45,
#        order.rows=rev(env_clust$rowInd),)
#dev.off()

png("../results/dendogram_bayenv_ordered.png", width=12, height=9, units = "in", res=500)
  superheat(t(abs(sample_bayenv)), 
          order.rows=rev(env_clust$rowInd),
          heat.pal = (brewer.pal(5, "Blues")),
          heat.na.col = "white",
          col.dendrogram = TRUE, row.dendrogram = TRUE,
            left.label.size = 0.45,
        order.cols = snp_order$order)
dev.off()

```


## MAT vs. MAP with clusters

```{r}

pdf(file = "../results/Clustering_inMATvMAP.pdf", width=7, height=15)




table(results_pine3$type[results_pine3$Group1_logical], clusMember)

quartz()
par(mfrow=c(1,1), mar=c(4,4,1,1))
x<- results_pine3$MAT_raw_rho*results_pine3$Ances_flipNum
y <- results_pine3$MAP_raw_rho*results_pine3$Ances_flipNum

plot2Dcov(x=x, y=y, 
       xlab= "Spearman's rho between allele frequency and MAT", 
       ylab= "Spearman's rho between allele frequency and MAP" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       plot_greenCOV = FALSE, plot_orangeCOV = FALSE, 
       plot_blueCOV = FALSE, plot_yellowCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))

plot2Dcov(x=x, y=y, 
       xlab= "Spearman's rho betweenallele frequency and MAT", 
       ylab= "Spearman's rho between allele frequency and MAP" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=x[results_pine3$cluster==1], 
       y_sub_orange=y[results_pine3$cluster==1],
       x_sub_blue=x[results_pine3$cluster==3], 
       y_sub_blue=y[results_pine3$cluster==3],
       x_sub_green=x[results_pine3$cluster==4], 
       y_sub_green=y[results_pine3$cluster==4], 
       x_sub_yellow=x[results_pine3$cluster==2], 
       y_sub_yellow=y[results_pine3$cluster==2], 
       plot_greenCOV = FALSE, plot_orangeCOV = FALSE, 
       plot_blueCOV = FALSE, plot_yellowCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))
  #text(-0.25,0.25, "C",cex=1.5)
#dev.off()

quartz()
par(mfrow=c(1,1), mar=c(4,6,1,1))
x<- results_pine3$MAT_rhoave*results_pine3$Ances_flipNum
y <- results_pine3$MAP_rhoave*results_pine3$Ances_flipNum
plot2Dcov(x=x, y=y, 
       xlab= "Spearman's rho between structure-corrected allele frequency and MAT", 
       ylab= "Spearman's rho between structure-corrected\nallele frequency and MAP" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=x[results_pine3$cluster==1], 
       y_sub_orange=y[results_pine3$cluster==1],
       x_sub_blue=x[results_pine3$cluster==3], 
       y_sub_blue=y[results_pine3$cluster==3],
       x_sub_green=x[results_pine3$cluster==4], 
       y_sub_green=y[results_pine3$cluster==4], 
       x_sub_yellow=x[results_pine3$cluster==2], 
       y_sub_yellow=y[results_pine3$cluster==2], 
       plot_greenCOV = FALSE, plot_orangeCOV = FALSE, 
       plot_blueCOV = FALSE, plot_yellowCOV = FALSE,
       PE=0.33, PElab=c("MAT", "MAP"))

```



## Galaxy in multivariate environment raw 4 clusters
```{r}
# loads env_clust
head(env_clust$carpet)

rawp_names

whichvars <- c("LAT", "ELEVATION", "MAT","MAP", "CMD") #Eref

pdf("../results/multvariateEnviGalaxy4clusters.pdf", width=22, height=20)
par(mfrow=c(length(whichvars), length(whichvars)), mar=c(0.3,0.7,0,0.7), oma=c(10,10,0,0))
for (i in 1:length(whichvars)){
  for (j in 1:length(whichvars)){
    if(i==j){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      text(1,1, whichvars[i], cex=5)  
    }
    if(j>i){
      plot(1,1,bty="n", xaxt="n", yaxt="n", xlab="", ylab="", col="white")
      if(j==2 & i==1){
        legend(1,1, legend = c("Freezing", "Aridity", "Multi", "Geography"), 
               pch=c(22,24,21), cex=4, bty="n",xjust=0.5,yjust=0.5,
               col=c("blue", "brown", "darkgreen", "brown"), 
               pt.bg=c("lightblue", "orange", "lightgreen", "yellow"))
      }
    }
      
    if(i>j){
      print(c(i,j))
      print(c(whichvars[j], whichvars[i]))
      # Get EE corr
      ri <- grep(paste("^",whichvars[j],sep=""),rownames(env_clust$carpet))
      cj <- grep(paste("^",whichvars[i],sep=""),colnames(env_clust$carpet))
      print(c(rownames(env_clust$carpet)[ri], colnames(env_clust$carpet)[cj]))
      PE<- round(env_clust$carpet[ri,cj],2)
      
      whichx <- which(names(results_pine3)==paste(whichvars[j], "_raw_rho", sep=""))
      whichy <- which(names(results_pine3)==paste(whichvars[i], "_raw_rho", sep=""))
      print(names(results_pine3)[c(whichx,whichy)])
      x <- results_pine3[,whichx]*results_pine3$Ances_flipNum
      y <- results_pine3[,whichy]*results_pine3$Ances_flipNum
      
      plot2Dcov(x=x,
       y=y, 
       xlab= "", 
       ylab= "" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=x[results_pine3$cluster==1], 
       y_sub_orange=y[results_pine3$cluster==1],
       x_sub_blue=x[results_pine3$cluster==3], 
       y_sub_blue=y[results_pine3$cluster==3],
       x_sub_green=x[results_pine3$cluster==4], 
       y_sub_green=y[results_pine3$cluster==4], 
       x_sub_yellow=x[results_pine3$cluster==2], 
       y_sub_yellow=y[results_pine3$cluster==2], 
       plot_greenCOV = FALSE, plot_orangeCOV = FALSE, 
       plot_blueCOV = FALSE, plot_yellowCOV = FALSE,
       PE=PE, PElab=c(whichvars[i], whichvars[j]))
    } # end if
  } # end j
} # end i
  mtext(text = "Correlation between allele frequency and environment (in column)", side=1, outer=TRUE, line=5, cex=3, adj=0.1)
  mtext(text = "Correlation between allele frequency and environment (in row)", side=2, outer=TRUE, line=5, cex=3, adj=0.1)
dev.off()
```



## Create Venn Tables
```{r}

limit.df <- results_pine3[results_pine3$Group1_logical,]
dim(limit.df)

### Overlap among SNPs
a<- vennCounts(cbind(limit.df$cluster==1, # aridity
                     limit.df$cluster==2, # geography
                     limit.df$cluster==3, # freezing
                     limit.df$cluster==4)) # multi
vennDiagram(a)

### Overlap among contigs
contig.group.df <- data.frame(
  Aridity = table(limit.df$gtcontig,limit.df$cluster==1)[,2],
  Geography = table(limit.df$gtcontig,limit.df$cluster==2)[,2],
  Freezing = table(limit.df$gtcontig,limit.df$cluster==3)[,2],
  Multi = table(limit.df$gtcontig,limit.df$cluster==4)[,2]
)

contig.group.df$index=1:nrow(contig.group.df)

contig.group.df<- contig.group.df[order(contig.group.df$Multi, contig.group.df$Freezing,  contig.group.df$Aridity, contig.group.df$Geography, decreasing=TRUE),]

head(contig.group.df)
nrow(contig.group.df)

# Num SNPs
pdf("../results/PineVenn_cluster.pdf", width=6, height=10)
  par(mar=c(4,6, 0,1))
  barplot(t(as.matrix(contig.group.df[,c(4,3,1,2)])), beside = FALSE, horiz = TRUE, las=1, col=c("lightgreen", "lightblue",  "orange", "yellow"), cex.names=0.3, xlab="Number of SNPs", border = c("darkgreen", "blue",  "brown", "brown"))
  #legend(30, 120, c("Both", "Freezing", "Aridity", "Geography"), bty="n", fill=c("lightgreen", "lightblue",  "orange"), border = c( "darkgreen", "blue", "brown"), cex=1.5)
  text(40, 5, "Both", col="darkgreen", cex=1.5, adj=0)
  text(15, 25, "Freezing", col="blue", cex=1.5, adj=0)
  text(15, 65, "Aridity", col="darkorange", cex=1.5, adj=0)
  text(10, 110, "Geography", col="brown", cex=1.5, adj=0)
dev.off()
```
