---
title: "PEGA for Pine Analysis"
author: "Katie Lotterhos"
date: "August 24, 2015"
output: html_document
---

Does the degree to which SNPs evolve to be correlated with the environment depend on the phenotype-environment association of the phenotypes they are associated with?

Basic algorithm:

1. choose environment
2. choose phenotype
3. get PE association
4. get relevant column data for each SNP (bayenv rho, bayenv bf, gwas slope, gwas p-value) using system calls
5. correct rho for different coding from gwas using flip.cor using merge
6. correct rho for PE/GWAS direction (polarize)
    + (PE-correlation sign) * (GWAS-slope sign) = correction sign
7. (remove large values that are not significant)
8. plot GWAS slope (x axis) vs GEA (y axis)
    + color points in 1st and 3rd quadrants in green (signs match expected direction)
    + color points in 2nd and 4th quadrants in purple (signs don't match expected direction)
    
#### Note
Note that the code runs fastest if you work through R in a terminal window and upload the "results_pine" object first.  Once it is uploaded, when the markdown is created it won't try to load the "results_pine" dataframe, which is very large.  To make this markdown file:
```
library(knitr)
library(markdown)
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis")
knit("AnalysisPEGA.Rmd")  # produces the md file
markdownToHTML("AnalysisPEGA.md", "AnalysisPEGA.html")  # converts an md file to html
```
  
Install packages  
```{r, label="install packages"}
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA")
  if (!("hexbin" %in% installed.packages())){install.packages("hexbin")}
  library(hexbin)
  if (!("gridExtra" %in% installed.packages())){install.packages("ash")}
  library(gridExtra)
  if (!("ash" %in% installed.packages())){install.packages("ash")}
  library(ash)
```

Load data, if not already loaded.  The phenotype and environment labels were not consistent among studies.  The `matchXXX` objects are used to match inconsistent labeling.

```{r, label="load data"}
### load data
PE <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2.csv", header=TRUE)
#reorder to make largest effects on top
PE <- PE[rev(order(abs(PE$pine_correlation))),]
head(PE)
plot(PE$pine_correlation, PE$spruce_correlation, pch=20, col="blue")
abline(0,1)
abline(lm(PE$spruce_correlation~PE$pine_correlation), col="blue")

PE_matchGWAS <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGWAS.csv", header=TRUE)
PE_matchGEA <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGEA.csv", header=TRUE)

flip_pine <- read.table("data/large_files/flip_pine.cor", header=TRUE)
head(flip_pine)

flip_spruce <- read.table("data/large_files/flip_spruce.cor", header=TRUE)
head(flip_spruce)

#### read in results_pine if it doesn't exist
if (!("results_pine" %in% ls())){
results_pine <- read.table("data/large_files/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS", header=TRUE, comment.char=" ")
  }
dim(results_pine)

if (!("results_spruce" %in% ls())){
results_spruce <- read.table("data/large_files/var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS", header=TRUE, comment.char=" ")
  }
dim(results_spruce)

### Upload superoutliers file pine 
  pine_raw_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x_raw.txt", header=TRUE)
  pine_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x.txt", header=TRUE)
  pine_raw_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x_raw.txt", header=TRUE)
  pine_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x.txt", header=TRUE)

  head(pine_tcontig)
  head(pine_gcontig)

### Upload superoutliers file spruce 
  spruce_raw_tcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_100x_raw.txt", header=TRUE)
  spruce_tcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_100x.txt", header=TRUE)
  spruce_raw_gcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_gcontigs_100x_raw.txt", header=TRUE)
  spruce_gcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_gcontigs_100x.txt", header=TRUE)

  head(spruce_tcontig)
  head(spruce_gcontig)

  results_pine_head <- names(results_pine)
  results_spruce_head <- names(results_spruce)
  results_spruce$gcontig__gcontig_pos <- paste(results_spruce$gcontig, results_spruce$pos_gcontig, sep="__")
```

Get the list of superoutliers.  There are 4 types for each species: based on transcript or gcontig, and based on raw or corrected for population structure.

```{r, label="get list of superoutliers"}
### Get list of superoutliers pine
  pine_super_t <- as.character(pine_tcontig$pine[pine_tcontig$super_outlier_score > 1])
  pine_indexes_super_t <- which(as.character(results_pine$tcontig) %in% pine_super_t)

  pine_super_t_raw <- as.character(pine_raw_tcontig$pine[pine_raw_tcontig$super_outlier_score > 1])
  pine_indexes_super_t_raw <- which(as.character(results_pine$tcontig) %in% pine_super_t_raw)

  pine_super_g <- as.character(pine_gcontig$pine[pine_gcontig$super_outlier_score > 1])
  pine_indexes_super_g <- which(as.character(results_pine$gcontig) %in% pine_super_g)

  pine_super_g_raw <- as.character(pine_raw_gcontig$pine[pine_raw_gcontig$super_outlier_score > 1])
  pine_indexes_super_g_raw <- which(as.character(results_pine$gcontig) %in% pine_super_g_raw)

  PE$pine_score_ALL <- PE$pine_score_super_t <- PE$pine_score_super_t_raw <- PE$pine_score_super_g <- PE$pine_score_super_g_raw <- NA

### Get list of superoutliers spruce
  spruce_super_t <- as.character(spruce_tcontig$spruce[spruce_tcontig$super_outlier_score > 1])
  spruce_indexes_super_t <- which(as.character(results_spruce$tcontig) %in% spruce_super_t)

  spruce_super_t_raw <- as.character(spruce_raw_tcontig$spruce[spruce_raw_tcontig$super_outlier_score > 1])
  spruce_indexes_super_t_raw <- which(as.character(results_spruce$tcontig) %in% spruce_super_t_raw)

  spruce_super_g <- as.character(spruce_gcontig$spruce[spruce_gcontig$super_outlier_score > 1])
  spruce_indexes_super_g <- which(as.character(results_spruce$gcontig) %in% spruce_super_g)

  spruce_super_g_raw <- as.character(spruce_raw_gcontig$spruce[spruce_raw_gcontig$super_outlier_score > 1])
  spruce_indexes_super_g_raw <- which(as.character(results_spruce$gcontig) %in% spruce_super_g_raw)

  PE$spruce_score_ALL <- PE$spruce_score_super_t <- PE$spruce_score_super_t_raw <- PE$spruce_score_super_g <- PE$spruce_score_super_g_raw <- NA
```

Next, loop through each phenotype-environment combo.  For each combo, flip and polarize the GEA scores so they are in matching direction to the GWAS and PE.  Then, standardize the GWAS so that each phenotype has a variance of 1 and all slopes will be on the same scale.  Plot the architecture and calculate a score.  The score is based on the `sum(GWAS_effect_pine_standardized * GEA_rho_pine_polarized)/n_snps`.  So larger and more positive scores indicate architectures that have more and/or larger effect SNPs for that phenotype in that environment.

```{r, label="Loop through PE file"}
### Loop through PE file

for (i in 1:nrow(PE)){
### Match names
  PE_envi_name <- as.character(PE$Environment[i])
  GEA_envi_name <- as.character(PE_matchGEA$Environment_GEA[which(PE_matchGEA$Environment_PE==PE_envi_name)])
  
  PE_pheno_name <- as.character(PE$Phenotype[i])
  GWAS_pheno_name <- as.character(PE_matchGWAS$Phenotype_GWAS[which(PE_matchGWAS$Phenotype_PE==PE_pheno_name)])
  
  print(c(i, GEA_envi_name, GWAS_pheno_name))

### Get relevant columns from large file (pine)

  GWAS_effect_col_pine_corrected <- which(results_pine_head==gsub("_p","_snp_effect", GWAS_pheno_name))
  if (length(GWAS_effect_col_pine_corrected)!=1){print("ERROR: GWAS_effect_col_pine_corrected does not equal 1")}
  GWAS_effect_pine_corrected <- results_pine[,GWAS_effect_col_pine_corrected]

  GEA_effect_cols_pine_corrected <- grep(paste("^",GEA_envi_name,"_rho", sep=""), results_pine_head)
  if (length(GEA_effect_cols_pine_corrected)!=2){print("ERROR: GEA_effect_cols_pine_corrected does not equal 2")}
  GEA_rho_pine_corrected <- rowMeans(results_pine[,GEA_effect_cols_pine_corrected], na.rm=TRUE)

  GWAS_effect_col_pine_raw <- which(results_pine_head==paste(GWAS_pheno_name, "_raw_rho", sep=""))
  if (length(GWAS_effect_col_pine_raw)!=1){print("ERROR: GWAS_effect_col_pine_raw does not equal 1")}
  GWAS_effect_pine_raw <- results_pine[,GWAS_effect_col_pine_raw]

  GEA_effect_cols_pine_raw <- which(results_pine_head==paste(GEA_envi_name,"_raw_rho", sep=""))
  if (length(GEA_effect_cols_pine_raw)!=1){print("ERROR: GEA_effect_cols_pine_raw does not equal 1")}
  GEA_rho_pine_raw <- results_pine[,GEA_effect_cols_pine_raw]

### Get relevant columns from large file (spruce)
  GWAS_effect_col_spruce_corrected <- which(results_spruce_head==gsub("_p","_snp_effect", GWAS_pheno_name))
  if (length(GWAS_effect_col_spruce_corrected)!=1){print("ERROR: GWAS_effect_col_spruce_corrected does not equal 1")}
  GWAS_effect_spruce_corrected <- results_spruce[,GWAS_effect_col_spruce_corrected]

  GEA_effect_cols_spruce_corrected <- grep(paste("^",GEA_envi_name,"_rho", sep=""), results_spruce_head)
  if (length(GEA_effect_cols_spruce_corrected)!=2){print("ERROR: GEA_effect_cols_spruce_corrected does not equal 2")}
  GEA_rho_spruce_corrected <- rowMeans(results_spruce[,GEA_effect_cols_spruce_corrected], na.rm=TRUE)

  GWAS_effect_col_spruce_raw <- which(results_spruce_head==paste(GWAS_pheno_name, "_raw_rho", sep=""))
  if (length(GWAS_effect_col_spruce_raw)!=1){print("ERROR: GWAS_effect_col_spruce_raw does not equal 1")}
  GWAS_effect_spruce_raw <- results_spruce[,GWAS_effect_col_spruce_raw]

  GEA_effect_cols_spruce_raw <- which(results_spruce_head==paste(GEA_envi_name,"_raw_rho", sep=""))
  if (length(GEA_effect_cols_spruce_raw)!=1){print("ERROR: GEA_effect_cols_spruce_raw does not equal 1")}
  GEA_rho_spruce_raw <- results_spruce[,GEA_effect_cols_spruce_raw]

### flip_pine GEA_rho_pine so sign matches GWAS_effect_pine
  GEA_rho_flip_pine_corrected <- GEA_rho_pine_corrected * flip_pine$X1
  GEA_rho_flip_spruce_corrected <- GEA_rho_spruce_corrected * flip_spruce$X1

### plot corrected vs. raw values
  plot_compare<- function(x,y, xlab, ylab, xlim, ylim){
  data1 <- cbind(x, y)
  data1b <- data1[complete.cases(data1),]
        binned <- bin2(data1b, 
                 matrix(c(-xlim,xlim,-ylim,ylim), 2,2, byrow=TRUE), 
                 nbin=c(100,100))
    binned$nc[binned$nc==0]=NA
    image(seq(-xlim,xlim,length.out = 100), seq(-ylim,ylim, length.out=100),binned$nc,
             xlab=xlab, ylab=ylab, add=FALSE, col=tim.colors(75))
  }

### output plot to pdf
gwas_file <- paste("analysis/figure_raw_vs_corrected/", GWAS_pheno_name, ".pdf",sep="")
if(!(file.exists(gwas_file))){
pdf(gwas_file, width=5, height=8)
  par(mfrow=c(2,1), mar=c(4,4,1,1),oma=c(0,0,3,0))
  plot_compare(GWAS_effect_pine_raw,GWAS_effect_pine_corrected, 
               xlab="GWAS_effect_pine_raw", ylab="GWAS_effect_pine_corrected",
               xlim=1, ylim=3)
  plot_compare(GWAS_effect_spruce_raw,GWAS_effect_spruce_corrected, 
               xlab="GWAS_effect_spruce_raw", ylab="GWAS_effect_spruce_corrected",
               xlim=1, ylim=8)
  mtext(paste(GWAS_pheno_name), side=3, outer=TRUE)
dev.off()
}

gea_file <- paste("analysis/figure_raw_vs_corrected/", GEA_envi_name, ".pdf",sep="")
if(!(file.exists(gea_file))){
pdf(gea_file, width=5, height=8)
  par(mfrow=c(2,1), mar=c(4,4,1,1),oma=c(0,0,3,0))
  plot_compare(GEA_rho_pine_raw, #GEA_rho_pine_corrected, #OR GEA_rho_flip_pine_corrected, 
               xlab="GEA_rho_pine_raw", ylab="GEA_rho_pine_corrected",
               xlim=1, ylim=1)

  plot_compare(GEA_rho_spruce_raw, #GEA_rho_spruce_corrected OR GEA_rho_flip_spruce_corrected, 
               xlab="GEA_rho_spruce_raw", ylab="GEA_rho_spruce_corrected",
               xlim=1, ylim=1)
  mtext(paste(GEA_envi_name), side=3, outer=TRUE)
dev.off()
}
```

Flip and polarize

```{r}
### check that "flip_pine"" order matches "results_pine" order
if (!(identical(flip_pine$gcontig__gcontig_pos, results_pine$gcontig__gcontig_pos))){
  print("ERROR!  SNP order in flip_pine file does not match results_pine file"); break
}
if (!(identical(as.character(flip_spruce$snp_id), results_spruce$gcontig__gcontig_pos))){
  print("ERROR!  SNP order in flip_spruce file does not match results_spruce file"); break
}

### Polarize GEA so sign aligns with PE and GWAS sign: FOR CORRECTED VALUES ONLY
  PE_cor_pine <- PE$pine_correlation[i]
  PE_cor_spruce <- PE$spruce_correlation[i]

  GWAS_sign_pine_corrected <- sign(GWAS_effect_pine_corrected)
  GWAS_sign_pine_corrected[GWAS_sign_pine_corrected==0]=1
  GWAS_sign_spruce_corrected <- sign(GWAS_effect_spruce_corrected)
  GWAS_sign_spruce_corrected[GWAS_sign_spruce_corrected==0]=1

  if (sum(GWAS_sign_pine_corrected==0, na.rm=TRUE)>0){print("ERROR: Some GWAS have 0 sign for polarization, will skew results_pine")} #fixed this with code above
  if (sum(GWAS_sign_spruce_corrected==0, na.rm=TRUE)>0){print("ERROR: Some GWAS have 0 sign for polarization, will skew results_spruce")} #fixed this with code above

### Polarize GEA so sign aligns with PE and GWAS sign: FOR RAW VALUES ONLY
  GWAS_sign_pine_raw <- sign(GWAS_effect_pine_raw)
  GWAS_sign_pine_raw[GWAS_sign_pine_raw==0]=1
  GWAS_sign_spruce_raw <- sign(GWAS_effect_spruce_raw)
  GWAS_sign_spruce_raw[GWAS_sign_spruce_raw==0]=1

### Do Polarization
  if(PE_cor_pine != 0){
  GEA_rho_pine_polarized_corrected <- GEA_rho_flip_pine_corrected * sign(PE_cor_pine) * GWAS_sign_pine_corrected
  GEA_rho_pine_polarized_raw <- GEA_rho_pine_raw * sign(PE_cor_pine) * GWAS_sign_pine_raw
  }else{
  GEA_rho_pine_polarized_corrected <- GEA_rho_flip_pine_corrected * GWAS_sign_pine_corrected  
  GEA_rho_pine_polarized_raw <- GEA_rho_pine_raw  * GWAS_sign_pine_raw
  }

 if(PE_cor_spruce != 0){
  GEA_rho_spruce_polarized_corrected <- GEA_rho_flip_spruce_corrected * sign(PE_cor_spruce) * GWAS_sign_spruce_corrected
  GEA_rho_spruce_polarized_raw <- GEA_rho_spruce_raw * sign(PE_cor_spruce) * GWAS_sign_spruce_raw
  }else{
  GEA_rho_spruce_polarized_corrected <- GEA_rho_flip_spruce_corrected * GWAS_sign_spruce_corrected 
  GEA_rho_spruce_polarized_raw <- GEA_rho_spruce_raw * GWAS_sign_spruce_raw
  }

### Standardize GWAS for corrected values

GWAS_effect_pine_standardized_corrected <- GWAS_effect_pine_corrected/sd(GWAS_effect_pine_corrected, na.rm=TRUE)
GWAS_effect_spruce_standardized_corrected <- GWAS_effect_spruce_corrected/sd(GWAS_effect_spruce_corrected, na.rm=TRUE)





### Plotting 

plotter <- function(GWAS_effect, GEA_rho_polarized, main, xlim, ylim){
  data1 <- cbind(GWAS_effect, GEA_rho_polarized)
  data1b <- data1[complete.cases(data1),]
  score1 <- sum(data1b[,1]*data1b[,2])
  score1b <- round(score1/nrow(data1b)*10000,1)
  score2 <- round(sum((sign(data1b[,1])*sign(data1b[,2]))==1)/nrow(data1b),2)
    binned <- bin2(data1b, 
                 matrix(c(-xlim,xlim,-ylim,ylim), 2,2, byrow=TRUE), 
                 nbin=c(100,100))
    if(binned$nskip!=0){print("Error in binning, some values not counted. This affects score."); break}
    binned$nc[binned$nc==0]=NA
    image.plot(seq(-xlim,xlim,length.out = 100), seq(-ylim,ylim, length.out=100),binned$nc,
             xlab="GWAS effect standardized", ylab="GEA rho (polarized)", 
             main=paste(main, ", Score1 =",score1b, "Score2 =", score2))
  return(c(score1b, score2))
  }

  png(paste("analysis/figure_byPheno/PEGA_" ,PE_pheno_name,"_",i, "_", PE_envi_name,".png",sep=""), width=10, height=18, units="in", res=300)
  par(mfrow=c(2,2), mar=c(4,4,4,1), oma=c(0,0,3,0))

  #PE$pine_score_ALL[i]<-
  plotter(GWAS_effect_pine_standardized_corrected, GEA_rho_pine_polarized_corrected, main="All SNPs Standardized", xlim=13, ylim=0.5)
  #PE$spruce_score_ALL[i]<-
  plotter(GWAS_effect_spruce_standardized_corrected, GEA_rho_spruce_polarized_corrected, main="All SNPs Standardized", xlim=13, ylim=0.5)

  plotter(GWAS_effect_pine_raw, GEA_rho_pine_polarized_raw, main="All SNPs RAW", xlim=1, ylim=1)
  plotter(GWAS_effect_spruce_raw, GEA_rho_spruce_polarized_raw, main="All SNPs RAW", xlim=1, ylim=1)
  dev.off()

#  plotter(GWAS_effect_pine_corrected, GEA_rho_pine_polarized_corrected, main="PINE All SNPs", is.standard=FALSE)
#  plotter(GWAS_effect_spruce_corrected, GEA_rho_spruce_polarized_corrected, main="SPRUCE All SNPs", is.standard=FALSE)
 #  PE$pine_score_super_t[i]<-plotter(GWAS_effect_pine_standardized[pine_indexes_super_t], GEA_rho_pine_polarized[pine_indexes_super_t], main="Super tcontigs", is.standard=TRUE)
#   PE$spruce_score_super_t[i]<-plotter(GWAS_effect_spruce_standardized[spruce_indexes_super_t], GEA_rho_spruce_polarized[spruce_indexes_super_t], main="Super tcontigs", is.standard=TRUE)
# 
#   PE$pine_score_super_t_raw[i]<-plotter(GWAS_effect_pine_standardized[pine_indexes_super_t_raw], GEA_rho_pine_polarized[pine_indexes_super_t_raw], main="Super tcontigs RAW", is.standard=TRUE)
#   PE$spruce_score_super_t_raw[i]<-plotter(GWAS_effect_spruce_standardized[spruce_indexes_super_t_raw], GEA_rho_spruce_polarized[spruce_indexes_super_t_raw], main="Super tcontigs RAW", is.standard=TRUE)
# 
#   PE$pine_score_super_g[i]<-plotter(GWAS_effect_pine_standardized[pine_indexes_super_g], GEA_rho_pine_polarized[pine_indexes_super_g], main="Super gcontigs", is.standard=TRUE)
#   PE$spruce_score_super_g[i]<-plotter(GWAS_effect_spruce_standardized[spruce_indexes_super_g], GEA_rho_spruce_polarized[spruce_indexes_super_g], main="Super gcontigs", is.standard=TRUE)
# 
#   PE$pine_score_super_g_raw[i]<-plotter(GWAS_effect_pine_standardized[pine_indexes_super_g_raw], GEA_rho_pine_polarized[pine_indexes_super_g_raw], main="Super gcontigs RAW", is.standard=TRUE)
# 
#   PE$spruce_score_super_g_raw[i]<-plotter(GWAS_effect_spruce_standardized[spruce_indexes_super_g_raw], GEA_rho_spruce_polarized[spruce_indexes_super_g_raw], main="Super gcontigs RAW", is.standard=TRUE)
# 
#   title("") #not sure but needed to make mtext work
#   mtext(paste(PE_pheno_name, PE_envi_name, "\nPine", PE$pine_correlation[i], 
#               "\nSpruce", PE$spruce_correlation[i], sep=" "),
#         side=3, outer=TRUE, line=-1)


} #end loop
```

Now that all the scores have been calculated, we can look at the relationship between the PE association and the underlying architecture.  I expect that higher/larger PEs will have higher scores.  The code plots and summarizes the results of a linear model 

```{r, label="PE score analysis"}
head(PE)
  ylim=c(-70,70)

plotter_score <- function(x,y, xlab,ylab){
  plot(x, y,  ylim=c(-70,70), 
       col=as.numeric(PE$Environment), pch=as.numeric(PE$Environment),
       xlab=xlab, ylab=ylab)
  m1 <- lm(y~x + PE$Environment + PE$Phenotype)
  print(summary(m1))
  abline(lm(y~x))
  abline(0,0, col="grey")
}

png(paste("analysis/Scores_Pine_Spruce_Envi.png"), width=10, height=18, units="in", res=300)
 par(mfrow=c(5,2), mar=c(4,4,4,1), oma=c(0,0,2,0))
  ### All results
  plotter_score(abs(PE$pine_correlation), PE$pine_score_ALL, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE Score ALL")
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score_ALL, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score ALL")

  ### tcontig results
  plotter_score(abs(PE$pine_correlation), PE$pine_score_super_t, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_t")
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_t, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_t")
  ### tcontig results RAW
  plotter_score(abs(PE$pine_correlation), PE$pine_score_super_t_raw, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_t_raw")
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_t_raw, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_t_raw")

  ### gcontig results
  plotter_score(abs(PE$pine_correlation), PE$pine_score_super_g, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_g")
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_g, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_g")
  ### gcontig results RAW
  plotter_score(abs(PE$pine_correlation), PE$pine_score_super_g_raw, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_g_raw")
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_g_raw, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_g_raw")
dev.off()
```
#plot allsnps vs super, expect super to be above 1:1 line

# color by environment is different than by phenotype

plot(PE$pine_score_super_t, PE$spruce_score_super_t,col=as.numeric(PE$Environment), pch=as.numeric(PE$Environment))