---
title: "PEGA for Pine Analysis"
author: "Katie Lotterhos"
date: "August 24, 2015"
output: html_document
---

Does the degree to which SNPs evolve to be correlated with the environment depend on the phenotype-environment association of the phenotypes they are associated with?

Basic algorithm:

1. choose environment
2. choose phenotype
3. get PE association
4. get relevant column data for each SNP (bayenv rho, bayenv bf, gwas slope, gwas p-value) using system calls
5. correct rho for different coding from gwas using flip.cor using merge
6. correct rho for PE/GWAS direction (polarize)
    + (PE-correlation sign) * (GWAS-slope sign) = correction sign
7. (remove large values that are not significant)
8. plot GWAS slope (x axis) vs GEA (y axis)
    + color points in 1st and 3rd quadrants in green (signs match expected direction)
    + color points in 2nd and 4th quadrants in purple (signs don't match expected direction)
    
#### Note
Note that the code runs fastest if you work through R in a terminal window and upload the "results_pine" object first.  Once it is uploaded, when the markdown is created it won't try to load the "results_pine" dataframe, which is very large.  To make this markdown file:
```
library(knitr)
library(markdown)
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis")
knit("AnalysisPEGA.Rmd")  # produces the md file
markdownToHTML("AnalysisPEGA.md", "AnalysisPEGA.html")  # converts an md file to html
```
  
Install packages  
```{r, label="install packages"}
  if (!("hexbin" %in% installed.packages())){install.packages("hexbin")}
  library(hexbin)
  if (!("gridExtra" %in% installed.packages())){install.packages("ash")}
  library(gridExtra)
  if (!("ash" %in% installed.packages())){install.packages("ash")}
  library(ash)
  if (!("fields" %in% installed.packages())){install.packages("fields")}
  library(fields)

install.packages("solaR")
library(solaR)
library(hexbin)
```

Load data, if not already loaded.  The phenotype and environment labels were not consistent among studies.  The `matchXXX` objects are used to match inconsistent labeling.

Also load "flip" data.  A summary of the flip (i.e., allele coding) problem can be found in the `Log.Rmd`.  `XXX_flip_gwas` is an indicator (-1,1) to multiply the gwas slopes by to make them match the ways alleles were coded for bayenv GEAs.  `XXX_flip_raw` is an indicator (-1,1) to multiply Sam's raw correlations by to match the ways alleles were coded for the bayenv GEAs.

```{r, label="load data"}
### load data
PE <- read.csv("../data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2.csv", header=TRUE)

#reorder to make largest effects on top
PE <- PE[rev(order(abs(PE$pine_correlation))),]
head(PE)
plot(PE$pine_correlation, PE$spruce_correlation, pch=20, col="blue")
abline(0,1)
abline(lm(PE$spruce_correlation~PE$pine_correlation), col="blue")

PE_matchGWAS <- read.csv("../data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGWAS.csv", header=TRUE)
PE_matchGEA <- read.csv("../data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGEA.csv", header=TRUE)

#### Load PINE Flip and RESULTS files
  # Ran incorporateFlip.R to get files to right format

#### read in results_pine if it doesn't exist
  if (!("results_pine" %in% ls())){
  results_pine <- read.table("../data/large_files/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV_flipped", header=TRUE, comment.char=" ")
    }
  dim(results_pine)
  ### merge flip with results
  results_pine_head <- names(results_pine)
  results_pine_head

#### Load SPRUCE Flip and RESULTS files

if (!("results_spruce" %in% ls())){
results_spruce <- read.table("../data/large_files/var_out_GATK3_spruce_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV_flipped", header=TRUE, comment.char=" ")
  }
  dim(results_spruce)
 
  results_spruce_head <- names(results_spruce)
  head(results_spruce_head)

```

The following chunk of code is not evaluated:
```{r, eval=FALSE, echo=FALSE}
### Upload superoutliers file pine (NOT USED YET)
  pine_raw_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x_raw.txt", header=TRUE)
  pine_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x.txt", header=TRUE)
  pine_raw_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x_raw.txt", header=TRUE)
  pine_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x.txt", header=TRUE)

  head(pine_tcontig)
  head(pine_gcontig)

### Upload superoutliers file spruce 
  spruce_raw_tcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_100x_raw.txt", header=TRUE)
  spruce_tcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_100x.txt", header=TRUE)
  spruce_raw_gcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_gcontigs_100x_raw.txt", header=TRUE)
  spruce_gcontig <- read.table("data/large_files/super_outliers_per_variable_spruce_both_methods_gcontigs_100x.txt", header=TRUE)

  head(spruce_tcontig)
  head(spruce_gcontig)
```

(Currently not implemented) Get the list of superoutliers.  There are 4 types for each species: based on transcript or gcontig, and based on raw or corrected for population structure.

```{r, label="get list of superoutliers", eval=FALSE, echo=FALSE}
### Get list of superoutliers pine (NOT RUN OR ECHOED)
  pine_super_t <- as.character(pine_tcontig$pine[pine_tcontig$super_outlier_score > 1])
  pine_indexes_super_t <- which(as.character(results_pine$tcontig) %in% pine_super_t)

  pine_super_t_raw <- as.character(pine_raw_tcontig$pine[pine_raw_tcontig$super_outlier_score > 1])
  pine_indexes_super_t_raw <- which(as.character(results_pine$tcontig) %in% pine_super_t_raw)

  pine_super_g <- as.character(pine_gcontig$pine[pine_gcontig$super_outlier_score > 1])
  pine_indexes_super_g <- which(as.character(results_pine$gcontig) %in% pine_super_g)

  pine_super_g_raw <- as.character(pine_raw_gcontig$pine[pine_raw_gcontig$super_outlier_score > 1])
  pine_indexes_super_g_raw <- which(as.character(results_pine$gcontig) %in% pine_super_g_raw)


### Get list of superoutliers spruce
  spruce_super_t <- as.character(spruce_tcontig$spruce[spruce_tcontig$super_outlier_score > 1])
  spruce_indexes_super_t <- which(as.character(results_spruce$tcontig) %in% spruce_super_t)

  spruce_super_t_raw <- as.character(spruce_raw_tcontig$spruce[spruce_raw_tcontig$super_outlier_score > 1])
  spruce_indexes_super_t_raw <- which(as.character(results_spruce$tcontig) %in% spruce_super_t_raw)

  spruce_super_g <- as.character(spruce_gcontig$spruce[spruce_gcontig$super_outlier_score > 1])
  spruce_indexes_super_g <- which(as.character(results_spruce$gcontig) %in% spruce_super_g)

  spruce_super_g_raw <- as.character(spruce_raw_gcontig$spruce[spruce_raw_gcontig$super_outlier_score > 1])
  spruce_indexes_super_g_raw <- which(as.character(results_spruce$gcontig) %in% spruce_super_g_raw)

```

Function used for plotting
```{r}
  plot_compare<- function(x,y, xlab, ylab, xlim=NULL, ylim=NULL){
  if (sum(x,na.rm=TRUE)==0 | sum(y,na.rm=TRUE)==0){plot(0,0);print(head(x)); print(head(y))}else{
  data1 <- cbind(x, y)
  data1b <- data1[complete.cases(data1),]
  if(length(xlim)==0){xlim=max(abs(x)*1.01, na.rm=TRUE)}
  if(length(ylim)==0){ylim=max(abs(y)*1.01, na.rm=TRUE)}  
        binned <- bin2(data1b, 
                 matrix(c(-xlim,xlim,-ylim,ylim), 2,2, byrow=TRUE), 
                 nbin=c(100,100))
    binned$nc[binned$nc==0]=NA
    image(seq(-xlim,xlim,length.out = 100), seq(-ylim,ylim, length.out=100),binned$nc, 
             xlab=xlab, ylab=ylab, add=FALSE, col=tim.colors(75))
  }#end if
  }

### Plotter function
plotter <- function(x, y, main, xlim=NULL, ylim=NULL, xlab, ylab){
  data1 <- cbind(x, y)
  if (sum(x,na.rm=TRUE)==0 | sum(y,na.rm=TRUE)==0){plot(0,0);print(head(x));print(head(y))}else{
  if(length(xlim)==0){xlim=max(abs(x)*1.01, na.rm=TRUE)}
  if(length(ylim)==0){ylim=max(abs(y)*1.01, na.rm=TRUE)} 
  data1b <- data1[complete.cases(data1),]
  score1 <- sum(data1b[,1]*data1b[,2]) #score 1 based on sum of products
  score1b <- round(score1/nrow(data1b)*10000,1)
  score2 <- round(sum((sign(data1b[,1])*sign(data1b[,2]))==1)/nrow(data1b),4) #score 2 based on number of points
    binned <- bin2(data1b, 
                 matrix(c(-xlim,xlim,-ylim,ylim), 2,2, byrow=TRUE), 
                 nbin=c(100,100))
    if(binned$nskip!=0){print("Error in binning, some values not counted. This affects score."); break}
    binned$nc[binned$nc==0]=NA
  
  slope <- lm(y~x)
  
    image.plot(seq(-xlim,xlim,length.out = 100), seq(-ylim,ylim, length.out=100),binned$nc,
             xlab=xlab, ylab=ylab, 
             main=paste(main, "\n, Score1 =",score1b, 
                        "Score2 =", score2,
                        "Slope =",round(slope$coef[2],5)))

  abline(slope)
  return(c(score1b, score2, slope$coef[2]))
  }#end else
  }
```  

Next, loop through each phenotype-environment combo.  For each combo, flip GWAS to match the GEA and polarize the GEA scores so they are in matching direction to PE. (The 'flip' is going to be different for each SNP depending on how it was coded, while the 'polarization' multiplies all SNPs by -1 or by 1 depending on the slope between the phenotype and the environment.)

Then, standardize the GWAS so that each phenotype has a variance of 1 and all slopes will be on the same scale.  Plot the architecture and calculate a score:  

* score1 is `sum(GWAS_rho_bayenv * GEA_rho_bayenv)/n_snps`.  The null hypothesis for score1 is 0 (equal product of snp effects in all quadrants).  Positive scores indicate architectures that have more and/or larger effect SNPs for that phenotype in that environment.

* score2 is the proportion of points in the 1st and 3rd quadrants. Score2 varies between 0 and 1, with the null hypothesis of 0.5 (equal number of SNPs in all quadrants.)  Score2 > 0.5 indicates that the architecture is polarized in the same direction as the environment, while score2 < 0.5 indicates that the architecture is in the opposite direction as expected given the relationship between the phenotype and the environment.


```{r, label="Loop through PE file"}
PE$pine_score1_bayenv <- PE$pine_score2_bayenv <- PE$pine_slope_bayenv <- NA
PE$spruce_score1_bayenv <- PE$spruce_score2_bayenv <- PE$spruce_slope_bayenv <- NA

### Loop through PE file
for (i in 1:nrow(PE)){
### Match names
  PE_envi_name <- as.character(PE$Environment[i])
  GEA_envi_name <- as.character(PE_matchGEA$Environment_GEA[which(PE_matchGEA$Environment_PE==PE_envi_name)])
  
  PE_pheno_name <- as.character(PE$Phenotype[i])
  GWAS_pheno_name <- as.character(PE_matchGWAS$Phenotype_GWAS[which(PE_matchGWAS$Phenotype_PE==PE_pheno_name)])
  
  print(c(i, GEA_envi_name, GWAS_pheno_name))

###################
### Polarize GEA so sign aligns with PE and GWAS sign
###################
  PE_cor_pine <- PE$pine_correlation[i]
  PE_cor_spruce <- PE$spruce_correlation[i]

  polarize_pine <- sign(PE_cor_pine)
  if(polarize_pine==0){polarize_pine=1}
  polarize_spruce <- sign(PE_cor_spruce)
  if(polarize_spruce==0){polarize_spruce=1}

###################
### GWAS effects 
#####################
  ###################
  ### Get GCTA GWAS relevant columns from large file (pine) 
  ###################
    GWAS_effect_col_pine_gcta <- which(results_pine_head==gsub("_p","_snp_effect", GWAS_pheno_name))
    if (length(GWAS_effect_col_pine_gcta)!=1){print("ERROR: GWAS_effect_col_pine_corrected does not equal 1")}
    ### Flip due to allele coding ####
    GWAS_effect_pine_gcta <- results_pine[,GWAS_effect_col_pine_gcta]*results_pine$flip_pine_gwas
  ###################
  ### Get RAW GWAS relevant columns from large file (pine) 
  ###################
    GWAS_pheno_name_raw <- GWAS_pheno_name  
  # next lines fix a problem specific to pine df
    if(GWAS_pheno_name=="root_wt__shoot_wt_p"){
      GWAS_pheno_name_raw <- "root_wt_shoot_wt_p"}
    if(GWAS_pheno_name=="root_wt__shoot_wt_p_1"){
      GWAS_pheno_name_raw <- "root_wt_shoot_wt_p_1"} 
    GWAS_effect_col_pine_raw <- which(results_pine_head==paste(
      GWAS_pheno_name_raw, "_raw_rho", sep=""))
    if (length(GWAS_effect_col_pine_raw)!=1){
      print("ERROR: GWAS_effect_col_pine_raw does not equal 1")}
    ### Flip due to allele coding ####
    GWAS_effect_pine_raw <- results_pine[,GWAS_effect_col_pine_raw]*results_pine$flip_pine_raw
  ###################
  ### Get Bayenv GWAS relevant columns from large file (pine) 
  ###################
  GWAS_effect_cols_pine_bayenv <- grep(gsub("_p","_rho",paste("^",GWAS_pheno_name_raw, sep="")), results_pine_head)
    if (length(GWAS_effect_cols_pine_bayenv)!=2){print("ERROR: GWAS_effect_cols_pine_corrected does not equal 2"); GWAS_rho_pine_bayenv <- rep(NA, nrow(results_pine))}else{
    GWAS_rho_pine_bayenv <- rowMeans(results_pine[,GWAS_effect_cols_pine_bayenv], na.rm=TRUE)
    }
    ### no need to flip
###################
### GEA effects 
#####################
  ###################
  ### Get GEA raw (pine)
  #####################
    GEA_effect_cols_pine_raw <- which(results_pine_head==paste(
      GEA_envi_name,"_raw_rho", sep=""))
    if (length(GEA_effect_cols_pine_raw)!=1){
      print("ERROR: GEA_effect_cols_pine_raw does not equal 1")}
    ### flip and polarize
    GEA_rho_pine_raw <- results_pine[,GEA_effect_cols_pine_raw]*results_pine$flip_pine_raw*polarize_pine
  ###################
  ### Get GEA Bayenv relevant columns from large file (pine) 
  ###################
   GEA_effect_cols_pine_bayenv <- grep(paste("^",GEA_envi_name,"_rho", sep=""), results_pine_head)
    if (length(GEA_effect_cols_pine_bayenv)!=2){print("ERROR: GEA_effect_cols_pine_corrected does not equal 2")}
  # no flip but polarize
    GEA_rho_pine_bayenv <- rowMeans(results_pine[,GEA_effect_cols_pine_bayenv], na.rm=TRUE) * polarize_pine
  ###################
  ### Get GEA LFMM relevant columns from large file (pine)
  ###################
  GEA_effect_col_pine_lfmm <- grep(paste("^",GEA_envi_name,"_z_lfmm_beagle", sep=""), results_pine_head)
    if (length(GEA_effect_col_pine_lfmm)!=1){print("ERROR: GEA_effect_col_pine_lfmm does not equal 1")}
  # no flip but polarize
    GEA_z_pine_lfmm <- results_pine[,GEA_effect_col_pine_lfmm] *polarize_pine
  ###################
  ### Get GEA GCTA relevant columns from large file (pine)
  ###################
  GEA_effect_col_pine_gcta <- grep(paste("^",GEA_envi_name,"_snp_effect_GCTA", sep=""), results_pine_head)
    if (length(GEA_effect_col_pine_gcta)!=1){print("ERROR: GEA_effect_col_pine_gcta does not equal 1")}
    #flip and polarize
    GEA_effect_pine_gcta <- results_pine[,GEA_effect_col_pine_gcta]*results_pine$flip_pine_gwas*polarize_pine

###################
### Plot PINE GEA raw vs corrected
###################
  all_pine <- data.frame(GEA_rho_pine_raw, GEA_rho_pine_bayenv,
                     GEA_effect_pine_gcta, GEA_z_pine_lfmm,
                    GWAS_effect_pine_raw,GWAS_rho_pine_bayenv, 
                    GWAS_effect_pine_gcta)
 
  png(paste("figure_new/",
    PE_pheno_name,"_",i, "_", PE_envi_name,"pine.png",sep=""), width=15, height=15, units="in", res=300)
  par(mfcol=c(ncol(all_pine),ncol(all_pine)), mar=c(1,1,0,0), oma=c(4,4,4,3), cex.main=0.2)
  for (k in 1:ncol(all_pine)){
    for (j in 1:ncol(all_pine)){
        plot_compare(all_pine[,k], all_pine[,j])
        if (k==1){mtext(names(all_pine)[j],side=2,line = 2,cex=0.8)}
        if (j==1){mtext(names(all_pine)[k],side=3,line = 0,cex=0.8)}
    }# end j
  }# end i
  mtext(paste("pine", PE_pheno_name,"_",i, "_", PE_envi_name, PE$pine_correlation[i]), outer=TRUE,
        side=3, line=2)
  dev.off()
###################
### End Plot 
###################

###################
### START SPRUCE
###################
###################
### SPRUCE GWAS
###################
  ###################
  ### Get GWAS GCTA relevant columns from large file (spruce) 
  ###################
    GWAS_effect_col_spruce_gcta <- which(results_spruce_head==gsub("_p","_snp_effect", GWAS_pheno_name))
    if (length(GWAS_effect_col_spruce_gcta)!=1){print("ERROR: GWAS_effect_col_spruce_corrected does not equal 1")}
    # flip gwas
    GWAS_effect_spruce_gcta <- results_spruce[,GWAS_effect_col_spruce_gcta]*results_spruce$flip_spruce_gwas
  ###################
  ### Get GWAS RAW relevant columns from large file (spruce) 
  ###################
   GWAS_effect_col_spruce_raw <- which(results_spruce_head==paste(GWAS_pheno_name, "_raw_rho", sep=""))
  if (length(GWAS_effect_col_spruce_raw)!=1){print("ERROR: GWAS_effect_col_spruce_raw does not equal 1")}
  # flip raw
  GWAS_effect_spruce_raw <- results_spruce[,GWAS_effect_col_spruce_raw]*results_spruce$flip_spruce_raw
  ###################
  ### Get GWAS BAYENV relevant columns from large file (spruce) 
  ###################
  GWAS_effect_cols_spruce_bayenv <- grep(gsub("_p","_rho",paste("^",GWAS_pheno_name_raw, sep="")), results_spruce_head)
    if (length(GWAS_effect_cols_spruce_bayenv)!=2){print("ERROR: GWAS_effect_cols_spruce_corrected does not equal 2"); GWAS_rho_spruce_bayenv <- rep(NA, nrow(results_spruce))}else{
    GWAS_rho_spruce_bayenv <- rowMeans(results_spruce[,GWAS_effect_cols_spruce_bayenv], na.rm=TRUE)
    }
    ### no need to flip
###################
### SPRUCE GEA
###################
  ###################
  ### Get GEA Bayenv relevant columns from large file (spruce) 
  ###################
   GEA_effect_cols_spruce_bayenv <- grep(paste("^",GEA_envi_name,"_rho", sep=""), results_spruce_head)
    if (length(GEA_effect_cols_spruce_bayenv)!=2){print("ERROR: GEA_effect_cols_spruce_corrected does not equal 2")}
    GEA_rho_spruce_bayenv <- rowMeans(results_spruce[,GEA_effect_cols_spruce_bayenv], na.rm=TRUE)*polarize_spruce
  ###################
  ### Get LFMM relevant columns from large file (spruce)
  ###################
  GEA_effect_col_spruce_lfmm <- grep(paste("^",GEA_envi_name,".z.lfmm.beagle", sep=""), results_spruce_head)
    if (length(GEA_effect_col_spruce_lfmm)!=1){print("ERROR: GEA_effect_col_spruce_lfmm does not equal 1")}
    # no need to flip but need to polarize
    GEA_z_spruce_lfmm <- results_spruce[,GEA_effect_col_spruce_lfmm]*polarize_spruce
  ###################
  ### Get RAW GEA relevant columns from large file (spruce)
  ###################
    GEA_effect_cols_spruce_raw <- which(results_spruce_head==paste(GEA_envi_name,"_raw_rho", sep=""))
    if (length(GEA_effect_cols_spruce_raw)!=1){print("ERROR: GEA_effect_cols_spruce_raw does not equal 1")}
    # flip and polarize
    GEA_rho_spruce_raw <- results_spruce[,GEA_effect_cols_spruce_raw]*results_spruce$flip_spruce_raw*polarize_spruce
  ###################
  ### Get GEA GCTA relevant columns from large file (spruce)
  ###################
  GEA_effect_col_spruce_gcta <- grep(paste("^",GEA_envi_name,"_snp_effect_GCTA", sep=""), results_spruce_head)
    if (length(GEA_effect_col_spruce_gcta)!=1){print("ERROR: GEA_effect_col_spruce_gcta does not equal 1")}
    #flip and polarize
    GEA_effect_spruce_gcta <- results_spruce[,GEA_effect_col_spruce_gcta]*results_spruce$flip_spruce_gwas*polarize_spruce

###################
### Plot SPRUCE GEA raw vs corrected
###################
  all_spruce <- data.frame(GEA_rho_spruce_raw, GEA_rho_spruce_bayenv,
                     GEA_effect_spruce_gcta, GEA_z_spruce_lfmm,
                    GWAS_effect_spruce_raw,GWAS_rho_spruce_bayenv, 
                    GWAS_effect_spruce_gcta)
 
  png(paste("figure_new/",
    PE_pheno_name,"_",i, "_", PE_envi_name,"spruce.png",sep=""), width=15, height=15, units="in", res=300)
  par(mfcol=c(ncol(all_spruce),ncol(all_spruce)), mar=c(1,1,0,0), oma=c(4,4,4,3), cex.main=0.2)
  for (k in 1:ncol(all_spruce)){
    for (j in 1:ncol(all_spruce)){
        plot_compare(all_spruce[,k], all_spruce[,j])
        if (k==1){mtext(names(all_spruce)[j],side=2,line = 2,cex=0.8)}
        if (j==1){mtext(names(all_spruce)[k],side=3,line = 0,cex=0.8)}
    }# end j
  }# end i
  mtext(paste("spruce", PE_pheno_name,"_",i, "_", PE_envi_name, PE$spruce_correlation[i]), outer=TRUE,
        side=3, line=2)
  dev.off()
###################
### End Plot 
###################

###############
### Calc PEGA based on Bayenv GWAS and GEA
###############
if(sum(!is.na(GWAS_rho_pine_bayenv))>0 | sum(!is.na(GWAS_rho_spruce_bayenv))>0){
  PE$pine_score1_bayenv[i] <- sum(GWAS_rho_pine_bayenv*GEA_rho_pine_bayenv)/length(GWAS_rho_pine_bayenv) 
  PE$pine_score2_bayenv[i] <- round(sum(sign(GWAS_rho_pine_bayenv)*sign(GEA_rho_pine_bayenv)/length(GWAS_rho_pine_bayenv)) ,4) 
  PE$pine_slope_bayenv[i] <- lm(GEA_rho_pine_bayenv~GWAS_rho_pine_bayenv)$coef[2]
  
  PE$spruce_score1_bayenv[i] <- sum(GWAS_rho_spruce_bayenv*GEA_rho_spruce_bayenv)/length(GWAS_rho_spruce_bayenv)
  PE$spruce_score2_bayenv[i] <- round(sum(sign(GWAS_rho_spruce_bayenv)*sign(GEA_rho_spruce_bayenv)/length(GWAS_rho_spruce_bayenv)) ,4) 
  PE$spruce_slope_bayenv[i] <- lm(GEA_rho_spruce_bayenv~GWAS_rho_spruce_bayenv)$coef[2]
  }#end if
} #end loop

```

Now that all the scores have been calculated, we can look at the relationship between the PE association and the underlying genome-wide pattern.  I expect that higher/larger PEs will have higher scores.  The code plots and summarizes the results of a linear model 

```{r, label="PE score analysis"}
head(PE)
  ylim=c(-70,70)

plotter_score <- function(x,y, xlab,ylab, ymin, ymax){
  plot(x, y,  ylim=c(ymin,ymax), 
       col=as.numeric(PE$Phenotype), pch=as.numeric(PE$Phenotype),
       xlab=xlab, ylab=ylab)
  m1 <- lm(y~x + PE$Environment + PE$Phenotype)
  print(summary(m1))
  abline(lm(y~x))
  abline(0,0, col="grey", lwd=2)
  abline(0.5,0, col="grey", lwd=2)
}

plotter_score_envi <- function(x,y, xlab,ylab, ymin, ymax){
  plot(x, y,  ylim=c(ymin,ymax), 
       col=as.numeric(PE$Environment), pch=as.numeric(PE$Environment),
       xlab=xlab, ylab=ylab)
}

png(paste("../analysis/Score1_Pine_Spruce_Envi.png"), width=10, height=10, units="in", res=300)
 par(mfrow=c(2,2), mar=c(4,4,4,1), oma=c(0,0,2,0))
  ### All results
  plotter_score(abs(PE$pine_correlation), PE$pine_score1_ALL_corrected_bayenv, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (corrected) Score 1 ALL SNPs", -40,40)
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score1_ALL_corrected_bayenv, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (corrected) Score 1 ALL SNPs", -40,40)
  
  plotter_score(abs(PE$pine_correlation), PE$pine_score1_ALL_raw, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (raw) Score 1 ALL SNPs", -30,30)
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score1_ALL_raw, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (raw) Score 1 ALL SNPs", -550, 550)
dev.off()

png(paste("../analysis/Score2_Pine_Spruce_Envi.png"), width=10, height=10, units="in", res=300)
 par(mfrow=c(2,2), mar=c(4,4,4,1), oma=c(0,0,2,0))
  plotter_score(abs(PE$pine_correlation), PE$pine_score2_ALL_corrected_bayenv, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (corrected) Proportion All SNPs", .4,.6)
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score2_ALL_corrected_bayenv, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (corrected) Proportion All SNPs", .4,.6)
  plotter_score(abs(PE$pine_correlation), PE$pine_score2_ALL_raw, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (raw) Proportion All SNPs", 0.1, 1)
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score2_ALL_raw, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (raw) Proportion All SNPs", 0.1, 1)
dev.off()

#### Highlight points by environment
png(paste("../analysis/Score1_Pine_Spruce_Envi_colorByEnvi.png"), width=10, height=10, units="in", res=300)
 par(mfrow=c(2,2), mar=c(4,4,4,1), oma=c(0,0,2,0))
  ### All results
  plotter_score_envi(abs(PE$pine_correlation), PE$pine_score1_ALL_corrected_bayenv, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (corrected) Score 1 ALL SNPs", -40,40)
  plotter_score_envi(abs(PE$spruce_correlation), PE$spruce_score1_ALL_corrected_bayenv, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (corrected) Score 1 ALL SNPs", -40,40)
  
  plotter_score_envi(abs(PE$pine_correlation), PE$pine_score1_ALL_raw, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (raw) Score 1 ALL SNPs", -30,30)
  plotter_score_envi(abs(PE$spruce_correlation), PE$spruce_score1_ALL_raw, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (raw) Score 1 ALL SNPs", -550, 550)
dev.off()

png(paste("../analysis/Score2_Pine_Spruce_Envi_colorByEnvi.png"), width=10, height=10, units="in", res=300)
 par(mfrow=c(2,2), mar=c(4,4,4,1), oma=c(0,0,2,0))
  plotter_score_envi(abs(PE$pine_correlation), PE$pine_score2_ALL_corrected_bayenv, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (corrected) Proportion All SNPs", .4,.6)
  plotter_score_envi(abs(PE$spruce_correlation), PE$spruce_score2_ALL_corrected_bayenv, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (corrected) Proportion All SNPs", .4,.6)
  plotter_score_envi(abs(PE$pine_correlation), PE$pine_score2_ALL_raw, 
                "PINE Abs(Pheno-Envi) Correlation", "PINE (raw) Proportion All SNPs", 0.1, 1)
  plotter_score_envi(abs(PE$spruce_correlation), PE$spruce_score2_ALL_raw, 
                "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE (raw) Proportion All SNPs", 0.1, 1)
dev.off()

quartz()
plot(0,0)
phens <- levels(PE$Phenotype)
legend(-1,1, col=1:length(phens), pch=1:length(phens),
       legend=phens)


#   ### tcontig results
#   plotter_score(abs(PE$pine_correlation), PE$pine_score_super_t, 
#                 "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_t")
#   plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_t, 
#                 "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_t")
#   ### tcontig results RAW
#   plotter_score(abs(PE$pine_correlation), PE$pine_score_super_t_raw, 
#                 "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_t_raw")
#   plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_t_raw, 
#                 "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_t_raw")
# 
#   ### gcontig results
#   plotter_score(abs(PE$pine_correlation), PE$pine_score_super_g, 
#                 "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_g")
#   plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_g, 
#                 "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_g")
#   ### gcontig results RAW
#   plotter_score(abs(PE$pine_correlation), PE$pine_score_super_g_raw, 
#                 "PINE Abs(Pheno-Envi) Correlation", "PINE Score super_g_raw")
#   plotter_score(abs(PE$spruce_correlation), PE$spruce_score_super_g_raw, 
#                 "SPRUCE Abs(Pheno-Envi) Correlation", "SPRUCE Score super_g_raw")
```

```
#plot allsnps vs super, expect super to be above 1:1 line

# color by environment is different than by phenotype

plot(PE$pine_score_super_t, PE$spruce_score_super_t,col=as.numeric(PE$Environment), pch=as.numeric(PE$Environment))
```