---
title: "PEGA for Pine Analysis"
author: "Katie Lotterhos"
date: "August 24, 2015"
output: html_document
---

Basic algorithm:

1. choose environment
2. choose phenotype
3. get PE association
4. get relevant column data for each SNP (bayenv rho, bayenv bf, gwas slope, gwas p-value) using system calls
5. correct rho for different coding from gwas using flip.cor using merge
6. correct rho for PE/GWAS direction (polarize)
    + (PE-correlation sign) * (GWAS-slope sign) = correction sign
7. (remove large values that are not significant)
8. plot GWAS slope (x axis) vs GEA (y axis)
    + color points in 1st and 3rd quadrants in green (signs match expected direction)
    + color points in 2nd and 4th quadrants in purple (signs don't match expected direction)
    
#### Note
Note that the code runs fastest if you work through R in a terminal window and upload the "results_pine" object first.  Once it is uploaded, when the markdown is created it won't try to load the "results_pine" dataframe, which is very large.  To make this markdown file:
```
library(knitr)
library(markdown)
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis")
knit("AnalysisPEGA.Rmd")  # produces the md file
markdownToHTML("AnalysisPEGA.md", "AnalysisPEGA.html")  # converts an md file to html
```
  
Install packages  
```{r, label="install packages"}
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA")
  if (!("hexbin" %in% installed.packages())){install.packages("hexbin")}
  library(hexbin)
  if (!("gridExtra" %in% installed.packages())){install.packages("ash")}
  library(gridExtra)
  if (!("ash" %in% installed.packages())){install.packages("ash")}
  library(ash)
```

Load data, if not already loaded.  The phenotype and environment labels were not consistent among studies.  The `matchXXX` objects are used to match inconsistent labeling.

```{r, label="load data"}
### load data
PE <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2.csv", header=TRUE)
#reorder to make largest effects on top
PE <- PE[rev(order(abs(PE$pine_correlation))),]
head(PE)

PE_matchGWAS <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGWAS.csv", header=TRUE)
PE_matchGEA <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGEA.csv", header=TRUE)

flip_pine <- read.table("data/large_files/flip_pine.cor", header=TRUE)
head(flip_pine)

#### read in results_pine if it doesn't exist
if (!("results_pine" %in% ls())){
results_pine <- read.table("data/large_files/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS", header=TRUE, comment.char=" ")
  }
dim(results_pine)

### Upload superoutliers file
  pine_raw_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x_raw.txt", header=TRUE)
  pine_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x.txt", header=TRUE)
  pine_raw_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x_raw.txt", header=TRUE)
  pine_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x.txt", header=TRUE)

  head(pine_tcontig)
  head(pine_gcontig)
```

Get the list of superoutliers.  There are 4 types for each species: based on transcript or gcontig, and based on raw or corrected for population structure.

```{r, label="get list of superoutliers"}
### Get list of superoutliers
  pine_super_t <- as.character(pine_tcontig$pine[pine_tcontig$super_outlier_score > 1])
  pine_indexes_super_t <- which(as.character(results_pine$tcontig) %in% pine_super_t)

  pine_super_t_raw <- as.character(pine_raw_tcontig$pine[pine_raw_tcontig$super_outlier_score > 1])
  pine_indexes_super_t_raw <- which(as.character(results_pine$tcontig) %in% pine_super_t_raw)

  pine_super_g <- as.character(pine_gcontig$pine[pine_gcontig$super_outlier_score > 1])
  pine_indexes_super_g <- which(as.character(results_pine$gcontig) %in% pine_super_g)

  pine_super_g_raw <- as.character(pine_raw_gcontig$pine[pine_raw_gcontig$super_outlier_score > 1])
  pine_indexes_super_g_raw <- which(as.character(results_pine$gcontig) %in% pine_super_g_raw)

  PE$pine_score_ALL <- PE$pine_score_super_t <- PE$pine_score_super_t_raw <- PE$pine_score_super_g <- PE$pine_score_super_g_raw <- NA
```

Next, loop through each phenotype-environment combo.  For each combo, flip and polarize the GEA scores so they are in matching direction to the GWAS and PE.  Then, standardize the GWAS so that each phenotype has a variance of 1 and all slopes will be on the same scale.  Plot the architecture and calculate a score.  The score is based on the `sum(GWAS_effect_standardized * GEA_rho_polarized)/n_snps`.  So larger and more positive scores indicate architectures that have more and/or larger effect SNPs for that phenotype in that environment.

```{r, label="Loop through PE file"}
### Loop through PE file

for (i in 1:nrow(PE)){
### Match names
  PE_envi_name <- as.character(PE$Environment[i])
  GEA_envi_name <- as.character(PE_matchGEA$Environment_GEA[which(PE_matchGEA$Environment_PE==PE_envi_name)])
  
  PE_pheno_name <- as.character(PE$Phenotype[i])
  GWAS_pheno_name <- as.character(PE_matchGWAS$Phenotype_GWAS[which(PE_matchGWAS$Phenotype_PE==PE_pheno_name)])
  
  print(c(i, GEA_envi_name, GWAS_pheno_name))

### Get relevant columns from large file
  results_pine_head <- names(results_pine)
  GWAS_effect_col <- which(results_pine_head==gsub("_p","_snp_effect", GWAS_pheno_name))
  if (length(GWAS_effect_col)!=1){print("ERROR: GWAS_effect_col does not equal 1")}
  GWAS_effect <- results_pine[,GWAS_effect_col]

  GEA_effect_cols <- grep(paste("^",GEA_envi_name,"_rho", sep=""), results_pine_head)
  if (length(GEA_effect_cols)!=2){print("ERROR: GEA_effect_cols does not equal 2")}
  GEA_rho <- rowMeans(results_pine[,GEA_effect_cols], na.rm=TRUE)

### check that "flip_pine"" order matches "results_pine" order
if (!(identical(flip_pine$gcontig__gcontig_pos, results_pine$gcontig__gcontig_pos))){
  print("ERROR!  SNP order in flip_pine file does not match results_pine file")
}

### flip_pine GEA_rho so sign matches GWAS_effect
  GEA_rho_flip_pine <- GEA_rho * flip_pine$X1

### Polarize GEA so sign aligns with PE and GWAS sign
  PE_cor <- PE$pine_correlation[i]

  GWAS_sign <- sign(GWAS_effect)
  GWAS_sign[GWAS_sign==0]=1

  if (sum(GWAS_sign==0, na.rm=TRUE)>0){print("ERROR: Some GWAS have 0 sign for polarization, will skew results_pine")} #fixed this with code below

  if(PE_cor != 0){
  GEA_rho_polarized <- GEA_rho_flip_pine * sign(PE_cor) * GWAS_sign
  }else{
  GEA_rho_polarized <- GEA_rho_flip_pine * GWAS_sign  
  }

#  plot(0,0, xlim=c(-3,3), ylim=c(-0.5,0.5), col=0, main=paste(PE_envi_name, PE_pheno_name, PE$pine_correlation[i]))
  #plot(hexbin(GWAS_effect, GEA_rho_polarized,xbnds=c(-5,5), ybnds=c(-1,1), xbins=100), add=TRUE)


### Plotting 


GWAS_effect_standardized <- GWAS_effect/sd(GWAS_effect, na.rm=TRUE)

plotter <- function(GWAS_effect, GEA_rho_polarized, main, is.standard){
  data1 <- cbind(GWAS_effect, GEA_rho_polarized)
  data1b <- data1[complete.cases(data1),]
  score1 <- sum(data1b[,1]*data1b[,2])
  score <- score1/nrow(data1b)*10000
  if(is.standard==FALSE){
    binned <- bin2(data1b, 
                 matrix(c(-3,3,-0.5,0.5), 2,2, byrow=TRUE), 
                 nbin=c(100,100))
    binned$nc[binned$nc==0]=NA
    image.plot(seq(-3,3,length.out = 100), seq(-0.5, 0.5, length.out=100),binned$nc,
             xlab="GWAS effect", ylab="GEA rho (polarized)", main=main)
    #text(2.9,0.45, round(score,1))
  }
  if(is.standard==TRUE){
    binned <- bin2(data1b, 
                 matrix(c(-10,10,-0.5,0.5), 2,2, byrow=TRUE), 
                 nbin=c(100,100))
    binned$nc[binned$nc==0]=NA
    image.plot(seq(-10,10,length.out = 100), seq(-0.5, 0.5, length.out=100),binned$nc,
             xlab="GWAS effect standardized", ylab="GEA rho (polarized)", main=paste(main, ", Score = ",round(score,1)))
  }
  return(score)
  }

  png(paste("analysis/figure_byPheno/PEGA_" ,PE_pheno_name,"_",i, "_", PE_envi_name,".png",sep=""), width=5, height=18, units="in", res=300)
  par(mfrow=c(6,1), mar=c(4,4,4,1), oma=c(0,0,2,0))
  plotter(GWAS_effect, GEA_rho_polarized, main="All SNPs", is.standard=FALSE)
  PE$pine_score_ALL[i]<-plotter(GWAS_effect_standardized, GEA_rho_polarized, main="All SNPs Standardized", is.standard=TRUE)
  PE$pine_score_super_t[i]<-plotter(GWAS_effect_standardized[pine_indexes_super_t], GEA_rho_polarized[pine_indexes_super_t], main="Super tcontigs", is.standard=TRUE)
  PE$pine_score_super_t_raw[i]<-plotter(GWAS_effect_standardized[pine_indexes_super_t_raw], GEA_rho_polarized[pine_indexes_super_t_raw], main="Super tcontigs RAW", is.standard=TRUE)
  PE$pine_score_super_g[i]<-plotter(GWAS_effect_standardized[pine_indexes_super_g], GEA_rho_polarized[pine_indexes_super_g], main="Super gcontigs", is.standard=TRUE)
  PE$pine_score_super_g_raw[i]<-plotter(GWAS_effect_standardized[pine_indexes_super_g_raw], GEA_rho_polarized[pine_indexes_super_g_raw], main="Super gcontigs RAW", is.standard=TRUE)
  title("yo")
  mtext(paste(PE_pheno_name, PE_envi_name, "\nPine", PE$pine_correlation[i], sep=" "),
        side=3, outer=TRUE, line=-1)
  dev.off()

} #end loop
```

Now that all the scores have been calculated, we can look at the relationship between the PE association and the underlying architecture.  I expect that higher/larger PEs will have higher scores.  The code plots and summarizes the results of a linear model 

```{r, label="PE score analysis"}
head(PE)

png(paste("analysis/Scores.png"), width=5, height=18, units="in", res=300)
 par(mfrow=c(5,1), mar=c(4,4,4,1), oma=c(0,0,2,0))
  plot(abs(PE$pine_correlation), PE$pine_score_ALL,  ylim=c(-50,50), 
       col=as.numeric(PE$Phenotype), pch=20)
  m1 <- lm(PE$pine_score_ALL~abs(PE$pine_correlation) + PE$Environment + PE$Phenotype)
  summary(m1)
  abline(lm(PE$pine_score_ALL~abs(PE$pine_correlation)))

  plot(abs(PE$pine_correlation), PE$pine_score_super_t,  ylim=c(-50,50), 
       col=as.numeric(PE$Phenotype), pch=20)
  m2 <- lm(PE$pine_score_super_t~abs(PE$pine_correlation) + PE$Environment + PE$Phenotype)
  summary(m2)
  abline(lm(PE$pine_score_super_t~abs(PE$pine_correlation)))

  plot(abs(PE$pine_correlation), PE$pine_score_super_t_raw,  ylim=c(-50,50), 
       col=as.numeric(PE$Phenotype), pch=20)
  m3 <- lm(PE$pine_score_super_t_raw~abs(PE$pine_correlation) + PE$Environment + PE$Phenotype)
  summary(m3)
  abline(lm(PE$pine_score_super_t_raw~abs(PE$pine_correlation)))

  plot(abs(PE$pine_correlation), PE$pine_score_super_g,  ylim=c(-50,50), 
       col=as.numeric(PE$Phenotype), pch=20)
  m4<- lm(PE$pine_score_super_g~abs(PE$pine_correlation) + PE$Environment + PE$Phenotype)
  summary(m4)
  abline(lm(PE$pine_score_super_g~abs(PE$pine_correlation)))

  plot(abs(PE$pine_correlation), PE$pine_score_super_g_raw,  ylim=c(-50,50), 
       col=as.numeric(PE$Phenotype), pch=20)
  m5 <- lm(PE$pine_score_super_g_raw~abs(PE$pine_correlation) + PE$Environment + PE$Phenotype)
  summary(m5)
  abline(lm(PE$pine_score_super_g_raw~abs(PE$pine_correlation)))
dev.off()
```