---
title: "PEGA for Pine Analysis: Step 1 flip SNPs to alphabetical"
author: "Katie Lotterhos"
date: "April 19, 2016"
output: html_document
---

This script "flips" all the SNP effects so they are alphabetical (SNP with first letter in the alphabet is coded 0? or 1?).
    
#### Note
Note that the code runs fastest if you work through R in a terminal window and upload the "results_pine" object first.  Once it is uploaded, when the markdown is created it won't try to load the "results_pine" dataframe, which is very large.  To make this markdown file:
```
library(knitr)
library(markdown)
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis")
knit("AnalysisPEGA.Rmd")  # produces the md file
markdownToHTML("AnalysisPEGA.md", "AnalysisPEGA.html")  # converts an md file to html
```
  
Install packages  
```{r, label="install packages"}
  setwd("/Users/katie/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis")
  if (!("hexbin" %in% installed.packages())){install.packages("hexbin")}
  library(hexbin)
  if (!("gridExtra" %in% installed.packages())){install.packages("ash")}
  library(gridExtra)
  if (!("ash" %in% installed.packages())){install.packages("ash")}
  library(ash)
  if (!("fields" %in% installed.packages())){install.packages("fields")}
  library(fields)
  if (!("solaR" %in% installed.packages())){install.packages("solaR")}
  library(solaR)

plot_2D<- function(x,y, xlab, ylab, xlim=NULL, ylim=NULL, nbin, x_sub, y_sub){

  #if(!("ash" %in% installed.packages())){install.packages("ash")}
  require(ash)
  require(fields)

  data1 <- cbind(x, y)
  data1b <- data1[complete.cases(data1),]
  if(length(xlim)==0){
    xlim_up <- max(x, na.rm=TRUE)
    xlim_lower <- min(x, na.rm=TRUE)
  }
  if(length(ylim)==0){
    ylim_up <- max(y, na.rm=TRUE)
    ylim_lower <- min(y, na.rm=TRUE)
  }
        binned <- bin2(data1b,
                 matrix(c(xlim_lower,xlim_up,ylim_lower,ylim_up), 2,2, byrow=TRUE),
                 nbin=c(nbin,nbin))
    binned$nc[binned$nc==0]=NA
    image.plot(seq(xlim_lower,xlim_up,length.out = nbin), seq(ylim_lower,ylim_up, length.out=nbin),
               binned$nc,
             xlab=xlab, ylab=ylab, add=FALSE, col=grey.colors(75, 0.25,0.9))
    points(x_sub, y_sub, pch=24, cex=1.5, col="dodgerblue", bg="darkorange")
} # end .plot_2D



```

Load data, if not already loaded.  The phenotype and environment labels were not consistent among studies.  The `matchXXX` objects are used to match inconsistent labeling.

Also load "flip" data.  A summary of the flip (i.e., allele coding) problem can be found in the `Log.Rmd`.  `XXX_flip_gwas` is an indicator (-1,1) to multiply the gwas slopes by to make them match the ways alleles were coded for bayenv GEAs (alphabetical).  `XXX_flip_raw` is an indicator (-1,1) to multiply Sam's raw correlations by to match the ways alleles were coded for the bayenv GEAs (alphabetical).

```{r, label="load data"}
### load data
PE <- read.csv("../data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2.csv", header=TRUE)

#reorder to make largest effects on top
PE <- PE[rev(order(abs(PE$pine_correlation))),]
head(PE)
#plot(PE$pine_correlation, PE$spruce_correlation, pch=20, col="blue")
#abline(0,1)
#abline(lm(PE$spruce_correlation~PE$pine_correlation), col="blue")

PE_matchGWAS <- read.csv("../data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGWAS.csv", header=TRUE)
PE_matchGEA <- read.csv("../data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGEA.csv", header=TRUE)

head(PE_matchGWAS)
head(PE_matchGEA)
#### Load PINE Flip and RESULTS files
  # Ran incorporateFlip.R to get files to right format

#### read in results_pine if it doesn't exist
  if (!("results_pine" %in% ls())){
  results_pine <- read.table("../data/large_files/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS_altGCTABAYENV", header=TRUE, comment.char=" ")
    }
  dim(results_pine)
  ### merge flip with results
  results_pine_head <- names(results_pine)
  results_pine_head

#### Load PINE Flip and RESULTS files
flip_pine_gwas <- read.table("../data/large_files/flip_pine_v2.cor", header=TRUE)
  nrow(flip_pine_gwas)
  names(flip_pine_gwas)[2] <- "flip_pine_gwas"
  head(flip_pine_gwas)

flip_pine_raw <- read.table("../data/large_files/var_out_GATK3_allhet_pine688_ALL.table_filt10_p95_het7_passFILT2.2_LDformat_TOFLIP_alphabet_lowest_eq_1")
  nrow(flip_pine_raw)
  names(flip_pine_raw)[1] <- names(flip_pine_gwas)[1]
  names(flip_pine_raw)[2] <- "flip_pine_raw"
  head(flip_pine_raw)

flip_pine2 <- merge(flip_pine_gwas, flip_pine_raw, all.x=TRUE)
  nrow(flip_pine2)
  head(flip_pine2)
rm(flip_pine_gwas, flip_pine_raw)

 ### merge flip with results
  results_pine2 <- merge(flip_pine2, results_pine)
  dim(results_pine2)
  head(results_pine2[,1:10])
  results_pine <- results_pine2
  rm(results_pine2, flip_pine2)

  results_pine_head <- names(results_pine)
  results_pine_head
```


The following chunk of code further incorporates Sam's super outlier criteria into the dataframe.

```{r,echo=FALSE}
if ("pine_super_raw" %in% names(results_pine)){print("Check if you want to run this code"); break}
### Upload superoutliers 
 in_super_pine <- read.table ("../data/large_files/super_outliers_per_variable_pine_both_methods_100x.txt",T)

in_super_pine_raw <- read.table ("../data/large_files/super_outliers_per_variable_pine_both_methods_100x_raw.txt",T)


good_super_pine <- in_super_pine [(in_super_pine$outlier_count > in_super_pine$p9),]

good_super_pine_raw <- in_super_pine_raw [(in_super_pine_raw$outlier_count > in_super_pine_raw$p9),]

length(levels(factor(good_super_pine$pine)))
length(levels(factor(good_super_pine_raw$pine)))

head(good_super_pine)
head(good_super_pine_raw)

results_pine$pine_super_raw_p9 <-(results_pine$tcontig %in% as.character(good_super_pine_raw$pine))

results_pine$pine_super_p9 <-(results_pine$tcontig %in% as.character(good_super_pine$pine))

### One more less awesome level:
good_super_pine_p4 <- in_super_pine [(in_super_pine$outlier_count > in_super_pine$p4),]
good_super_pine_raw_p4 <- in_super_pine_raw [(in_super_pine_raw$outlier_count > in_super_pine_raw$p4),]

results_pine3$pine_super_raw_p4 <-(results_pine$tcontig %in% as.character(good_super_pine_raw_p4$pine))
results_pine3$pine_super_p4 <-(results_pine$tcontig %in% as.character(good_super_pine_p4$pine))

### Super SNPs
    ## number of SNPs
    sum(results_pine$pine_super_p9)
    
    ## number of gcontigs
    length(levels(factor(results_pine$gcontig[results_pine$pine_super_p9])))
           
    ## number of tcontigs
    length(levels(factor(results_pine$tcontig[results_pine$pine_super_p9])))

### Super RAW SNPs
    ## number of SNPs
    sum(results_pine$pine_super_raw_p9)
    
    ## number of gcontigs
    length(levels(factor(results_pine$gcontig[results_pine$pine_super_raw_p9])))
           
    ## number of tcontigs
    length(levels(factor(results_pine$tcontig[results_pine$pine_super_raw_p9])))


### Super AND Super RAW SNPs
    ## number of SNPs
    sum(results_pine$pine_super_raw_p9 & results_pine$pine_super_p9)
    
    ## number of gcontigs
    length(levels(factor(results_pine$gcontig[results_pine$pine_super_raw_p9& results_pine$pine_super_p9])))
           
    ## number of tcontigs
    length(levels(factor(results_pine$tcontig[results_pine$pine_super_raw_p9& results_pine$pine_super_p9])))

```

### Input convergent tcontigs and merge

```{r}
conv <- read.table("../data/pine_spruce_top25_convergence_q05.txt", header=TRUE)

results_pine$pine_super_convergent <- results_pine$tcontig %in% conv$pine_tcontig
### Check should == 25
length(unique(results_pine$tcontig[results_pine$pine_super_convergent]))
### number of SNPs
sum(results_pine$pine_super_convergent)
```

### Average together the 2 Bayenv rho estimates

```{r}
cols_rho1 <- c(grep("rho1", names(results_pine)))
cols_rho2 <- c(grep("rho2", names(results_pine)))
a<- cbind(names(results_pine)[cols_rho1],
      names(results_pine)[cols_rho2])

newname <- gsub("rho1", "rhoave", names(results_pine)[cols_rho1])

out <- matrix(NA, nrow(results_pine), length(newname))

for (a in 1:length(cols_rho1)){
  out[,a] <- (results_pine[,cols_rho1[a]]+ 
                  results_pine[,cols_rho2[a]])/2
}

outdf <- data.frame(out)
names(outdf) <- newname
head(outdf)

### Check
cor.test(outdf$Fall_cold_injury_rhoave_BAYENV, results_pine$Fall_cold_injury_rho1_BAYENV)

results_pine2 <- results_pine[,-c(cols_rho1, cols_rho2)]
results_pine2 <- data.frame(results_pine2, outdf)
head(results_pine2)
```



Next, flip all SNPs so they are in alphabetical order. 

```{r}
results_pine3 <- results_pine2
# grep "gcta" columns, multiply by flip
names(results_pine3)
cols_gcta <- grep("snp_effect", names(results_pine3))
names(results_pine3)[cols_gcta]
results_pine3[,cols_gcta] <- results_pine3[,cols_gcta]*results_pine3$flip_pine_gwas
head(results_pine3[,cols_gcta])

# grep "raw" columns, multiply by flip
cols_raw <- c(grep("raw", names(results_pine3)))
names(results_pine3[cols_raw])
cols_raw<-cols_raw[-c(1,length(cols_raw))] # keep flip_pine_raw and super_raw_p9
results_pine3[,cols_raw] <- results_pine3[,cols_raw]*results_pine3$flip_pine_raw
head(results_pine3[,cols_raw], 1)

## make a plot to check everything looks OK
names(results_pine3)[grep("LAT", names(results_pine3))]
names(results_pine3)[grep("Fall", names(results_pine3))]

# iconic relationship

super<- (results_pine3$Fall_cold_injury_rhoave_BAYENV*results_pine3$LAT_rhoave)< - 0.015
# sum(super)
# length(levels(factor(results_pine3$gcontig[super])))
# length(levels(factor(results_pine3$tcontig[super])))
# length(levels(factor(results_pine3$gcontig[super & (results_pine3$pine_super_raw_p9 | results_pine3$pine_super_p9)])))
super_FCILat <- data.frame(results_pine3$gcontig[super],
                           results_pine3$tcontig[super],
                           results_pine3$pine_super_p9[super],
                           results_pine3$pine_super_raw_p9[super], stat=(results_pine3$Fall_cold_injury_rhoave_BAYENV*results_pine3$LAT_rhoave)[super]*-1)

### Random sample of 10000
plot(sample(results_pine3$Fall_cold_injury_rhoave_BAYENV,10000),
     sample(results_pine3$LAT_rhoave, 10000),
     pch=19, col=rgb(0,0,0,0.5))



plot_2D(x=results_pine3$Fall_cold_injury_rhoave_BAYENV[samp],
       y=results_pine3$LAT_rhoave[samp], 
       xlab= "FCI rho Bayenv", 
       ylab= "LAT rho Bayenv" , 
       nbin=100, 
       x_sub=results_pine3$Fall_cold_injury_rhoave_BAYENV[super], 
       y_sub=results_pine3$LAT_rhoave[super])


plot_2D(x=results_pine3$Fall_cold_injury_rhoave_BAYENV,
       y=results_pine3$LAT_rhoave, 
       xlab= "FCI rho Bayenv", 
       ylab= "LAT rho Bayenv" , 
       nbin=100, 
       x_sub=results_pine3$Fall_cold_injury_rhoave_BAYENV[super], 
       y_sub=results_pine3$LAT_rhoave[super])


# Bayenv LAT Manhattan
plot_2D(x=1:length(results_pine3$LAT_rhoave),
        y=results_pine3$LAT_rhoave,
        xlab="Bp",
        ylab="LAT rho Bayenv",
        nbin=100,
        x_sub=(1:length(results_pine3$LAT_rhoave))[super], 
        y_sub=results_pine3$LAT_rhoave[super])

# before flip - should be no relationship
plot_2D(x=results_pine2$Fall_cold_injury_rhoave_BAYENV,
        y=results_pine2$LAT_raw_rho,
        xlab="FCI unflipped",
        ylab="LAT rho Bayenv",
        nbin=100)

# after flip - Bayenv vs. raw should look good for LAT

plot_2D(x=results_pine3$LAT_rhoave,
        y=results_pine3$LAT_raw_rho,
        xlab="Lat Bayenv",
        ylab="LAT raw rho",
        nbin=100,
        x_sub=results_pine3$LAT_rhoave[super],
        y_sub=results_pine3$LAT_raw_rho[super])

# after flip - bayenv vs. raw should look good for FCI
plot_2D(x=results_pine3$Fall_cold_injury_rhoave_BAYENV,
        y=results_pine3$Fall_cold_injury_p_raw_rho,
        xlab="FCI Bayenv",
        ylab="FCI raw rho",
        nbin=100,
        x_sub=results_pine3$Fall_cold_injury_rhoave_BAYENV[super],
        y_sub=results_pine3$Fall_cold_injury_p_raw_rho[super])

# after flip - gcta vs. raw should look good for FCI
plot_2D(x=results_pine3$Fall_cold_injury_snp_effect,
        y=results_pine3$Fall_cold_injury_p_raw_rho,
        xlab="FCI GCTA",
        ylab="FCI raw rho",
        nbin=100,
        x_sub=results_pine3$Fall_cold_injury_snp_effect[super],
        y_sub=results_pine3$Fall_cold_injury_p_raw_rho[super])

# after flip - gcta vs. raw should look good for LAT
plot_2D(x=results_pine3$LAT_snp_effect_GCTA,
        y=results_pine3$LAT_raw_rho,
        xlab="Lat GCTA",
        ylab="Lat raw rho",
        nbin=100,
        x_sub=results_pine3$LAT_snp_effect_GCTA[super],
        y_sub=results_pine3$LAT_raw_rho[super])
```

### Make array of gcontigs and tcontigs

First, merge gcontig and tcontig - use tcontig, if tcontig == NA, use gcontig.

```{r}
sum(is.na(results_pine3$tcontig))
sum(is.na(results_pine3$gcontig))

results_pine3$gtcontig <- as.character(results_pine3$tcontig)
results_pine3$gtcontig[is.na(results_pine3$tcontig)] <- as.character(results_pine3$gcontig[is.na(results_pine3$tcontig)])
sum(is.na(results_pine3$gtcontig))

### Make an array.  Each index is a gtcontig.  Within each index, list the number of SNPs.
getInd <- function(gtcontig){
  return(which(results_pine3$gtcontig %in% gtcontig))
}
### Check it works
  head(results_pine3$gtcontig, 10)
  sapply(unique(results_pine3$gtcontig[1:10]),FUN = getInd, simplify = "array")

### Apply to whole data
  gtcontig_array_names <- unique(results_pine3$gtcontig)
  length(gtcontig_array_names)
  gtcontig_array <- sapply(gtcontig_array_names,FUN = getInd, simplify = "array")  #this takes some time
    
```



Finally, write to new file and save R data!

```{r}
write.table(results_pine3,"../data/large_files/Pine_Alpha_AveRho_WithSuperLogical",col.names = TRUE, row.names=FALSE)
save(PE, gtcontig_array, gtcontig_array_names, results_pine3,PE_matchGWAS,PE_matchGEA, file="../data/large_files/Pine_Alpha_AveRho_WithSuperLogical.RData")
```

