---
title: "PEGA for Pine Analysis"
author: "Katie Lotterhos"
date: "August 24, 2015"
output: html_document
---

Basic algorithm:

1. choose environment
2. choose phenotype
3. get PE association
4. get relevant column data for each SNP (bayenv rho, bayenv bf, gwas slope, gwas p-value) using system calls
5. correct rho for different coding from gwas using flip.cor using merge
6. correct rho for PE/GWAS direction (polarize)
    + (PE-correlation sign) * (GWAS-slope sign) = correction sign
7. (remove large values that are not significant)
8. plot GWAS slope (x axis) vs GEA (y axis)
    + color points in 1st and 3rd quadrants in green (signs match expected direction)
    + color points in 2nd and 4th quadrants in purple (signs don't match expected direction)
    
#### Note
Note that the code runs fastest if you work through R in a terminal window and upload the "results" object first.  Once it is uploaded, when the markdown is created it won't try to load the "results" dataframe, which is very large.  To make this markdown file:
```
library(knitr)
library(markdown)
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis")
knit("AnalysisPEGA.Rmd")  # produces the md file
markdownToHTML("AnalysisPEGA.md", "AnalysisPEGA.html")  # converts an md file to html
```
    
```{r}
setwd("~/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA")
  if (!("hexbin" %in% installed.packages())){install.packages("hexbin")}
  library(hexbin)
library(gridExtra)
### load data
PE <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2.csv", header=TRUE)
#reorder to make largest effects on top
PE <- PE[rev(order(abs(PE$pine_correlation))),]
head(PE)

PE_matchGWAS <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGWAS.csv", header=TRUE)
PE_matchGEA <- read.csv("data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2_matchGEA.csv", header=TRUE)

flip <- read.table("data/large_files/flip.cor", header=TRUE)
head(flip)

results_head <- scan("data/large_files/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS", what = "character", nlines = 1)
results_head
#### read in results if it doesn't exist
if (!("results" %in% ls())){
results <- read.table("data/large_files/var_out_GATK3_allhet_pine688_ALL.summary.ALL.annots.sorted.GOOD.window_RESULTS", header=TRUE, comment.char=" ")
  }
dim(results)

### Upload superoutliers file
  pine_raw_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x_raw.txt", header=TRUE)
  pine_tcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_100x.txt", header=TRUE)
  pine_raw_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x_raw.txt", header=TRUE)
  pine_gcontig <- read.table("data/large_files/super_outliers_per_variable_pine_both_methods_gcontigs_100x.txt", header=TRUE)

  head(pine_tcontig)
  head(pine_gcontig)

### Get list of superoutliers
  pine_super_t <- as.character(pine_tcontig$pine[pine_tcontig$super_outlier_score > 1])
  pine_super_t_raw <- as.character(pine_raw_tcontig$pine[pine_raw_tcontig$super_outlier_score > 1])
  pine_super_g <- as.character(pine_gcontig$pine[pine_gcontig$super_outlier_score > 1])
  pine_super_g_raw <- as.character(pine_raw_gcontig$pine[pine_raw_gcontig$super_outlier_score > 1])

### Get indexes of superoutliers in 'results' dataframe
get_index <- function(superlist, t_or_g){
  # t_or_g is transcript or genomic contig
  super_results <- NULL
  for (j in 1:length(superlist)){
     if (t_or_g=="t"){ind <- which(as.character(results$tcontig) %in% superlist[j])}
     if (t_or_g=="g"){ind <- which(as.character(results$gcontig) %in% superlist[j])}
     super_results <- c(super_results, ind)
     if (j%%100==0){print(c(j, length(ind)))}
    }
  return(super_results)
}

  indexes_super_t <- get_index(pine_super_t, "t")


### Loop through PE file

for (i in 1:nrow(PE)){
### Match names
  PE_envi_name <- as.character(PE$Environment[i])
  GEA_envi_name <- as.character(PE_matchGEA$Environment_GEA[which(PE_matchGEA$Environment_PE==PE_envi_name)])
  
  PE_pheno_name <- as.character(PE$Phenotype[i])
  GWAS_pheno_name <- as.character(PE_matchGWAS$Phenotype_GWAS[which(PE_matchGWAS$Phenotype_PE==PE_pheno_name)])
  
  print(c(i, GEA_envi_name, GWAS_pheno_name))

### Get relevant columns from large file
  
  GWAS_effect_col <- which(results_head==gsub("_p","_snp_effect", GWAS_pheno_name))
  if (length(GWAS_effect_col)!=1){print("ERROR: GWAS_effect_col does not equal 1")}
  GWAS_effect <- results[,GWAS_effect_col]

  GEA_effect_cols <- grep(paste("^",GEA_envi_name,"_rho", sep=""), results_head)
  if (length(GEA_effect_cols)!=2){print("ERROR: GEA_effect_cols does not equal 2")}
  GEA_rho <- rowMeans(results[,GEA_effect_cols], na.rm=TRUE)

### check that "flip"" order matches "results" order
if (!(identical(flip$gcontig__gcontig_pos, results$gcontig__gcontig_pos))){
  print("ERROR!  SNP order in flip file does not match results file")
}

### Flip GEA_rho so sign matches GWAS_effect
  GEA_rho_flip <- GEA_rho * flip$X1

### Polarize GEA so sign aligns with PE and GWAS sign
  PE_cor <- PE$pine_correlation[i]

  GWAS_sign <- sign(GWAS_effect)
  GWAS_sign[GWAS_sign==0]=1

  if (sum(GWAS_sign==0, na.rm=TRUE)>0){print("ERROR: Some GWAS have 0 sign for polarization, will skew results")} #fixed this with code below

  if(PE_cor != 0){
  GEA_rho_polarized <- GEA_rho_flip * sign(PE_cor) * GWAS_sign
  }else{
  GEA_rho_polarized <- GEA_rho_flip * GWAS_sign  
  }

#  plot(0,0, xlim=c(-3,3), ylim=c(-0.5,0.5), col=0, main=paste(PE_envi_name, PE_pheno_name, PE$pine_correlation[i]))
  #plot(hexbin(GWAS_effect, GEA_rho_polarized,xbnds=c(-5,5), ybnds=c(-1,1), xbins=100), add=TRUE)


### Plotting 
install.packages("ash")
library(ash)

plotter <- function(GWAS_effect, GEA_rho_polarized, main){
  data1 <- cbind(GWAS_effect, GEA_rho_polarized)
  data1b <- data1[complete.cases(data1),]
  binned <- bin2(data1b, 
                 matrix(c(-3,3,-0.5,0.5), 2,2, byrow=TRUE), 
                 nbin=c(100,100))
  binned$nc[binned$nc==0]=NA
  image.plot(seq(-3,3,length.out = 100), seq(-0.5, 0.5, length.out=100),binned$nc,
             xlab="GWAS effect", ylab="GEA rho (polarized)", main=main)
  }

pdf("Pine.pdf", height=14, width=4)
  par(mfrow=c(5,1), mar=c(4,4,4,1))
  plotter(GWAS_effect, GEA_rho_polarized, main="All SNPs")
  plotter(GWAS_effect[indexes_super_t], GEA_rho_polarized[indexes_super_t], main="Super tcontigs")
dev.off()

  a<- hexbinplot(GEA_rho_polarized~GWAS_effect, xbnds=c(-5,5), ybnds=c(-1,1), xbins=100, main=paste(PE_envi_name, PE_pheno_name, PE$pine_correlation[i]) , xlim=c(-3,3), ylim=c(-0.5,0.5), aspect=1)
  a_super_t <- hexbinplot(GEA_rho_polarized[indexes_super_t]~GWAS_effect[indexes_super_t], xbnds=c(-5,5), ybnds=c(-1,1), xbins=100, main=paste(PE_envi_name, PE_pheno_name, PE$pine_correlation[i]) , xlim=c(-3,3), ylim=c(-0.5,0.5), aspect=1)
   
  par(mfrow=c(2,1), mar=c(4,4,1,1))
  plot(a)
  plot(a_super_t)
library(gridExtra)    
do.call(grid.arrange, c(list(plot(hxb), plot(hxb_super_t)), ncol=2))


 hxb<-hexbin(GWAS_effect, GEA_rho_polarized,xbnds=c(-5,5), ybnds=c(-1,1), xbins=100)
  hxb_super_t <- hexbin(GWAS_effect[indexes_super_t], GEA_rho_polarized[indexes_super_t],xbnds=c(-5,5), ybnds=c(-1,1), xbins=100)

plotscript <- function(superind){
  pp<- plot(hxb)
  pushHexport(pp$plot.vp)
  x <- GWAS_effect[superind]
  y <- GEA_rho_polarized[superind]
    grid.points(x,y,gp=gpar(col=rgb(0,0,1,0.3)),
                pch=20,size = unit(0.5, "char"))
  popViewport()
}
plotscript(indexes_super_t)



  png(paste("analysis/figure_byenvi/PEGA_",PE_envi_name,"_",i,"_",PE_pheno_name, ".png",sep=""), width=6, height=5, units="in", res=300)
  par(mar=c(1,1,1,1))
  plot(a)
  #do.call(grid.arrange, c(a, ncol=1))
  dev.off()

  png(paste("analysis/figure_byPEcor/PEGA_",i, "_", PE_envi_name,"_",PE_pheno_name,"_",i, ".png",sep=""), width=6, height=5, units="in", res=300)
  par(mar=c(1,1,1,1))
  plot(a)
  #do.call(grid.arrange, c(a, ncol=1))
  dev.off()


  png(paste("analysis/figure_byPheno/PEGA_" ,PE_pheno_name,"_",i, "_", PE_envi_name,".png",sep=""), width=6, height=5, units="in", res=300)
  par(mar=c(1,1,1,1))

  #do.call(grid.arrange, c(a, ncol=1))
  dev.off()

} #end loop