---
title: "PEGA for Pine Analysis: Step 2 choose outlier groups"
author: "Katie Lotterhos"
date: "April 24, 2016"
output: html_document
---

### Load data
setwd("/Users/katie/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis/")
```{r, label="load data"}
load("../data/large_files/Pine_Alpha_AveRho_WithSuperLogical.RData")
  #if(!("ash" %in% installed.packages())){install.packages("ash")}
  require(ash)
  require(fields)
  require(dplyr)
  require(cluster)
```
This loads:

* PE
* results_pine3
* gtcontig_array 
* gtcontig_array_names
* PE_matchGWAS
* PE_matchGEA

### Plotting function plot_2D.b
plot_2D.b does not flip the phenotype axis.  Instead it colors the quadrant for the expected direction based on the PE association. This version also draws the 95% CI on the covariance matrix.

```{r}
plot_2D.b <- function(x,y, xlab, ylab, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3), nbin, x_sub_orange=NULL, y_sub_orange=NULL, x_sub_green=NULL, y_sub_green=NULL, x_sub_blue=NULL,  y_sub_blue=NULL,x_sub_golden=NULL,y_sub_golden=NULL, PE=1, PElab = c("V1", "V2")){
  #if(PE<0){x <- -x; x_sub <- -x_sub}

  data1 <- cbind(x, y)
  data1b <- data1[complete.cases(data1),]
  if(length(xlim)==0){
    xlim_up <- max(x, na.rm=TRUE)+0.15*max(x, na.rm=TRUE)
    xlim_lower <- min(x, na.rm=TRUE)-0.15*max(x, na.rm=TRUE)
  }else{
    xlim_lower <- xlim[1]; xlim_up <- xlim[2]
  }
  if(length(ylim)==0){
    ylim_up <- max(y, na.rm=TRUE)+0.15*max(y, na.rm=TRUE)
    ylim_lower <- min(y, na.rm=TRUE)-0.15*max(y, na.rm=TRUE)
  }else{
    ylim_lower <- ylim[1]; ylim_up <- ylim[2]
  }
        binned <- bin2(data1b,
                 matrix(c(xlim_lower,xlim_up,ylim_lower,ylim_up), 2,2, byrow=TRUE),
                 nbin=c(nbin,nbin))
    binned$nc[binned$nc==0]=NA

    ### Start plotting 
    xlocs <- seq(xlim_lower,xlim_up,length.out = nbin)
    ylocs <- seq(ylim_lower,ylim_up, length.out=nbin)
    image(xlocs,  ylocs,
               z=matrix(0, nrow=length(ylocs), ncol=length(xlocs)),
          col=(rgb(0,0,0,0)),
             xlab=xlab, ylab=ylab, add=FALSE, bty="n")#, 
             #xaxt="n", yaxt="n", bty="n")
    
    if(PE >= 0.1){
      polygon(c(0,1,1,0,0), c(0,0,1,1,0),col=rgb(0,0,0,0.05))
      polygon(c(0,-1,-1,0,0), c(0,0,-1,-1,0), col=rgb(0,0,0,0.05))
      }
    if(PE <= -0.1){
      polygon(c(0,-1,-1,0,0), c(0,0,1,1,0),col=rgb(0,0,0,0.05))
      polygon(c(0,1,1,0,0), c(0,0,-1,-1,0),col=rgb(0,0,0,0.05))
    }
    lines(c(-100,100), c(0,0), col="black") #add x axis
    lines(c(0,0), c(-100,100), col="black") # add y axis
    
    image(seq(xlim_lower,xlim_up,length.out = nbin), 
               seq(ylim_lower,ylim_up, length.out=nbin),
               binned$nc,
             xlab=xlab, ylab=ylab, add=TRUE, 
             col=grey.colors(75, 0.8,0.1))
             #xaxt="l", yaxt="l", bty="n")
    
    ### Function for plotting covariance matrix
    plot_mycov <- function(data1, CI = 0.95, color="black", ...){
      # data1 is a data frame with x, y in columns
      C.ls2 <- cov(data1[complete.cases(data1),])
      m.ls2 <- colMeans(data1[complete.cases(data1),])
      d2.95 <- qchisq(CI, df = 2)
      #d2.95 <- qnorm(0.99)
      lines(ellipsoidPoints(C.ls2, d2.95, loc=m.ls2), lwd=2, col=color, ...)
    }
    
    ### Make cov plot for all data
    plot_mycov(data1)
    
    ### Make function for plotting sub points
    plot_mycov_sub <- function(x_sub, y_sub, outlinecolor, bgcolor, linecolor,  mylty, mypch, ...){
      if(length(x_sub)>0){
      points(x_sub, y_sub, cex=1, col=outlinecolor, bg=bgcolor, pch=mypch)
      plot_mycov(data.frame(x_sub, y_sub), col=linecolor, lty=mylty,...)
      }
    }
    
    ### Make cov plot for x_sub_orange, y_sub_orange
    plot_mycov_sub(x_sub_orange, y_sub_orange, outlinecolor=adjustcolor("darkorange",0.7), bgcolor=adjustcolor("orange",0.7), mypch=24, mylty=2, linecolor="#CC6600")
    
    plot_mycov_sub(x_sub_green, y_sub_green, linecolor="darkgreen", bgcolor="lightgreen", mypch=21, mylty=3)
    
    plot_mycov_sub(x_sub_blue, y_sub_blue, outlinecolor=adjustcolor("blue",0.7), bgcolor=adjustcolor("lightblue",0.7), mypch=22, mylty=4, linecolor="darkblue")
    
    plot_mycov_sub(x_sub_golden, y_sub_golden, linecolor="darkred", bgcolor="gold", mypch=23, mylty=6)
    

    #text(xlim_up*0.05, ylim_up*0.95, "+", cex=2, col="black")
    #text(xlim_up*0.05, ylim_lower*0.95, "-", cex=2, col="black")
    
    #text(xlim_up*0.95,ylim_up*0.05,"+", cex=2, col="black")       
    #text(xlim_lower*0.95,ylim_up*0.05,"-", cex=2, col="black")

    t <- bquote(~rho ~ "(" ~ .(PElab[1]) ~ "," ~ .(PElab[2]) ~ ") =" ~ .(PE))
    text(xlim_up*0.5,ylim_lower*0.95, t, cex=1.2)
} # end plot_2D.b
```



### Plotting function for muliple panels, ordered by dendogram of covariances among phenotypes and environments

```{r}
head(PE)
load("../../20160125 heatmap denograms envi pheno/PineHeatmap.Rdata")
# loads env,env_clust,phen,phen_clust,EvP, EvPclust
  # env, phen, and EvP are correlation matrices based on raw data
  # env_clust and phen_clust are heatmap.2 objects showing how variables were ordered
names(env_clust)
(xorder <- rownames(env_clust$carpet)) # plotted in x axis in this order
colnames(env_clust$carpet) # plotted in y axis in this order

(yorder <- rev(rownames(phen_clust$carpet))) # plotted phenotypes in this order
#yorder <- yorder[-4] # remove root weight
pheno_ind <- c(1:3,5:17)
colnames(phen_clust$carpet) # plotted in y axis in this order

#### Function for plotting according to dendogram order
MakeOutlierPlot <- function(filename, Group_logical, envi_ind=1:22, pheno_ind=c(1:3,5:17), xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),  width=50, height=66, type="Bayenv", Green_logical=NULL, Blue_logical=NULL, Golden_logical=NULL){
    #type = "Bayenv" or "Raw"
  
  pdf(file = filename,width=width,height=height)
  par(mfrow=c(length(envi_ind),length(pheno_ind)), mar=c(2,2,0,0), oma=c(6,6,6,0))
  for (i in envi_ind){ ## loop through environments
    for (j in pheno_ind){ ## loop through phenotypes
     
      ### y = get GEA column
      if(type=="Bayenv"){gea_col <- grep(paste("^",xorder[i],"_rhoave", sep=""),names(results_pine3))}
      if(type=="Raw"){gea_col <- grep(paste("^",xorder[i],"_raw_rho", sep=""),names(results_pine3))}
      
      y <- results_pine3[,gea_col] #carrot is needed for FFP
      
      ### x = get GPA column
        gpa1 <- PE_matchGWAS$Phenotype_GWAS[which(PE_matchGWAS$Abb == yorder[j])]
        if(gpa1=="root_wt__shoot_wt_p"){
          gpa1 <- "root_wt_shoot_wt_p"}
        if(gpa1=="root_wt__shoot_wt_p_1"){
          gpa1 <- "root_wt_shoot_wt_p_1"} 
        
        if(type=="Bayenv"){
          gpa2<- gsub("_p","",gpa1)
          gpa_col <- grep(paste("^",gpa2, "_rhoave_BAYENV", sep=""), names(results_pine3))
        }
        if(type=="Raw"){gpa_col <- grep(paste("^",gpa1,"_raw_rho", sep=""), names(results_pine3))}
  
      x <- results_pine3[,gpa_col]
    # pecorr = get PE
      pecorr <- PE$pine_correlation[(PE$Environment==PE_matchGEA$Environment_PE[PE_matchGEA$Environment_GEA == xorder[i]]) & PE$PhenAbb==yorder[j]]
      
      # make plot2D
      plot_2D.b(x,y, 
         xlab= "", xlim=xlim,
         ylab= "" , ylim=ylim,
         nbin=100, 
         x_sub_orange=x[Group_logical], 
         y_sub_orange=y[Group_logical],
         PE=pecorr, 
         x_sub_green=x[Green_logical],
         y_sub_green=y[Green_logical],
         x_sub_blue=x[Blue_logical],
         y_sub_blue=y[Blue_logical],
         x_sub_golden=x[Golden_logical],
         y_sub_golden=y[Golden_logical]
         ) 
      mtext(xorder[i], side=2)
      mtext(yorder[j], side=1)
    } #end j loop
    print(i)
    } #end i loop
  dev.off()

}# end function
```

### Group 1: Top SNP's from Sam's top candidates

```{r,label="Top SNP's from Sam's top candidates" }
  superdf <- results_pine3[results_pine3$pine_super_p9 | 
                           results_pine3$pine_super_raw_p9,]
  names(results_pine3)[grep("raw_rho", names(results_pine3))]
  cols_raw <- grep("raw_rho", names(results_pine3))
  is.superdf <- matrix(NA, nrow(superdf), length(cols_raw))
  colnames(is.superdf) <- names(results_pine3)[cols_raw]
  tail <- 0.999
  for (i in 1:length(cols_raw)){
    a<- sort(abs(results_pine3[,cols_raw[i]]), na.last = NA)
    sorted.index <- round(length(a)*.999)
    rho_cutoff <- a[sorted.index]
    is.superdf[,i] <- abs(superdf[,cols_raw[i]])>rho_cutoff
    print(i)
  }
  num.var.raw.out <- rowSums(is.superdf,na.rm = TRUE)
  nrow(superdf) #total number of SNPs in super contigs
  sum(num.var.raw.out>0) # number of super SNPs in super contigs
  
#number gcontigs represented by all super vs. super SNPs
  length(unique(superdf$gtcontig)) # all Sam's super contigs
  length(unique(superdf$gtcontig[num.var.raw.out>0])) # super contigs with superSNPs

  ### Dataframe of Group 1
  Group1_allSuperSNPs <- superdf[num.var.raw.out>0,]
  dim(Group1_allSuperSNPs)
  ### Logical Group 1
  Group1_logical <- results_pine3$gcontig__gcontig_pos %in% Group1_allSuperSNPs$gcontig__gcontig_pos


  ### Group 1 array
  #png(file = "GroupEnvByPheno.png",width=50,height=66, unit="in", res=300)
  Group1_allSuperSNPs_array_names <- unique(Group1_allSuperSNPs$gtcontig)
  Group1_allSuperSNPs_array <- sapply(Group1_allSuperSNPs_array_names, getInd2, Group1_allSuperSNPs)
  
    ### snps per contig in top candidates
  G1_df <- data.frame(contig=Group1_allSuperSNPs_array_names,
             numOutSnps=as.numeric(sapply(Group1_allSuperSNPs_array, length)))
  dim(G1_df)
  G1_df <- G1_df[order(G1_df$numOutSnps, decreasing = TRUE),]
  G1_df$rank_numOut <- rank(1/G1_df$numOutSnps) 
    # index shows order by number of SNPs - contig with most outlier SNPs # 1
  G1_df$numTotSnps <- sapply(1:nrow(G1_df), function(i){sum(results_pine3$gtcontig==G1_df$contig[i])})
  G1_df$prop <- G1_df$numOutSnps/G1_df$numTotSnps
    # proportion of outlier SNPs in that contig
  
  ### Print top candidate list to html file, ordered by the contigs with the higest proportion of outliers SNPS
  G1_df$rank_prop <- rank(1/G1_df$prop)
  G1_df[order(G1_df$prop, decreasing=TRUE),]

   #topcan <- as.character(G1_df$contig[G1_df$prop>0.05])
   #length(topcan)
  
  ### Which ones are super convergent outliers
  results_pine3$gtcontig[Group1_logical & results_pine3$pine_super_convergent]

  names(results_pine3)
  results_pine3$Group1_logical <- Group1_logical
```


### Group 1 Make some comparison plots
```{r}

names(results_pine3)[grep("MAT", names(results_pine3))]

plot_2D.b(x=results_pine3$MAT_rhoave,
       y=results_pine3$MAP_rhoave, 
       xlab= "MAT rho Bayenv", 
       ylab= "MAP rho Bayenv" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=results_pine3$MAT_rhoave[Group1_logical], 
       y_sub_orange=results_pine3$MAP_rhoave[Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))

plot_2D.b(x=results_pine3$MAT_raw_rho,
       y=results_pine3$MAP_raw_rho, 
       xlab= "MAT rho raw", 
       ylab= "MAP rho raw" , 
       nbin=100, xlim=c(-0.8,0.8), ylim=c(-0.8,0.8),
       x_sub_orange=results_pine3$MAT_raw_rho[Group1_logical], 
       y_sub_orange=results_pine3$MAP_raw_rho[Group1_logical],
       PE=0.33, PElab=c("MAT", "MAP"))

plot_2D.b(x=results_pine3$Fall_cold_injury_rhoave_BAYENV,
       y=results_pine3$LAT_rhoave, 
       xlab= "FCI rho Bayenv", 
       ylab= "LAT rho Bayenv" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=superdf$Fall_cold_injury_rhoave_BAYENV[num.var.raw.out>0], 
       y_sub_orange=superdf$LAT_rhoave[num.var.raw.out>0],
       PE=-0.42, PElab=c("FCI", "LAT"))

plot_2D.b(x=results_pine3$Fall_cold_injury_rhoave_BAYENV,
       y=results_pine3$MAP_rhoave, 
       xlab= "FCI rho Bayenv", 
       ylab= "LAT rho Bayenv" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=superdf$Fall_cold_injury_rhoave_BAYENV[num.var.raw.out>0], 
       y_sub_orange=superdf$MAP_rhoave[num.var.raw.out>0],
       PE=0.25)

plot_2D.b(x=results_pine3$Fall_cold_injury_p_raw_rho,
       y=results_pine3$MAP_raw_rho, 
       xlab= "FCI rho Bayenv", 
       ylab= "LAT rho Bayenv" , 
       nbin=100, xlim=c(-1,1), ylim=c(-1,1),
       x_sub_orange=superdf$Fall_cold_injury_p_raw_rho[num.var.raw.out>0], 
       y_sub_orange=superdf$MAP_raw_rho[num.var.raw.out>0],
       PE=0.25)

plot_2D.b(x=results_pine3$Fall_cold_injury_rhoave_BAYENV,
       y=results_pine3$Eref_rhoave, 
       xlab= "FCI rho Bayenv", 
       ylab= "LAT rho Bayenv" , 
       nbin=100, xlim=c(-0.3,0.3), ylim=c(-0.3,0.3),
       x_sub_orange=superdf$Fall_cold_injury_rhoave_BAYENV[num.var.raw.out>0], 
       y_sub_orange=superdf$Eref_rhoave[num.var.raw.out>0],
       PE=0.26)


plot_2D.b(x=results_pine3$Fall_cold_injury_p_raw_rho,
       y=results_pine3$LAT_raw_rho, 
       xlab= "FCI rho raw", 
       ylab= "LAT rho raw" , 
       nbin=100, xlim=c(-1,1), ylim=c(-1,1),
       x_sub_orange=superdf$Fall_cold_injury_p_raw_rho[num.var.raw.out>0], 
       y_sub_orange=superdf$LAT_raw_rho[num.var.raw.out>0],
       PE=-0.42)


plot_2D.b(x=results_pine3$Fall_cold_injury_snp_effect,
       y=results_pine3$LAT_snp_effect_GCTA, 
       xlab= "FCI gcta", 
       ylab= "LAT gcta" , 
       nbin=100, xlim=NULL, ylim=NULL,
       x_sub_orange=superdf$Fall_cold_injury_snp_effect[num.var.raw.out>0], 
       y_sub_orange=superdf$LAT_snp_effect_GCTA[num.var.raw.out>0],
       PE=-0.42)

````

### Make some outlier plots for everything

```{r, eval=FALSE}
MakeOutlierPlot("Group1_EnvByPhenoBayenv1.pdf", Group1_logical, type="Bayenv")
MakeOutlierPlot("Group1_EnvByPhenoRaw1.pdf", Group1_logical, type="Raw", xlim=c(-1,1), ylim=c(-1,1))
```


### Group 2: Lat and Low-temp variables
```{r}
colnames(is.superdf)
Group2cols.superdf <- colnames(is.superdf) %in% c(
                                          "LAT_raw_rho",
                                          "DD_0_raw_rho")

 ### Dataframe of Group 2
  Group2_allSuperSNPs <- superdf[rowSums(is.superdf[,Group2cols.superdf],na.rm = TRUE)>0,]
  
  ### Logical Group 2
  Group2_logical <- results_pine3$gcontig__gcontig_pos %in% Group2_allSuperSNPs$gcontig__gcontig_pos
  ## Check these two are equal
  nrow(Group2_allSuperSNPs)
  sum(Group2_logical)

  ### Group2 array for boostrap covariance
  Group2_allSuperSNPs_array_names <- unique(Group2_allSuperSNPs$gtcontig)
  Group2_allSuperSNPs_array <- sapply(Group2_allSuperSNPs_array_names, getInd2, Group2_allSuperSNPs)
  length(Group2_allSuperSNPs_array)
  
  #snps per contig
  Group2_df <- data.frame(contig=Group2_allSuperSNPs_array_names,
             numOutSnps=as.numeric(sapply(Group2_allSuperSNPs_array, length)))
  dim(Group2_df)
  Group2_df$rank_numOut <- rank(1/Group2_df$numOutSnps)
    # index shows order by number of SNPs - contig with most outlier SNPs # 1
  Group2_df$numTotSnps <- sapply(1:nrow(Group2_df), function(i){sum(results_pine3$gtcontig==Group2_df$contig[i])})
  Group2_df$prop <- Group2_df$numOutSnps/Group2_df$numTotSnps
    # proportion of outlier SNPs in that contig
  
  ### Print top candidate list to html file, ordered by the contigs with the higest proportion of outliers SNPS
  Group2_df$rank_prop <- rank(1/Group2_df$prop)
  Group2_df[order(Group2_df$prop, decreasing=TRUE),]
  
  ### Add column to results_pine3
  results_pine3$Group2_logical <- Group2_logical
```


### Designate small groups of focus variables
```{r}
xorder
envi_ind=c(2, 5, 6, 9, 10, 11, 12, 13, 17)
xorder[envi_ind]

yorder
pheno_ind=c(1,15, 16,17)
yorder[pheno_ind]

envi_ind_sm <- c(2,9,10)
xorder[envi_ind_sm]

pheno_ind_sm <- c(1,16)
yorder[pheno_ind_sm]

sum(Group2_logical & results_pine3$gtcontig=="comp60064_c0_seq1") 
#should equal 46
```

```{r, eval=FALSE}
### All 
MakeOutlierPlot("Group2_Lat_DD0_EnvByPhenoBayenv2.pdf", Group2_logical, type="Bayenv", envi_ind = envi_ind, pheno_ind=pheno_ind, width=20, height=24,
    Blue_logical=Group2_logical & results_pine3$gtcontig=="comp60064_c0_seq1",
    Green_logical= Group2_logical & results_pine3$gtcontig=="comp1563_c0_seq1",
    Golden_logical=Group2_logical & results_pine3$gtcontig=="comp73926_c0_seq1")

MakeOutlierPlot("Group2_Lat_DD0_EnvByPhenoRaw2.pdf", Group2_logical, type="Raw", envi_ind = envi_ind, pheno_ind=pheno_ind, width=20, height=24, xlim=c(-1,1), ylim=c(-1,1),
                    Blue_logical=Group2_logical & results_pine3$gtcontig=="comp60064_c0_seq1",
    Green_logical= Group2_logical & results_pine3$gtcontig=="comp1563_c0_seq1",
    Golden_logical=Group2_logical & results_pine3$gtcontig=="comp73926_c0_seq1")
```


### Group 3 CMD/MAP RAW outliers
```{r}
  Group3cols.superdf <- colnames(is.superdf) %in% c("CMD_raw_rho", "MAP_raw_rho")

 ### Dataframe of Group 3
  Group3_allSuperSNPs <- superdf[rowSums(is.superdf[,Group3cols.superdf],na.rm = TRUE)>0,]
   
  ### Logical Group 3
  Group3_logical <- results_pine3$gcontig__gcontig_pos %in% Group3_allSuperSNPs$gcontig__gcontig_pos
  ## Check these two are equal
  nrow(Group3_allSuperSNPs)
  sum(Group3_logical)

  ### Group3 array contigs
  Group3_allSuperSNPs_array_names <- unique(Group3_allSuperSNPs$gtcontig)
  Group3_allSuperSNPs_array <- sapply(Group3_allSuperSNPs_array_names, getInd2, Group3_allSuperSNPs)
  length(Group3_allSuperSNPs_array)
    # number of contigs in Group 3
  
  #snps per contig
  Group3_df <- data.frame(contig=Group3_allSuperSNPs_array_names,
             numOutSnps=as.numeric(sapply(Group3_allSuperSNPs_array, length)))
  #34 contigs
  
  dim(Group3_df)
  Group3_df$rank_numOut <- rank(1/Group3_df$numOutSnps)
    # index shows order by number of SNPs - contig with most outlier SNPs # 1
  Group3_df$numTotSnps <- sapply(1:nrow(Group3_df), function(i){sum(results_pine3$gtcontig==Group3_df$contig[i])})
  Group3_df$prop <- Group3_df$numOutSnps/Group3_df$numTotSnps
    # proportion of outlier SNPs in that contig
  
  ### Print top candidate list to html file, ordered by the contigs with the higest proportion of outliers SNPS
  Group3_df$rank_prop <- rank(1/Group3_df$prop)
  Group3_df[order(Group3_df$prop, decreasing=TRUE),]
  
  ### Add column to results_pine3
  results_pine3$Group3_logical <- Group3_logical
  
```


### Compare Group 2 and Group 3
```{r, eval=FALSE}
MakeOutlierPlot("Group2_vs_Group3_EnvByPhenoBayenv1.pdf", Group2_logical , type="Bayenv", envi_ind = envi_ind, pheno_ind=pheno_ind, width=20, height=24,
    Blue_logical=Group3_logical) 
MakeOutlierPlot("Group2_vs_Group3_EnvByPhenoBayenv_sm.pdf", Group2_logical , type="Bayenv", envi_ind = envi_ind_sm, pheno_ind=pheno_ind_sm, width=10, height=12,
    Blue_logical=Group3_logical) 

MakeOutlierPlot("Group2_vs_Group3_EnvByPhenoRaw1.pdf", Group2_logical, type="Raw", envi_ind = envi_ind, pheno_ind=pheno_ind, width=20, height=24, xlim=c(-1,1), ylim=c(-1,1),
    Blue_logical=Group3_logical)
MakeOutlierPlot("Group2_vs_Group3_EnvByPhenoRaw1_sm.pdf", Group2_logical, type="Raw", envi_ind = envi_ind_sm, pheno_ind=pheno_ind_sm, width=10, height=12, xlim=c(-1,1), ylim=c(-1,1),
    Blue_logical=Group3_logical)

MakeOutlierPlot("Group1sm_EnvByPhenoBayenv1.pdf", Group1_logical , type="Bayenv", envi_ind = envi_ind_sm, pheno_ind=pheno_ind_sm, width=10, height=12) 

MakeOutlierPlot("Group1sm_EnvByPhenoRaw1.pdf", Group1_logical, type="Raw", envi_ind = envi_ind_sm, pheno_ind=pheno_ind_sm, width=10, height=12, xlim=c(-1,1), ylim=c(-1,1))
```




```{r}
x=results_pine3$root_wt_shoot_wt_1_rhoave_BAYENV 
  # Total weight (ratio is without 1)
y=results_pine3$Fall_cold_injury_rhoave_BAYENV
par(mar=c(4,4,1,1))
plot_2D.b(x,y,xlab="Total biomass", ylab="FCI", 
                      xlim=c(-0.3,0.3), 
                      ylim=c(-0.3,0.3), 
                      nbin=100, 
                      x_sub_orange=x[Group2_logical], 
                      y_sub_orange=y[Group2_logical],
                      x_sub_blue=x[Group3_logical], 
                      y_sub_blue=y[Group3_logical],
                      PE=-1)

# other outliers that aren't in the two groups
plot_2D.b(x,y,xlab="Total Weight", ylab="FCI", 
                      xlim=c(-0.3,0.3), 
                      ylim=c(-0.3,0.3), 
                      nbin=100, 
                      x_sub_orange=x[Group1_logical & !Group2_logical & !Group3_logical], 
                      y_sub_orange=y[Group1_logical & !Group2_logical & !Group3_logical],
                      PE=-1)

x=results_pine3$root_wt_shoot_wt_p_1_raw_rho 
y=results_pine3$Fall_cold_injury_p_raw_rho
plot_2D.b(x,y,xlab="Total weight", ylab="FCI", 
                      xlim=c(-1,1), 
                      ylim=c(-1,1), 
                      nbin=100, 
                      x_sub_orange=x[Group2_logical], 
                      y_sub_orange=y[Group2_logical],
                      x_sub_blue=x[Group3_logical], 
                      y_sub_blue=y[Group3_logical],
                      PE=-1)
plot_2D.b(x,y,xlab="Total weight", ylab="FCI", 
                      xlim=c(-1,1), 
                      ylim=c(-1,1), 
                      nbin=100, 
                      x_sub_orange=x[Group1_logical & !Group2_logical & !Group3_logical], 
                      y_sub_orange=y[Group1_logical & !Group2_logical & !Group3_logical], 
                      PE=-1)
```


### Make outlier plots - one for each candidate gene
This takes a while. To run this code, set MakeThesePlots <- TRUE below
```{r, eval=FALSE, fig.width=12}
MakeThesePlots <- FALSE
if(MakeThesePlots){
 ### Loop through top genes but for subset of envi
  for (i in 1:length(topcan)){
    magLogical <- results_pine3$gtcontig==topcan[i] & Group1_logical #outlier SNPs
    blueLogical <- results_pine3$gtcontig==topcan[i] & !Group1_logical #other SNPs in the gene
    nameBay <- paste(round(G1_df$prop[G1_df$contig==topcan[i]],2),
                     G1_df$numOutSnps[G1_df$contig==topcan[i]], 
                     G1_df$numTotSnps[G1_df$contig==topcan[i]],
                     topcan[i], "BayenvSub.pdf", sep="_")
    nameRaw <- paste(round(G1_df$prop[G1_df$contig==topcan[i]],2),
                     G1_df$numOutSnps[G1_df$contig==topcan[i]], 
                     G1_df$numTotSnps[G1_df$contig==topcan[i]],
                     topcan[i], "RawSub.pdf", sep="_")
    MakeOutlierPlot(nameBay, magLogical, type="Bayenv", Blue_logical = blueLogical,envi_ind = envi_ind_sm, pheno_ind=pheno_ind_sm, width=10, height=12)
    MakeOutlierPlot(nameRaw, magLogical, type="Raw", Blue_logical = blueLogical, xlim=c(-1,1), ylim=c(-1,1), envi_ind = envi_ind_sm, pheno_ind=pheno_ind_sm, width=10, height=12)
    print(c("done with", i, "of", length(topcan)))
  } #end for loop
}# end if
```

```{r, eval=FALSE}
write.table(G1_df, "../PineTopCandidates.txt", col.names=TRUE, row.names=FALSE)
G1_df

### 
```

### Merge G1_df with Kay's annotations
```{r}
go <- read.table("../GOanalysis/tair_pine_katie2", header=TRUE)
names(go)
names(G1_df)
names(go)[1] <- "contig2"
head(go$contig2)
head(G1_df$contig)
### Remove seq1, but check "seq1" in all rows first
length(grep("_seq1", G1_df$contig))==nrow(G1_df)

G1_df$contig2 <- sub("_seq1", "", G1_df$contig)
G1_df_go <- merge(G1_df, go, all=TRUE)
head(G1_df_go, 1)
tail(G1_df_go, 1)
nrow(G1_df_go) # should equal 107
write.table(G1_df_go, "../PineTopCandidates_WithGO.txt", col.names=TRUE, row.names=FALSE)
```

```{r, label="Plot MAT vs MAP"}
ff <- read.table ("../data/MAT06-BLUEs&Climate-pine.csv", sep = ",", header = T)
head(ff)
PE_MAT_MAP = cor(ff$MAT, ff$log.MAP.,use = "pairwise.complete.obs")

par(mar=c(4,5,1,1))

plot_2D.b(results_pine3$MAT_rhoave, results_pine3$MAP_rhoave, 
          expression(rho*"["*p*", mean annual temperature (MAT)]"), 
          expression(rho*"["*p*", mean annual precipitation (MAP)]"), 
          nbin=100, 
          x_sub_orange=results_pine3$MAT_rhoave[Group2_logical], 
          y_sub_orange=results_pine3$MAP_rhoave[Group2_logical], 
          x_sub_blue = results_pine3$MAT_rhoave[Group3_logical], 
          y_sub_blue=results_pine3$MAP_rhoave[Group3_logical], 
          PE = round(PE_MAT_MAP,2), PElab=c("MAT", "MAP"))
```

### Save .Rdata for further analysis (optional)
```{r}
save=FALSE
if (save==TRUE){
  out <- "../data/large_files/Pine_Alpha_AveRho_WithSuperLogical_OutputPEGA_2FindOutlierGroups.RData"
  save.image(file=out)
}

load=FALSE
if (load==TRUE){
  load(out)
}
```
Note that saving the image in this way uses a lot of unneccesary hard drive space.

### Summary of what was done so far:
- took Sam's list of super outliers - genes that have more outlier snps than predicted by random based on raw or corrected correlations at p9 level
- reduced list further to genes with SNPs in > 0.999 on raw rho (Group 1)
- I noticed two groups of behavior.  Chose variables that had strongest PE associations.
  - Group temp: subset of Group 1 that are raw outliers in LAT and DD0
  - Group moisture: subset of Group 1 that are raw outliers in MAP and CMD

### Next steps: assess significance
- Compare: Group 2 vs. random match set; Group 3 vs. random match set; other super SNPs in Group 1 but not in groups 2 or 3.
- Loop through pairs of variables

Function ()
  Loop over SNPs:
  - make a focal df and a dummy df for all data
  - Sample a SNP from the focal group and store
  - Sample a matching SNP by allele frequency from the dataset and store
  - reduce the focal df and dummy df by removing all SNPs in the gtcontig associated with     focal SNPs
  End loop
  return covariances for focal group and null group
End function

replicate or mapply function 1000 times
calculate 95% quantiles
paired t-test

```{r}
### Part 1: Get the right columns ####
  p.bob <- PE_matchGWAS$Phenotype_GWAS[match(yorder[pheno_ind], PE_matchGWAS$Abb)]
    #because nothing was standardized, need to adjust names
  (p.bob1 <- gsub("__", "_", p.bob))
  (gpa2<- gsub("_p","",p.bob1))
  
  (phen_names <- yorder[pheno_ind])
  (pheno_bayenv_cols <- match(paste(gpa2, "_rhoave_BAYENV", sep=""), names(results_pine3)))
  (pheno_raw_cols <-  match(paste(p.bob1,"_raw_rho", sep=""), names(results_pine3)))
  ## check
  phen_df <- data.frame(name=unlist(phen_names), 
             bayenv_name=as.character(names(results_pine3)[pheno_bayenv_cols]),
             raw_name=as.character(names(results_pine3)[pheno_raw_cols]),
             bayenv_col=pheno_bayenv_cols,
             raw_col=pheno_raw_cols, stringsAsFactors=FALSE)
  phen_df <- phen_df[c(3,2,4,1),] #reorder for analysis
  phen_df$type <- "P"
  
  (env_names <- xorder[envi_ind])
  (env_bayenv_cols <- match( paste(xorder[envi_ind],"_rhoave", sep=""), names(results_pine3)))
  (env_raw_cols <-  match( paste(xorder[envi_ind],"_raw_rho", sep=""), names(results_pine3)))
  ## unit test
  env_df <- data.frame(name=env_names, 
                   bayenv_name=as.character(names(results_pine3)[env_bayenv_cols]),
                       raw_name=as.character(names(results_pine3)[env_raw_cols]),
                       bayenv_col=env_bayenv_cols,
                       raw_col=env_raw_cols, stringsAsFactors=FALSE)
  env_df$type <- "E"
  
  ### final dataframe
  final_df <- rbind(phen_df, env_df)
  str(final_df)

  # test_set <- results_pine3$Group2_logical
  info1 <- results_pine3[,c("minor_freq","gtcontig", "pos_gcontig")]
  # head(info1)
  
### Part 2: Get correlation between variables ####
  ls()
  pine_corrs <- load("../data/PineHeatmap.Rdata")
  # this data was analyzed for the Science Paper and is in the folder
  # folder: /20160125 heatmap denograms envi pheno/
  # R script: AdaptreeEnviPhenoCorrs_pine.R
  # loads env,env_clust,phen,phen_clust,EvP, EvPclust
  head(env) ## E vs E correlations
    colnames(env)
  head(phen) ## P vs P correlations
    colnames(phen)
  head(EvP) ## E vs P correlations
    colnames(EvP)
    rownames(EvP)
  
### Part 3: Make graphs ####
    ## Here I do all combos of i and j, because it is useful to be able
    ## to compare all variables along any chosen axis
  for (i in 1:nrow(final_df)){
    for (j in 1:nrow(final_df)){
      if (i != j){
      print(c(final_df$name[i], final_df$name[j]))
      
      if (final_df$type[i]=="P" & final_df$type[j]=="P"){
        rowN <- which(colnames(phen)==final_df$name[i])
        colN <- which(colnames(phen)==final_df$name[j])
        pecorr = phen[rowN,colN]
      }
      
      if (final_df$type[i]=="E" & final_df$type[j]=="E"){
        rowN <- which(colnames(env)==final_df$name[i])
        colN <- which(colnames(env)==final_df$name[j])
        pecorr = env[rowN,colN]
      }
      
      if (final_df$type[i]=="P" & final_df$type[j]=="E"){
        colN <- which(colnames(phen)==final_df$name[i]) #phenotype in column of EvP
        rowN <- which(colnames(env)==final_df$name[j]) #envi in row of EvP
        pecorr = EvP[rowN,colN]
      }
      
      if (final_df$type[i]=="E" & final_df$type[j]=="P"){
        colN <- which(colnames(phen)==final_df$name[j]) #phenotype in column of EvP
        rowN <- which(colnames(env)==final_df$name[i]) #envi in row of EvP
          # note switch of i and j compared to last 'if' statement
        pecorr = EvP[rowN,colN]
      }
          
      ### Making a four panel plot ####
        ### Upper left plot: plot 2D bayenv
        ### Upper right plot: plot randomizations bayenv
        ### Lower left plot: plot 2D raw
        ### Lower right plot: plot randomizations raw
      pdf(paste("../figures201611/", final_df$name[i],"_", 
                final_df$name[j], ".pdf", sep=""), width=8, height=8)
      par(mfrow=c(2,2), mar=c(2,2,2,2), oma=c(0,0,0,0))
      for(type in c("Bayenv", "Raw")){
        if(type=="Bayenv"){colz=final_df$bayenv_col}
        if(type=="Raw"){colz=final_df$raw_col}
        
        ### There is some redundancy by looping over types this way
        ### because matching SNPs are based on maf and would be the same
  
          x1 <- results_pine3[,colz[i]]
          y1 <- results_pine3[,colz[j]]
          
          ### Upper left plot: plot 2d Bayenv ####
          if(type=="Bayenv"){lim=0.3}
          if(type=="Raw"){lim=0.8}
          plot_2D.b(x1,y1, 
             xlab= "",  xlim=c(-lim,lim),
             ylab= "" , ylim=c(-lim,lim),
             nbin=100, 
             x_sub_orange=x1[Group2_logical], 
             y_sub_orange=y1[Group2_logical],
             PE=round(pecorr,2),
             x_sub_blue=x1[Group3_logical],
             y_sub_blue=y1[Group3_logical],
             ) 
          mtext(final_df$name[j], side=2) #y
          mtext(final_df$name[i], side=1) #x
          if(type=="Bayenv"){mtext("Bayenv", side=3, adj=0)}
          if(type=="Raw"){mtext("Raw", side=3, adj=0)}
  
        ### Function for Randomization tests ####
        GetSampleCov <- function(test_set, getNull=TRUE){
          x1_test <- results_pine3[test_set,colz[i]]
          y1_test <- results_pine3[test_set,colz[j]]
          info1_test <- info[test_set,]
          
          info <- data.frame(info1, x1, y1)
          info_test <- data.frame(info1_test, x1_test, y1_test)
          info_test <- info_test[complete.cases(info_test),]
          # REASSIGNMENT HERE TO REMOVE ALL SNPS FROM THE 
          # SAME CONTIG EVERY TIME A SNP IS CHOSEN FOR THE NEXT STEP
          ### sample test SNPs
          rand_test_set <- info_test %>% group_by(gtcontig) %>% sample_n(size = 1)
          ### unit test
          if(length(unique(info_test$gtcontig)) != nrow(rand_test_set)){
            print("Error")
            break
          } #end if 
          test_cov <- cov(rand_test_set$x1_test, rand_test_set$y1_test)
          
          if(getNull==TRUE){
            snp_mat <- matrix(NA, nrow= nrow(rand_test_set), ncol=2)
              # 1st column match SNP x
              # 2nd column match SNP y
            
              #get matching SNPs
            for (k in 1:nrow(rand_test_set)){
              test_maf <- rand_test_set$minor_freq[k]
              b <- info %>% filter((minor_freq < (test_maf+0.025)) & (minor_freq > test_maf-0.025)) %>% sample_n(size = 1)
              snp_mat[k,] <- c(b$x1, b$y1)
              # could update info here by removing columns with gtcontig in b
              
              #info <- info %>% filter(gtcontig!=b$gtcontig)
                # ensure not to sample from same contig twice
                # this line of code slows down 2x
                # removed for first initial run, can add back in later
            }#end for loop
            null_cov <- cov(snp_mat[,1], snp_mat[,2])
          }else{
            null_cov=NA
          }
          return(c(test_cov=test_cov, null_cov=null_cov))
        }
        ### End Function for Randomization tests ####
        
        #test_set
        #replicate(10, GetSampleCov(Group2_logical))
        
        ### Do Randomization tests ####
        compare_Grp2 <-  data.frame(t(replicate(100, GetSampleCov(Group2_logical))))
        compare_Grp3 <-  data.frame(t(replicate(100, GetSampleCov(Group3_logical))))
        compare_SuperSnpsNot2or3 <- data.frame(t(replicate(100, GetSampleCov(Group1_logical & !Group2_logical & !Group3_logical, getNull=FALSE))))
          #~1 sec per replicate
          # 
        (max_x <- max(abs(unlist(c(compare_Grp2, compare_Grp3, compare_SuperSnpsNot2or3))), na.rm=TRUE))*1.1
        
        compare_Grp2 <- compare_Grp2[complete.cases(compare_Grp2),]
        compare_Grp3 <- compare_Grp3[complete.cases(compare_Grp3),]
        
        ### Plot Randomization tests ####
        g2_nden <- density(compare_Grp2$null_cov)
        g3_nden <- density(compare_Grp3$null_cov)
        
        g2_tden <- density(compare_Grp2$test_cov)
        g3_tden <- density(compare_Grp3$test_cov)
        
        g0_den <- density(compare_SuperSnpsNot2or3$test_cov)
        
        (max_y <- max(c(g2_nden$y, g3_nden$y, g2_tden$y, g3_tden$y, g0_den$y)))
        
        plot(density(compare_Grp2$null_cov), main="", xlab="Covariance",
             xlim=c(-max_x, max_x), ylim=c(0,max_y),
             col=rgb(1,0,1,0.4), lwd=1, bty='n', yaxt="n")
        
        lines(density(compare_Grp2$test_cov),
               col="magenta", lwd=2, type="l")
        
        lines(density(compare_Grp3$null_cov),
              col=rgb(0,0,1,0.4), lwd=1)
        
        lines(density(compare_Grp3$test_cov),
               col="blue", lwd=2, type="l")
        
        lines(g0_den,
               col=rgb(0,0,0,0.2), lwd=1, type="l")
        ### End Plot Randomization tests ####
        
        ### Significance of Randomization tests ####
        g2_nq <- quantile(compare_Grp2$null_cov, c(0.025, 0.975), type=1)
        g2_tq <- quantile(compare_Grp2$test_cov, c(0.025, 0.975), type=1)  
        if(mean(compare_Grp2$test_cov)>mean(compare_Grp2$null_cov)){
          g2_qtest <- g2_nq[2] < g2_tq[1]
        }
        if(mean(compare_Grp2$test_cov)<mean(compare_Grp2$null_cov)){
          g2_qtest <- g2_nq[1] > g2_tq[2]
        }
        
        g3_nq <- quantile(compare_Grp3$null_cov, c(0.025, 0.975), type=1)
        g3_tq <- quantile(compare_Grp3$test_cov, c(0.025, 0.975), type=1)  
        if(mean(compare_Grp3$test_cov)>mean(compare_Grp3$null_cov)){
          (g3_qtest <- g3_nq[2] < g3_tq[1])
        }
        if(mean(compare_Grp3$test_cov)<mean(compare_Grp3$null_cov)){
          g3_qtest <- g3_nq[1] > g3_tq[2]
        }      
        
        g2_ttest <- t.test(compare_Grp2$test_cov, compare_Grp2$null_cov, paired=TRUE)
        g3_ttest <- t.test(compare_Grp3$test_cov, compare_Grp3$null_cov, paired=TRUE)
      
        plotText <- paste("Temp P<e",round(log(g2_ttest$p.value,10)),"\n",
                  "Temp quant. test ", g2_qtest, "\n",
                  "Precip P<e", round(log(g3_ttest$p.value,10)), "\n",
                  "Precip quant. test ", g3_qtest, sep="")
        
        text(-max_x,max_y, labels=plotText, cex=0.7, adj=c(0,1))
          # Add significance to Plot
    
      }#end loop through type
    dev.off()
      }# end if (i != j)
    }# end loop through i
  }# end loop through j
```







### Plot GCTA effect sizes

```{r}
pdf("GCTA_effect_graphs.pdf")

### Compare effect sizes and p-values
x1 <- -log(results_pine3$LAT_p_GCTA,10)
y1 <- results_pine3$LAT_snp_effect_GCTA
plot_2D.b(x1, y1, xlab= "LAT gcta -log10 P-value",  
              ylab= "LAT gcta effect size" , xlim=c(0,10), ylim=c(-5,5),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

x1 <- -log(results_pine3$MAT_p_GCTA,10)
y1 <- results_pine3$MAT_snp_effect_GCTA
plot_2D.b(x1, y1, xlab= "MAT gcta -log10 p-value",  
              ylab= "MAT gcta effect size" , xlim=c(0,10), ylim=c(-5,5),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

x1 <- -log(results_pine3$MAP_p_GCTA,10)
y1 <- scale(results_pine3$MAP_snp_effect_GCTA)
plot_2D.b(x1, y1, xlab= "MAP gcta -log10 p-value",  
              ylab= "MAP gcta effect size" , xlim=c(0,8), ylim=c(-10,10),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

x1 <- -log(results_pine3$Fall_cold_injury_p,10)
y1 <- results_pine3$Fall_cold_injury_snp_effect
plot_2D.b(x1, y1, xlab= "FCI gcta -log10 P-value",  
              ylab= "FCI gcta snp effect" , xlim=c(0,10), ylim=c(-5,5),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 


x1 <- -log(results_pine3$Winter_cold_injury_p,10)
y1 <- results_pine3$Winter_cold_injury_snp_effect
plot_2D.b(x1, y1, xlab= "WCI gcta -log10 P-value",  
              ylab= "WCI gcta effect" , xlim=c(0,10), ylim=c(-5,5),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

x1 <- -log(results_pine3$Spring_cold_injury_p,10)
y1 <- results_pine3$Spring_cold_injury_snp_effect
plot_2D.b(x1, y1, xlab= "SCI gcta -log10 P-value",  
              ylab= "SCI gcta effect" , xlim=c(0,10), ylim=c(-5,5),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

### Compare MAT and MAP

x1 <- -log(results_pine3$MAT_p_GCTA,10)
y1 <- -log(results_pine3$MAP_p_GCTA,10)
plot_2D.b(x1, y1, xlab= "MAT gcta -log10 P-value",  
              ylab= "MAP gcta -log10 P-value" , xlim=c(0,8), ylim=c(0,8),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 


x1 <- scale(results_pine3$MAT_snp_effect_GCTA)
y1 <- scale(results_pine3$MAP_snp_effect_GCTA)
plot_2D.b(x1, y1, xlab= "MAT snp effect gcta",  
              ylab= "MAP snp effect gcta" , xlim=c(-15,15), ylim=c(-15,15),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

### Compare MAT and FCI

x1 <- -log(results_pine3$MAT_p_GCTA,10)
y1 <- -log(results_pine3$Fall_cold_injury_p,10)
plot_2D.b(x1, y1, xlab= "MAT gcta -log10 P-value",  
              ylab= "FCI gcta -log10 P-value" , xlim=c(0,8), ylim=c(0,8),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

x1 <- scale(results_pine3$MAT_snp_effect_GCTA)
y1 <- scale(results_pine3$Fall_cold_injury_snp_effect)
plot_2D.b(x1, y1, xlab= "MAT snp effect gcta",  
              ylab= "FCI snp effect gcta" , xlim=c(-15,15), ylim=c(-15,15),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

x1 <- results_pine3$MAT_rhoave
y1 <- results_pine3$Fall_cold_injury_rhoave_BAYENV
plot_2D.b(x1, y1, xlab= "MAT bayenv rho",  
              ylab= "FCI bayenv rho" , #xlim=c(0,8), ylim=c(0,8),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 


### Compare FCI and MAP

x1 <- -log(results_pine3$MAP_p_GCTA,10)
y1 <- -log(results_pine3$Fall_cold_injury_p,10)
plot_2D.b(x1, y1, xlab= "MAP gcta -log10 P-value",  
              ylab= "FCI gcta -log10 P-value" , xlim=c(0,8), ylim=c(0,8),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 

x1 <- scale(results_pine3$MAP_snp_effect_GCTA)
y1 <- scale(results_pine3$Fall_cold_injury_snp_effect)
plot_2D.b(x1, y1, xlab= "MAP snp effect gcta",  
              ylab= "FCI snp effect gcta" , xlim=c(-15,15), ylim=c(-15,15),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 


x1 <- results_pine3$MAP_rhoave
y1 <- results_pine3$Fall_cold_injury_rhoave_BAYENV
plot_2D.b(x1, y1, xlab= "MAP bayenv rho",  
              ylab= "FCI bayenv rho" , #xlim=c(0,8), ylim=c(0,8),
              nbin=100, 
              x_sub_orange=x1[Group2_logical], 
              y_sub_orange=y1[Group2_logical],
              PE=1,
              x_sub_blue=x1[Group3_logical],
              y_sub_blue=y1[Group3_logical],
              ) 
dev.off()
```

### Cross-check with orthologs
Checked whether the genes that you have found as being in the 2 groups in your galaxy plots are also found in the ortholog list? If not, that would explain why they have so few genes that are also convergence candidates. 

```{r}
head(results_pine3)[,1:10]
tail(results_pine3)[,1:10]

orthos <- read.table("../data/all_orthologs_analysed.txt", header=TRUE)
head(orthos)
op <- as.character(orthos$pine)
head(op)

### Total number of orthologs
length(op)

### Total number of orthologs in results_pine3
length(which(op %in% results_pine3$tcontig))

### Total number of contigs in top SNPs
length(levels(factor(results_pine3$tcontig[Group1_logical])))
### Total number of orthologs in all top SNPs
length(which(op %in% results_pine3$tcontig[Group1_logical]))

### Total number of contigs in temperature group
length(levels(factor(results_pine3$tcontig[Group2_logical])))
### Total number of orthologs in temperature group
length(which(op %in% results_pine3$tcontig[Group2_logical]))

### Total number of contigs in precipitation group
length(levels(factor(results_pine3$tcontig[Group3_logical])))
### Total number of orthologs in precipitation group
length(which(op %in% results_pine3$tcontig[Group3_logical]))

### Total number of super convergent outliers 
length(levels(factor(results_pine3$tcontig[results_pine3$pine_super_convergent])))

```