---
title: "Untitled"
author: "Katie Lotterhos"
date: "January 19, 2016"
output: html_document
---
Now that all the scores have been calculated, we can look at the relationship between the PE association and the underlying genome-wide pattern.  I expect that higher/larger PEs will have higher scores.  The code plots and summarizes the results of a linear model 
### read in datafram with scores
```{r}
setwd("/Users/katie/Desktop/CurrResearch/1-AdaptreeData/201509_PEGA/analysis")
library(tiff)
 
PE2<-read.table(file = "../data/EnvironmentPhenotypeCorrelations-PineSpruce-spearman_v2.WITHSCORES.txt", header=TRUE)
head(PE2)
PE2$pine_score1_bayenv <- PE2$pine_score1_bayenv*1000
PE2$spruce_score1_bayenv <- PE2$spruce_score1_bayenv*1000

phens <- levels(PE2$Phenotype)

#SPECIFY
infolder <- "figure_new_noaxes_highlightLatFallColdInj/"
files <- list.files(infolder)
outfolder <- "figure_PEGA_highlightLatFallColdInj/"

```

```{r}

for (j in seq_along(phens)){
    print(j)
    ## subset dataframe to get just this Phenotype
    PE_sub <- PE2[PE2$Phenotype==phens[j],]
    if(!(is.na(sum(PE_sub$pine_score1_bayenv)))){
    PE_pheno_name <- as.character(phens[j])
    files_sub <- files[grep(PE_pheno_name,files)]
  ### pine
    m <- max(c(PE_sub$pine_score1_bayenv,PE_sub$spruce_score1_bayenv))
    m_min <- min(c(PE_sub$pine_score1_bayenv,PE_sub$spruce_score1_bayenv))
  m_range <- m-m_min
  mx <- max(c(abs(PE_sub$pine_correlation),abs(PE_sub$spruce_correlation)))
  pdf(paste(outfolder,PE_pheno_name,".pdf", sep=""), width = 7, height=14,bg="transparent")    
    par(mfrow=c(2,1), mar=c(3,3,2,1), oma=c(2,2,2,0))
    scale <- 10
  ### PINE #####
  ##############
    plot(c(0-mx/10,mx+mx/10),c(m_min-m_range/10,m+m_range/10), type='n', bty="n", main="Pine", las=1)
  ### Match names
  for (i in 1:nrow(PE_sub)){
    PE_envi_name <- as.character(PE_sub$Environment[i])
    ### pine
    f_pat <- paste("*", PE_pheno_name,"__", PE_envi_name,"pine.tiff",sep="")
    f1 <- gsub("\\(","\\\\\\(", f_pat)
    f2 <- gsub("\\)","\\\\\\)", f1)
    f_pat2 <- f2
    f <- files[grep(f_pat2, files)]
    img <- readTIFF(paste(infolder,f, sep=""))
    xleft <- abs(PE_sub$pine_correlation[i])-mx/scale
    xright <- abs(PE_sub$pine_correlation[i])+mx/scale
    ybottom <- PE_sub$pine_score1_bayenv[i]-m_range/scale
    ytop <- PE_sub$pine_score1_bayenv[i]+m_range/scale
      transparent <- img[,,4] == 0
      img <- as.raster(img[,,1:3])
      img[transparent] <- NA
    rasterImage(img, xleft, ybottom, xright, ytop, interpolate=TRUE)
  }
  abline(lm(PE_sub$pine_score1_bayenv~abs(PE_sub$pine_correlation)))
  
  ### SPRUCE #####
  ##############
    plot(c(0-mx/10,mx+mx/10),c(m_min-m/10,m+m/10), type='n', bty="n", main="Spruce", las=1)
  ### Match names
  for (i in 1:nrow(PE_sub)){
    PE_envi_name <- as.character(PE_sub$Environment[i])
    ### spruce
    f_pat <- paste("*", PE_pheno_name,"__", PE_envi_name,"spruce.tiff",sep="")
    f1 <- gsub("\\(","\\\\\\(", f_pat)
    f2 <- gsub("\\)","\\\\\\)", f1)
    f_pat2 <- f2
    f <- files[grep(f_pat2, files)]
    img <- readTIFF(paste(infolder,f, sep=""))
    xleft <- abs(PE_sub$spruce_correlation[i])-mx/scale
    xright <- abs(PE_sub$spruce_correlation[i])+mx/scale
    ybottom <- PE_sub$spruce_score1_bayenv[i]-m_range/scale
    ytop <- PE_sub$spruce_score1_bayenv[i]+m_range/scale
      transparent <- img[,,4] == 0
      img <- as.raster(img[,,1:3])
      img[transparent] <- NA
    rasterImage(img, xleft, ybottom, xright, ytop, interpolate=TRUE)
  }
  abline(lm(PE_sub$spruce_score1_bayenv~abs(PE_sub$spruce_correlation)))

  mtext("Phenotype correlation with environment",side = 1, outer=TRUE)
  mtext(PE_pheno_name,side = 3, outer=TRUE)
  mtext("Genome Score",side = 2, outer=TRUE)
  dev.off()
  }#end if not NA
}#end for loop
```    


Next chunk of code needs to be rewritten...

```{r, label="PE score analysis"}
par(mfcol=c(3,2), mar=c(3,3,0,0), oma=c(0,0,0,0))
plot(abs(PE$pine_correlation), PE$pine_score1_bayenv)
abline(0,0)
plot(abs(PE$pine_correlation), PE$pine_score2_bayenv)
abline(0,0)
plot(abs(PE$pine_correlation), PE$pine_slope_bayenv)
abline(0,0)
plot(abs(PE$spruce_correlation), PE$spruce_score1_bayenv)
plot(abs(PE$spruce_correlation), PE$spruce_score2_bayenv)
abline(0,0)
plot(abs(PE$spruce_correlation), PE$spruce_slope_bayenv)
abline(0,0)



head(PE)
  ylim=c(-70,70)

plotter_score <- function(x,y, xlab,ylab, ymin, ymax, data){
  plot(x, y,  ylim=c(ymin,ymax), 
       col=as.numeric(data$Environment), pch=as.numeric(data$Environment),  yaxt="n", xlim=c(0, max(x)),
       xlab=xlab, ylab=ylab, las=1, cex.axis=0.6)
  #m1 <- lm(y~x + data$Environment + data$Phenotype)
  #print(summary(m1))
  abline(lm(y~x))
  abline(0,0, col="grey", lwd=2)
  abline(0.5,0, col="grey", lwd=2)
}

png(paste("../analysis/Score1_Pine_Spruce_Envi.png"), width=10, height=10, units="in", res=300)
 par(mfrow=c(2,1), mar=c(4,4,1,1), oma=c(0,0,2,0))
  ### All results
  plotter_score(abs(PE$pine_correlation), PE$pine_score1_bayenv, 
                "PINE Abs(Pheno-Envi) Correlation", 
                "PINE Score 1 ALL SNPs", -0.0003,0.0003)
  plotter_score(abs(PE$spruce_correlation), PE$spruce_score1_bayenv, 
                "SPRUCE Abs(Pheno-Envi) Correlation", 
                "SPRUCE Score 1 ALL SNPs", -0.00035,0.00035)
dev.off()


png(paste("../analysis/Score1_Pine_Spruce_Envi_byPheno.png"), width=11, height=3, units="in", res=300)
phen <- levels(PE$Phenotype)
par(mfcol=c(2, 15), mar=c(2,0.5,1,0.5), oma=c(2,2,0,0))
for (j in seq_along(phen)){
  sub <- PE[PE$Phenotype==phen[j],]
  if(sum(!is.na(sub$pine_score1_bayenv))>0){
  plotter_score(abs(sub$pine_correlation), sub$pine_score1_bayenv, 
                "", "", -0.00035,0.00035, sub)
  text(0,0.00035, phen[j], cex=0.4,adj=0)
  plotter_score(abs(sub$spruce_correlation), sub$spruce_score1_bayenv, 
                "SPRUCE Abs(Pheno-Envi) Correlation", 
                "SPRUCE Score 1 ALL SNPs", -0.00035,0.00035, sub)
  text(0,0.00035, phen[j], cex=0.4,adj=0)
  }
}
mtext("Phenotype-Environment Correlation", side=1, outer=TRUE)
mtext("(Spruce) GEA/GWAS Score (Pine)", side=2, outer=TRUE)
dev.off()

png(paste("../analysis/Score1_Pine_Spruce_Envi_byPheno_legend.png"), width=3, height=6, units="in", res=300)
  envi<- levels(PE$Environment)
  par(mfrow=c(1,1))
  plot(rep(1, length(envi)), 1:length(envi), col=1:length(envi), pch=1:length(envi))
  text(rep(1.03, length(envi)),1:length(envi) ,labels= envi, adj=0)
dev.off()
```

```
#plot allsnps vs super, expect super to be above 1:1 line

# color by environment is different than by phenotype

plot(PE$pine_score_super_t, PE$spruce_score_super_t,col=as.numeric(PE$Environment), pch=as.numeric(PE$Environment))
```